{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "toc_visible": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/DCEG-workshops/statgen_workshop_tutorial/blob/main/src/02_qc.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Setting up a few things"
      ],
      "metadata": {
        "id": "SHxtxszu9sk6"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "_3pFsIaX9w-b"
      },
      "outputs": [],
      "source": [
        "#load python libraries\n",
        "import os"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#load R magic\n",
        "%load_ext rpy2.ipython"
      ],
      "metadata": {
        "id": "wZppOK6ki7dx"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#mount google drive\n",
        "from google.colab import drive\n",
        "drive.mount('/content/drive/')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Wl2agaRa6hOd",
        "outputId": "491485af-b953-48ac-8760-dd6903ce3def"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive/\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#take a look at the files\n",
        "%ls drive/MyDrive/statgen_workshop/data/workshop1/inputs/penncath_withPheno/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fzFXMpJ383QJ",
        "outputId": "ff59261c-2682-4b90-9214-d0d6b6851893"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "causal_variants.txt  penncath.bim  plink_fix.log  run_plink.sh   sim_pheno.phen\n",
            "gcta_sim.log         penncath.fam  run_gcta.log   sim_pheno.log\n",
            "penncath.bed         penncath.log  run_gcta.sh    sim_pheno.par\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#take a look at my path\n",
        "%%bash\n",
        "echo $PATH\n",
        "ls /tools/node/bin\n",
        "touch /tools/node/bin/test.txt\n",
        "ls /tools/node/bin"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2cLZ4_SkfG5_",
        "outputId": "b5ce9c4a-ee28-4c93-bfd6-2f31820602ce"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/opt/bin:/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/tools/node/bin:/tools/google-cloud-sdk/bin\n",
            "node\n",
            "npm\n",
            "npx\n",
            "test.txt\n",
            "node\n",
            "npm\n",
            "npx\n",
            "test.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#download plink1.9\n",
        "%%bash\n",
        "curl -o /tools/node/bin/plink.zip https://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20230116.zip\n",
        "cd /tools/node/bin/\n",
        "unzip plink.zip"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Kh0Eyw5Mpvnw",
        "outputId": "11471ddd-3225-4a38-b47f-4bf9c6531ac0"
      },
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Archive:  plink.zip\n",
            "  inflating: plink                   \n",
            "  inflating: LICENSE                 \n",
            "  inflating: toy.ped                 \n",
            "  inflating: toy.map                 \n",
            "  inflating: prettify                \n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r  0 8743k    0 53756    0     0  41107      0  0:03:37  0:00:01  0:03:36 41097\r 25 8743k   25 2223k    0     0   999k      0  0:00:08  0:00:02  0:00:06  999k\r100 8743k  100 8743k    0     0  3340k      0  0:00:02  0:00:02 --:--:-- 3339k\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#see if we can execute plink\n",
        "%%bash\n",
        "plink --help"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1udPXdDjp7g_",
        "outputId": "db962c3e-ec40-4732-c301-e29f31be4f3a"
      },
      "execution_count": 40,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "\n",
            "In the command line flag definitions that follow,\n",
            "  * <angle brackets> denote a required parameter, where the text between the\n",
            "    angle brackets describes its nature.\n",
            "  * ['square brackets + single-quotes'] denotes an optional modifier.  Use the\n",
            "    EXACT text in the quotes.\n",
            "  * [{bar|separated|braced|bracketed|values}] denotes a collection of mutually\n",
            "    exclusive optional modifiers (again, the exact text must be used).  When\n",
            "    there are no outer square brackets, one of the choices must be selected.\n",
            "  * ['quoted_text='<description of value>] denotes an optional modifier that\n",
            "    must begin with the quoted text, and be followed by a value with no\n",
            "    whitespace in between.  '|' may also be used here to indicate mutually\n",
            "    exclusive options.\n",
            "  * [square brackets without quotes or braces] denote an optional parameter,\n",
            "    where the text between the brackets describes its nature.\n",
            "  * An ellipsis (...) indicates that you may enter multiple parameters of the\n",
            "    specified type.\n",
            "\n",
            "  plink <input flag(s)...> [command flag(s)...] [other flag(s)...]\n",
            "  plink --help [flag name(s)...]\n",
            "\n",
            "Most PLINK runs require exactly one main input fileset.  The following flags\n",
            "are available for defining its form and location:\n",
            "\n",
            "  --bfile [prefix] : Specify .bed + .bim + .fam prefix (default 'plink').\n",
            "  --bed <filename> : Specify full name of .bed file.\n",
            "  --bim <filename> : Specify full name of .bim file.\n",
            "  --fam <filename> : Specify full name of .fam file.\n",
            "\n",
            "  --keep-autoconv  : With --file/--tfile/--lfile/--vcf/--bcf/--data/--23file,\n",
            "                     don't delete autogenerated binary fileset at end of run.\n",
            "\n",
            "  --file [prefix]  : Specify .ped + .map filename prefix (default 'plink').\n",
            "  --ped <filename> : Specify full name of .ped file.\n",
            "  --map <filename> : Specify full name of .map file.\n",
            "\n",
            "  --no-fid         : .fam/.ped file does not contain column 1 (family ID).\n",
            "  --no-parents     : .fam/.ped file does not contain columns 3-4 (parents).\n",
            "  --no-sex         : .fam/.ped file does not contain column 5 (sex).\n",
            "  --no-pheno       : .fam/.ped file does not contain column 6 (phenotype).\n",
            "\n",
            "  --tfile [prefix] : Specify .tped + .tfam filename prefix (default 'plink').\n",
            "  --tped <fname>   : Specify full name of .tped file.\n",
            "  --tfam <fname>   : Specify full name of .tfam file.\n",
            "\n",
            "  --lfile [prefix] : Specify .lgen + .map + .fam (long-format fileset) prefix.\n",
            "  --lgen <fname>   : Specify full name of .lgen file.\n",
            "  --reference <fn> : Specify default allele file accompanying .lgen input.\n",
            "  --allele-count   : When used with --lfile/--lgen + --reference, specifies\n",
            "                     that the .lgen file contains reference allele counts.\n",
            "\n",
            "  --vcf <filename> : Specify full name of .vcf or .vcf.gz file.\n",
            "  --bcf <filename> : Specify full name of BCF2 file.\n",
            "\n",
            "  --data [prefix]  : Specify Oxford .gen + .sample prefix (default 'plink').\n",
            "  --gen <filename> : Specify full name of .gen or .gen.gz file.\n",
            "  --bgen <f> ['snpid-chr'] : Specify full name of .bgen file.\n",
            "  --sample <fname> : Specify full name of .sample file.\n",
            "\n",
            "  --23file <fname> [FID] [IID] [sex] [pheno] [pat. ID] [mat. ID] :\n",
            "    Specify 23andMe input file.\n",
            "\n",
            "  --grm-gz [prfx]  : Specify .grm.gz + .grm.id (GCTA rel. matrix) prefix.\n",
            "  --grm-bin [prfx] : Specify .grm.bin + .grm.N.bin + .grm.id (GCTA triangular\n",
            "                     binary relationship matrix) filename prefix.\n",
            "\n",
            "  --dummy <sample ct> <SNP ct> [missing geno freq] [missing pheno freq]\n",
            "          [{acgt | 1234 | 12}] ['scalar-pheno']\n",
            "    This generates a fake input dataset with the specified number of samples\n",
            "    and SNPs.  By default, the missing genotype and phenotype frequencies are\n",
            "    zero, and genotypes are As and Bs (change the latter with\n",
            "    'acgt'/'1234'/'12').  The 'scalar-pheno' modifier causes a normally\n",
            "    distributed scalar phenotype to be generated instead of a binary one.\n",
            "\n",
            "  --simulate <simulation parameter file> [{tags | haps}] [{acgt | 1234 | 12}]\n",
            "  --simulate-qt <sim. parameter file> [{tags | haps}] [{acgt | 1234 | 12}]\n",
            "    --simulate generates a fake input dataset with disease-associated SNPs,\n",
            "    while --simulate-qt generates a dataset with quantitative trait loci.\n",
            "\n",
            "Output files have names of the form 'plink.<extension>' by default.  You can\n",
            "change the 'plink' prefix with\n",
            "\n",
            "  --out <prefix>   : Specify prefix for output files.\n",
            "\n",
            "Most runs also require at least one of the following commands:\n",
            "\n",
            "  --make-bed\n",
            "    Create a new binary fileset.  Unlike the automatic text-to-binary\n",
            "    converters (which only heed chromosome filters), this supports all of\n",
            "    PLINK's filtering flags.\n",
            "  --make-just-bim\n",
            "  --make-just-fam\n",
            "    Variants of --make-bed which only write a new .bim or .fam file.  Can be\n",
            "    used with only .bim/.fam input.\n",
            "    USE THESE CAUTIOUSLY.  It is very easy to desynchronize your binary\n",
            "    genotype data and your .bim/.fam indexes if you use these commands\n",
            "    improperly.  If you have any doubt, stick with --make-bed.\n",
            "\n",
            "  --recode <output format> [{01 | 12}] [{tab | tabx | spacex | bgz | gen-gz}]\n",
            "           ['include-alt'] ['omit-nonmale-y']\n",
            "    Create a new text fileset with all filters applied.  The following output\n",
            "    formats are supported:\n",
            "    * '23': 23andMe 4-column format.  This can only be used on a single\n",
            "      sample's data (--keep may be handy), and does not support multicharacter\n",
            "      allele codes.\n",
            "    * 'A': Sample-major additive (0/1/2) coding, suitable for loading from R.\n",
            "      If you need uncounted alleles to be named in the header line, add the\n",
            "      'include-alt' modifier.\n",
            "    * 'AD': Sample-major additive (0/1/2) + dominant (het=1/hom=0) coding.\n",
            "      Also supports 'include-alt'.\n",
            "    * 'A-transpose': Variant-major 0/1/2.\n",
            "    * 'beagle': Unphased per-autosome .dat and .map files, readable by early\n",
            "      BEAGLE versions.\n",
            "    * 'beagle-nomap': Single .beagle.dat file.\n",
            "    * 'bimbam': Regular BIMBAM format.\n",
            "    * 'bimbam-1chr': BIMBAM format, with a two-column .pos.txt file.  Does not\n",
            "      support multiple chromosomes.\n",
            "    * 'fastphase': Per-chromosome fastPHASE files, with\n",
            "      .chr-<chr #>.recode.phase.inp filename extensions.\n",
            "    * 'fastphase-1chr': Single .recode.phase.inp file.  Does not support\n",
            "      multiple chromosomes.\n",
            "    * 'HV': Per-chromosome Haploview files, with .chr-<chr #>{.ped,.info}\n",
            "      filename extensions.\n",
            "    * 'HV-1chr': Single Haploview .ped + .info file pair.  Does not support\n",
            "      multiple chromosomes.\n",
            "    * 'lgen': PLINK 1 long-format (.lgen + .fam + .map), loadable with --lfile.\n",
            "    * 'lgen-ref': .lgen + .fam + .map + .ref, loadable with --lfile +\n",
            "       --reference.\n",
            "    * 'list': Single genotype-based list, up to 4 lines per variant.  To omit\n",
            "      nonmale genotypes on the Y chromosome, add the 'omit-nonmale-y' modifier.\n",
            "    * 'rlist': .rlist + .fam + .map fileset, where the .rlist file is a\n",
            "      genotype-based list which omits the most common genotype for each\n",
            "      variant.  Also supports 'omit-nonmale-y'.\n",
            "    * 'oxford': Oxford-format .gen + .sample.  With the 'gen-gz' modifier, the\n",
            "      .gen file is gzipped.\n",
            "    * 'ped': PLINK 1 sample-major (.ped + .map), loadable with --file.\n",
            "    * 'compound-genotypes': Same as 'ped', except that the space between each\n",
            "      pair of same-variant allele codes is removed.\n",
            "    * 'structure': Structure-format.\n",
            "    * 'transpose': PLINK 1 variant-major (.tped + .tfam), loadable with\n",
            "      --tfile.\n",
            "    * 'vcf', 'vcf-fid', 'vcf-iid': VCFv4.2.  'vcf-fid' and 'vcf-iid' cause\n",
            "      family IDs or within-family IDs respectively to be used for the sample\n",
            "      IDs in the last header row, while 'vcf' merges both IDs and puts an\n",
            "      underscore between them.  If the 'bgz' modifier is added, the VCF file is\n",
            "      block-gzipped.\n",
            "      The A2 allele is saved as the reference and normally flagged as not based\n",
            "      on a real reference genome (INFO:PR).  When it is important for reference\n",
            "      alleles to be correct, you'll also want to include --a2-allele and\n",
            "      --real-ref-alleles in your command.\n",
            "    In addition,\n",
            "    * The '12' modifier causes A1 (usually minor) alleles to be coded as '1'\n",
            "      and A2 alleles to be coded as '2', while '01' maps A1 -> 0 and A2 -> 1.\n",
            "    * The 'tab' modifier makes the output mostly tab-delimited instead of\n",
            "      mostly space-delimited.  'tabx' and 'spacex' force all tabs and all\n",
            "      spaces, respectively.\n",
            "\n",
            "  --flip-scan ['verbose']\n",
            "    (alias: --flipscan)\n",
            "    LD-based scan for case/control strand inconsistency.\n",
            "\n",
            "  --write-covar\n",
            "    If a --covar file is loaded, --make-bed/--make-just-fam and --recode\n",
            "    automatically generate an updated version (with all filters applied).\n",
            "    However, if you do not wish to simultaneously generate a new genotype file,\n",
            "    you can use --write-covar to just produce a pruned covariate file.\n",
            "\n",
            "  --write-cluster ['omit-unassigned']\n",
            "    If clusters are specified with --within/--family, this generates a new\n",
            "    cluster file (with all filters applied).  The 'omit-unassigned' modifier\n",
            "    causes unclustered samples to be omitted from the file; otherwise their\n",
            "    cluster is 'NA'.\n",
            "\n",
            "  --write-set\n",
            "  --set-table\n",
            "    If sets have been defined, --write-set dumps 'END'-terminated set\n",
            "    membership lists to <output prefix>.set, while --set-table writes a\n",
            "    variant-by-set membership table to <output prefix>.set.table.\n",
            "\n",
            "  --merge <.ped filename> <.map filename>\n",
            "  --merge <text fileset prefix>\n",
            "  --bmerge <.bed filename> <.bim filename> <.fam filename>\n",
            "  --bmerge <binary fileset prefix>\n",
            "    Merge the given fileset with the initially loaded fileset, writing the\n",
            "    result to <output prefix>.bed + .bim + .fam.  (It is no longer necessary to\n",
            "    simultaneously specify --make-bed.)\n",
            "  --merge-list <filename>\n",
            "    Merge all filesets named in the text file with the reference fileset, if\n",
            "    one was specified.  (However, this can also be used *without* a reference;\n",
            "    in that case, the newly created fileset is then treated as the reference by\n",
            "    most other PLINK operations.)  The text file is interpreted as follows:\n",
            "    * If a line contains only one name, it is assumed to be the prefix for a\n",
            "      binary fileset.\n",
            "    * If a line contains exactly two names, they are assumed to be the full\n",
            "      filenames for a text fileset (.ped first, then .map).\n",
            "    * If a line contains exactly three names, they are assumed to be the full\n",
            "      filenames for a binary fileset (.bed, then .bim, then .fam).\n",
            "\n",
            "  --write-snplist\n",
            "  --list-23-indels\n",
            "    --write-snplist writes a .snplist file listing the names of all variants\n",
            "    which pass the filters and inclusion thresholds you've specified, while\n",
            "    --list-23-indels writes the subset with 23andMe-style indel calls (D/I\n",
            "    allele codes).\n",
            "\n",
            "  --list-duplicate-vars ['require-same-ref'] ['ids-only'] ['suppress-first']\n",
            "    --list-duplicate-vars writes a .dupvar file describing all groups of\n",
            "    variants with matching positions and allele codes.\n",
            "    * By default, A1/A2 allele assignments are ignored; use 'require-same-ref'\n",
            "      to override this.\n",
            "    * Normally, the report contains position and allele codes.  To remove them\n",
            "      (and produce a file directly usable with e.g. --extract/--exclude), use\n",
            "      'ids-only'.  Note that this command will fail in 'ids-only' mode if any\n",
            "      of the reported IDs are not unique.\n",
            "    * 'suppress-first' causes the first variant ID in each group to be omitted\n",
            "      from the report.\n",
            "\n",
            "  --freq [{counts | case-control}] ['gz']\n",
            "  --freqx ['gz']\n",
            "    --freq generates a basic allele frequency (or count, if the 'counts'\n",
            "    modifier is present) report.  This can be combined with --within/--family\n",
            "    to produce a cluster-stratified allele frequency/count report instead, or\n",
            "    the 'case-control' modifier to report case and control allele frequencies\n",
            "    separately.\n",
            "    --freqx generates a more detailed genotype count report, designed for use\n",
            "    with --read-freq.\n",
            "\n",
            "  --missing ['gz']\n",
            "    Generate sample- and variant-based missing data reports.  If clusters are\n",
            "    defined, the variant-based report is cluster-stratified.  'gz' causes the\n",
            "    output files to be gzipped.\n",
            "    Unlike most other commands, this doesn't treat het. haploids as missing.\n",
            "\n",
            "  --test-mishap\n",
            "    Check for association between missing calls and flanking haplotypes.\n",
            "\n",
            "  --hardy ['midp'] ['gz']\n",
            "    Generate a Hardy-Weinberg exact test p-value report.  (This does NOT\n",
            "    simultaneously filter on the p-value any more; use --hwe for that.)  With\n",
            "    the 'midp' modifier, the test applies the mid-p adjustment described in\n",
            "    Graffelman J, Moreno V (2013) The mid p-value in exact tests for\n",
            "    Hardy-Weinberg Equilibrium.\n",
            "\n",
            "  --mendel ['summaries-only']\n",
            "    Generate a Mendel error report.  The 'summaries-only' modifier causes the\n",
            "    .mendel file (listing every single error) to be skipped.\n",
            "\n",
            "  --het ['small-sample'] ['gz']\n",
            "  --ibc\n",
            "    Estimate inbreeding coefficients.  --het reports method-of-moments\n",
            "    estimates, while --ibc calculates all three values described in Yang J, Lee\n",
            "    SH, Goddard ME and Visscher PM (2011) GCTA: A Tool for Genome-wide Complex\n",
            "    Trait Analysis.  (That paper also describes the relationship matrix\n",
            "    computation we reimplement.)\n",
            "    * These functions require decent MAF estimates.  If there are very few\n",
            "      samples in your immediate fileset, --read-freq is practically mandatory\n",
            "      since imputed MAFs are wildly inaccurate in that case.\n",
            "    * They also assume the marker set is in approximate linkage equilibrium.\n",
            "    * By default, --het omits the n/(n-1) multiplier in Nei's expected\n",
            "      homozygosity formula.  The 'small-sample' modifier causes it to be\n",
            "      included, while forcing --het to use MAFs imputed from founders in the\n",
            "      immediate dataset.\n",
            "\n",
            "  --check-sex [female max F] [male min F]\n",
            "  --check-sex ycount [female max F] [male min F] [female max Y obs]\n",
            "                     [male min Y obs]\n",
            "  --check-sex y-only [female max Y obs] [male min Y obs]\n",
            "  --impute-sex [female max F] [male min F]\n",
            "  --impute-sex ycount [female max F] [male min F] [female max Y obs]\n",
            "                      [male min Y obs]\n",
            "  --impute-sex y-only [female max Y obs] [male min Y obs]\n",
            "    --check-sex normally compares sex assignments in the input dataset with\n",
            "    those imputed from X chromosome inbreeding coefficients.\n",
            "    * Make sure that the X chromosome pseudo-autosomal region has been split\n",
            "      off (with e.g. --split-x) before using this.\n",
            "    * You also need decent MAF estimates (so, with very few samples in your\n",
            "      immediate fileset, use --read-freq), and your marker set should be in\n",
            "      approximate linkage equilibrium.\n",
            "    * By default, F estimates smaller than 0.2 yield female calls, and values\n",
            "      larger than 0.8 yield male calls.  If you pass numeric parameter(s) to\n",
            "      --check-sex, the first two control these thresholds.\n",
            "    There are now two modes which consider Y chromosome data.\n",
            "    * In 'ycount' mode, gender is still imputed from the X chromosome, but\n",
            "      female calls are downgraded to ambiguous whenever more than 0 nonmissing\n",
            "      Y genotypes are present, and male calls are downgraded when fewer than 0\n",
            "      are present.  (Note that these are counts, not rates.)  These thresholds\n",
            "      are controllable with --check-sex ycount's optional 3rd and 4th numeric\n",
            "      parameters.\n",
            "    * In 'y-only' mode, gender is imputed from nonmissing Y genotype counts.\n",
            "      The male minimum threshold defaults to 1 instead of zero in this case.\n",
            "    --impute-sex changes sex assignments to the imputed values, and is\n",
            "    otherwise identical to --check-sex.  It must be used with\n",
            "    --make-bed/--recode/--write-covar.\n",
            "\n",
            "  --fst ['case-control']\n",
            "    (alias: --Fst)\n",
            "    Estimate Wright's Fst for each autosomal diploid variant using the method\n",
            "    introduced in Weir BS, Cockerham CC (1984) Estimating F-statistics for the\n",
            "    analysis of population structure, given a set of subpopulations defined via\n",
            "    --within.  Raw and weighted global means are also reported.\n",
            "    * If you're interested in the global means, it is usually best to perform\n",
            "      this calculation on a marker set in approximate linkage equilibrium.\n",
            "    * If you have only two subpopulations, you can represent them with\n",
            "      case/control status and use the 'case-control' modifier.\n",
            "\n",
            "  --indep <window size>['kb'] <step size (variant ct)> <VIF threshold>\n",
            "  --indep-pairwise <window size>['kb'] <step size (variant ct)> <r^2 threshold>\n",
            "  --indep-pairphase <window size>['kb'] <step size (variant ct)> <r^2 thresh>\n",
            "    Generate a list of markers in approximate linkage equilibrium.  With the\n",
            "    'kb' modifier, the window size is in kilobase instead of variant count\n",
            "    units.  (Pre-'kb' space is optional, i.e. \"--indep-pairwise 500 kb 5 0.5\"\n",
            "    and \"--indep-pairwise 500kb 5 0.5\" have the same effect.)\n",
            "    Note that you need to rerun PLINK using --extract or --exclude on the\n",
            "    .prune.in/.prune.out file to apply the list to another computation.\n",
            "\n",
            "  --r [{square | square0 | triangle | inter-chr}] [{gz | bin | bin4}]\n",
            "      ['spaces'] ['in-phase'] [{d | dprime | dprime-signed}] ['with-freqs']\n",
            "      ['yes-really']\n",
            "  --r2 [{square | square0 | triangle | inter-chr}] [{gz | bin | bin4}]\n",
            "       ['spaces'] ['in-phase'] [{d | dprime | dprime-signed}] ['with-freqs']\n",
            "       ['yes-really']\n",
            "    LD statistic reports.  --r yields raw inter-variant correlations, while\n",
            "    --r2 reports their squares.  You can request results for all pairs in\n",
            "    matrix format (if you specify 'bin' or one of the shape modifiers), all\n",
            "    pairs in table format ('inter-chr'), or a limited window in table format\n",
            "    (default).\n",
            "    * The 'gz' modifier causes the output text file to be gzipped.\n",
            "    * 'bin' causes the output matrix to be written in double-precision binary\n",
            "      format, while 'bin4' specifics single-precision binary.  The matrix is\n",
            "      square if no shape is explicitly specified.\n",
            "    * By default, text matrices are tab-delimited; 'spaces' switches this.\n",
            "    * 'in-phase' adds a column with in-phase allele pairs to table-formatted\n",
            "      reports.  (This cannot be used with very long allele codes.)\n",
            "    * 'dprime' adds the absolute value of Lewontin's D-prime statistic to\n",
            "      table-formatted reports, and forces both r/r^2 and D-prime to be based on\n",
            "      the maximum likelihood solution to the cubic equation discussed in Gaunt\n",
            "      T, Rodriguez S, Day I (2007) Cubic exact solutions for the estimation of\n",
            "      pairwise haplotype frequencies.\n",
            "      'dprime-signed' keeps the sign, while 'd' skips division by D_{max}.\n",
            "    * 'with-freqs' adds MAF columns to table-formatted reports.\n",
            "    * Since the resulting file can easily be huge, you're required to add the\n",
            "      'yes-really' modifier when requesting an unfiltered, non-distributed all\n",
            "      pairs computation on more than 400k variants.\n",
            "    * These computations can be subdivided with --parallel (even when the\n",
            "      'square' modifier is active).\n",
            "  --ld <variant ID> <variant ID> ['hwe-midp']\n",
            "    This displays haplotype frequencies, r^2, and D' for a single pair of\n",
            "    variants.  When there are multiple biologically possible solutions to the\n",
            "    haplotype frequency cubic equation, all are displayed (instead of just the\n",
            "    maximum likelihood solution identified by --r/--r2), along with HWE exact\n",
            "    test statistics.\n",
            "\n",
            "  --show-tags <filename>\n",
            "  --show-tags all\n",
            "    * If a file is specified, list all variants which tag at least one variant\n",
            "      named in the file.  (This will normally be a superset of the original\n",
            "      list, since a variant is considered to tag itself here.)\n",
            "    * If 'all' mode is specified, for each variant, each *other* variant which\n",
            "      tags it is reported.\n",
            "\n",
            "  --blocks ['no-pheno-req'] ['no-small-max-span']\n",
            "    Estimate haplotype blocks, via Haploview's interpretation of the block\n",
            "    definition suggested by Gabriel S et al. (2002) The Structure of Haplotype\n",
            "    Blocks in the Human Genome.\n",
            "    * Normally, samples with missing phenotypes are not considered by this\n",
            "      computation; the 'no-pheno-req' modifier lifts this restriction.\n",
            "    * Normally, size-2 blocks may not span more than 20kb, and size-3 blocks\n",
            "      are limited to 30kb.  The 'no-small-max-span' modifier removes these\n",
            "      limits.\n",
            "    The .blocks file is valid input for PLINK 1.07's --hap command.  However,\n",
            "    the --hap... family of flags has not been reimplemented in PLINK 1.9 due to\n",
            "    poor phasing accuracy relative to other software; for now, we recommend\n",
            "    using BEAGLE instead of PLINK for case/control haplotype association\n",
            "    analysis.  (You can use \"--recode beagle\" to export data to BEAGLE 3.3.)\n",
            "    We apologize for the inconvenience, and plan to develop variants of the\n",
            "    --hap... flags which handle pre-phased data effectively.\n",
            "\n",
            "  --distance [{square | square0 | triangle}] [{gz | bin | bin4}] ['ibs']\n",
            "             ['1-ibs'] ['allele-ct'] ['flat-missing']\n",
            "    Write a lower-triangular tab-delimited table of (weighted) genomic\n",
            "    distances in allele count units to <output prefix>.dist, and a list of the\n",
            "    corresponding sample IDs to <output prefix>.dist.id.  The first row of the\n",
            "    .dist file contains a single <genome 1-genome 2> distance, the second row\n",
            "    has the <genome 1-genome 3> and <genome 2-genome 3> distances in that\n",
            "    order, etc.\n",
            "    * It is usually best to perform this calculation on a marker set in\n",
            "      approximate linkage equilibrium.\n",
            "    * If the 'square' or 'square0' modifier is present, a square matrix is\n",
            "      written instead; 'square0' fills the upper right triangle with zeroes.\n",
            "    * If the 'gz' modifier is present, a compressed .dist.gz file is written\n",
            "      instead of a plain text file.\n",
            "    * If the 'bin' modifier is present, a binary (square) matrix of\n",
            "      double-precision floating point values, suitable for loading from R, is\n",
            "      instead written to <output prefix>.dist.bin.  ('bin4' specifies\n",
            "      single-precision numbers instead.)  This can be combined with 'square0'\n",
            "      if you still want the upper right zeroed out, or 'triangle' if you don't\n",
            "      want to pad the upper right at all.\n",
            "    * If the 'ibs' modifier is present, an identity-by-state matrix is written\n",
            "      to <output prefix>.mibs.  '1-ibs' causes distances expressed as genomic\n",
            "      proportions (i.e. 1 - IBS) to be written to <output prefix>.mdist.\n",
            "      Combine with 'allele-ct' if you want to generate the usual .dist file as\n",
            "      well.\n",
            "    * By default, distance rescaling in the presence of missing genotype calls\n",
            "      is sensitive to allele count distributions: if variant A contributes, on\n",
            "      average, twice as much to other pairwise distances as variant B, a\n",
            "      missing call at variant A will result in twice as large of a missingness\n",
            "      correction.  To turn this off (because e.g. your missing calls are highly\n",
            "      nonrandom), use the 'flat-missing' modifier.\n",
            "    * The computation can be subdivided with --parallel.\n",
            "  --distance-matrix\n",
            "  --ibs-matrix\n",
            "    These deprecated commands are equivalent to \"--distance 1-ibs flat-missing\n",
            "    square\" and \"--distance ibs flat-missing square\", respectively, except that\n",
            "    they generate space- instead of tab-delimited text matrices.\n",
            "\n",
            "  --make-rel [{square | square0 | triangle}] [{gz | bin | bin4}]\n",
            "             [{cov | ibc2 | ibc3}]\n",
            "    Write a lower-triangular variance-standardized realized relationship matrix\n",
            "    to <output prefix>.rel, and corresponding IDs to <output prefix>.rel.id.\n",
            "    * It is usually best to perform this calculation on a marker set in\n",
            "      approximate linkage equilibrium.\n",
            "    * 'square', 'square0', 'triangle', 'gz', 'bin', and 'bin4' act as they do\n",
            "      on --distance.\n",
            "    * The 'cov' modifier removes the variance standardization step, causing a\n",
            "      covariance matrix to be calculated instead.\n",
            "    * By default, the diagonal elements in the relationship matrix are based on\n",
            "      --ibc's Fhat1; use the 'ibc2' or 'ibc3' modifiers to base them on Fhat2\n",
            "      or Fhat3 instead.\n",
            "    * The computation can be subdivided with --parallel.\n",
            "  --make-grm-gz ['no-gz'] [{cov | ibc2 | ibc3}]\n",
            "  --make-grm-bin [{cov | ibc2 | ibc3}]\n",
            "    --make-grm-gz writes the relationships in GCTA's original gzipped list\n",
            "    format, which describes one pair per line, while --make-grm-bin writes them\n",
            "    in GCTA 1.1+'s single-precision triangular binary format.  Note that these\n",
            "    formats explicitly report the number of valid observations (where neither\n",
            "    sample has a missing call) for each pair, which is useful input for some\n",
            "    scripts.\n",
            "    These computations can be subdivided with --parallel.\n",
            "\n",
            "  --rel-cutoff [val]\n",
            "    (alias: --grm-cutoff)\n",
            "    Exclude one member of each pair of samples with relatedness greater than\n",
            "    the given cutoff value (default 0.025).  If no later operation will cause\n",
            "    the list of remaining samples to be written to disk, this will save it to\n",
            "    <output prefix>.rel.id.\n",
            "    Note that maximizing the remaining sample size is equivalent to the NP-hard\n",
            "    maximum independent set problem, so we use a greedy algorithm instead of\n",
            "    guaranteeing optimality.  (Use the --make-rel and --keep/--remove flags if\n",
            "    you want to try to do better.)\n",
            "\n",
            "  --ibs-test [permutation count]\n",
            "  --groupdist [iters] [d]\n",
            "    Given case/control phenotype data, these commands consider three subsets of\n",
            "    the distance matrix: pairs of affected samples, affected-unaffected pairs,\n",
            "    and pairs of unaffected samples.  Each of these subsets has a distribution\n",
            "    of pairwise genomic distances; --ibs-test uses permutation to estimate\n",
            "    p-values re: which types of pairs are most similar, while --groupdist\n",
            "    focuses on the differences between the centers of these distributions and\n",
            "    estimates standard errors via delete-d jackknife.\n",
            "\n",
            "  --regress-distance [iters] [d]\n",
            "    Linear regression of pairwise genomic distances on pairwise average\n",
            "    phenotypes and vice versa, using delete-d jackknife for standard errors.  A\n",
            "    scalar phenotype is required.\n",
            "    * With less than two parameters, d is set to <number of people>^0.6 rounded\n",
            "      down.  With no parameters, 100k iterations are run.\n",
            "  --regress-rel [iters] [d]\n",
            "    Linear regression of pairwise genomic relationships on pairwise average\n",
            "    phenotypes, and vice versa.  Defaults for iters and d are the same as for\n",
            "    --regress-distance.\n",
            "\n",
            "  --genome ['gz'] ['rel-check'] ['full'] ['unbounded'] ['nudge']\n",
            "    Generate an identity-by-descent report.\n",
            "    * It is usually best to perform this calculation on a marker set in\n",
            "      approximate linkage equilibrium.\n",
            "    * The 'rel-check' modifier excludes pairs of samples with different FIDs\n",
            "      from the final report.\n",
            "    * 'full' adds raw pairwise comparison data to the report.\n",
            "    * The P(IBD=0/1/2) estimator employed by this command sometimes yields\n",
            "      numbers outside the range [0,1]; by default, these are clipped.  The\n",
            "      'unbounded' modifier turns off this clipping.\n",
            "    * Then, when PI_HAT^2 < P(IBD=2), 'nudge' adjusts the final P(IBD=0/1/2)\n",
            "      estimates to a theoretically possible configuration.\n",
            "    * The computation can be subdivided with --parallel.\n",
            "\n",
            "  --homozyg [{group | group-verbose}] ['consensus-match'] ['extend']\n",
            "            ['subtract-1-from-lengths']\n",
            "  --homozyg-snp <min var count>\n",
            "  --homozyg-kb <min length>\n",
            "  --homozyg-density <max inverse density (kb/var)>\n",
            "  --homozyg-gap <max internal gap kb length>\n",
            "  --homozyg-het <max hets>\n",
            "  --homozyg-window-snp <scanning window size>\n",
            "  --homozyg-window-het <max hets in scanning window hit>\n",
            "  --homozyg-window-missing <max missing calls in scanning window hit>\n",
            "  --homozyg-window-threshold <min scanning window hit rate>\n",
            "    These commands request a set of run-of-homozygosity reports, and allow you\n",
            "    to customize how they are generated.\n",
            "    * If you're satisfied with all the default settings described below, just\n",
            "      use --homozyg with no modifiers.  Otherwise, --homozyg lets you change a\n",
            "      few binary settings:\n",
            "      * 'group[-verbose]' adds a report on pools of overlapping runs of\n",
            "        homozygosity.  (Automatically set when --homozyg-match is present.)\n",
            "      * With 'group[-verbose]', 'consensus-match' causes pairwise segmental\n",
            "        matches to be called based on the variants in the pool's consensus\n",
            "        segment, rather than the variants in the pairwise intersection.\n",
            "      * Due to how the scanning window algorithm works, it is possible for a\n",
            "        reported ROH to be adjacent to a few homozygous variants.  The 'extend'\n",
            "        modifier causes them to be included in the reported ROH if that\n",
            "        wouldn't cause a violation of the --homozyg-density bound.\n",
            "      * By default, segment bp lengths are calculated as <end bp position> -\n",
            "        <start bp position> + 1.  Therefore, reports normally differ slightly\n",
            "        from PLINK 1.07, which does not add 1 at the end.  For testing\n",
            "        purposes, you can use the 'subtract-1-from-lengths' modifier to apply\n",
            "        the old formula.\n",
            "    * By default, only runs of homozygosity containing at least 100 variants,\n",
            "      and of total length >= 1000 kilobases, are noted.  You can change these\n",
            "      minimums with --homozyg-snp and --homozyg-kb, respectively.\n",
            "    * By default, a ROH must have at least one variant per 50 kb on average;\n",
            "      change this bound with --homozyg-density.\n",
            "    * By default, if two consecutive variants are more than 1000 kb apart, they\n",
            "      cannot be in the same ROH; change this bound with --homozyg-gap.\n",
            "    * By default, a ROH can contain an unlimited number of heterozygous calls;\n",
            "      you can impose a limit with --homozyg-het.\n",
            "    * By default, the scanning window contains 50 variants; change this with\n",
            "      --homozyg-window-snp.\n",
            "    * By default, a scanning window hit can contain at most 1 heterozygous\n",
            "      call and 5 missing calls; change these limits with --homozyg-window-het\n",
            "      and --homozyg-window-missing, respectively.\n",
            "    * By default, for a variant to be eligible for inclusion in a ROH, the hit\n",
            "      rate of all scanning windows containing the variant must be at least\n",
            "      0.05; change this threshold with --homozyg-window-threshold.\n",
            "\n",
            "  --cluster ['cc'] [{group-avg | old-tiebreaks}] ['missing'] ['only2']\n",
            "    Cluster samples using a pairwise similarity statistic (normally IBS).\n",
            "    * The 'cc' modifier forces every cluster to have at least one case and one\n",
            "      control.\n",
            "    * The 'group-avg' modifier causes clusters to be joined based on average\n",
            "      instead of minimum pairwise similarity.\n",
            "    * The 'missing' modifier causes clustering to be based on\n",
            "      identity-by-missingness instead of identity-by-state, and writes a\n",
            "      space-delimited identity-by-missingness matrix to disk.\n",
            "    * The 'only2' modifier causes only a .cluster2 file (which is valid input\n",
            "      for --within) to be written; otherwise 2 other files will be produced.\n",
            "    * By default, IBS ties are not broken in the same manner as PLINK 1.07, so\n",
            "      final cluster solutions tend to differ.  This is generally harmless.\n",
            "      However, to simplify testing, you can use the 'old-tiebreaks' modifier to\n",
            "      force emulation of the old algorithm.\n",
            "\n",
            "  --pca [count] ['header'] ['tabs'] ['var-wts']\n",
            "    Calculates a variance-standardized relationship matrix (use\n",
            "    --make-rel/--make-grm-gz/--make-grm-bin to dump it), and extracts the top\n",
            "    20 principal components.\n",
            "    * It is usually best to perform this calculation on a marker set in\n",
            "      approximate linkage equilibrium.\n",
            "    * You can change the number of PCs by passing a numeric parameter.\n",
            "    * The 'header' modifier adds a header line to the .eigenvec output file.\n",
            "      (For compatibility with the GCTA flag of the same name, the default is no\n",
            "      header line.)\n",
            "    * The 'tabs' modifier causes the .eigenvec file(s) to be tab-delimited.\n",
            "    * The 'var-wts' modifier requests an additional .eigenvec.var file with PCs\n",
            "      expressed as variant weights instead of sample weights.\n",
            "\n",
            "  --neighbour <n1> <n2>\n",
            "    (alias: --neighbor)\n",
            "    Report IBS distances from each sample to their n1th- to n2th-nearest\n",
            "    neighbors, associated Z-scores, and the identities of those neighbors.\n",
            "    Useful for outlier detection.\n",
            "\n",
            "  --assoc ['perm' | 'mperm='<value>] ['perm-count'] [{fisher | fisher-midp}]\n",
            "          ['counts'] ['set-test']\n",
            "  --assoc ['perm' | 'mperm='<value>] ['perm-count'] ['qt-means'] ['lin']\n",
            "          ['set-test']\n",
            "  --model ['perm' | 'mperm='<value>] ['perm-count']\n",
            "          [{fisher | fisher-midp | trend-only}] ['set-test']\n",
            "          [{dom | rec | gen | trend}]\n",
            "    Basic association analysis report.\n",
            "    Given a case/control phenotype, --assoc performs a 1df chi-square allelic\n",
            "    test, while --model performs 4 other tests as well (1df dominant gene\n",
            "    action, 1df recessive gene action, 2df genotypic, Cochran-Armitage trend).\n",
            "    * With 'fisher'/'fisher-midp', Fisher's exact test is used to generate\n",
            "      p-values.  'fisher-midp' also applies Lancaster's mid-p adjustment.\n",
            "    * 'perm' causes an adaptive permutation test to be performed.\n",
            "    * 'mperm='<value> causes a max(T) permutation test with the specified\n",
            "      number of replications to be performed.\n",
            "    * 'perm-count' causes the permutation test report to include counts instead\n",
            "      of frequencies.\n",
            "    * 'counts' causes --assoc to report allele counts instead of frequencies.\n",
            "    * 'set-test' tests the significance of variant sets.  Requires permutation;\n",
            "      can be customized with --set-p/--set-r2/--set-max.\n",
            "    * 'dom', 'rec', 'gen', and 'trend' force the corresponding test to be used\n",
            "      as the basis for --model permutation.  (By default, the most significant\n",
            "      result among the allelic, dominant, and recessive tests is used.)\n",
            "    * 'trend-only' causes only the trend test to be performed.\n",
            "    Given a quantitative phenotype, --assoc normally performs a Wald test.\n",
            "    * In this case, the 'qt-means' modifier causes trait means and standard\n",
            "      deviations stratified by genotype to be reported as well.\n",
            "    * 'lin' causes the Lin statistic to be computed, and makes it the basis for\n",
            "      multiple-testing corrections and permutation tests.\n",
            "    Several other flags (most notably, --aperm) can be used to customize the\n",
            "    permutation test.\n",
            "\n",
            "  --mh ['perm' | 'mperm='<value>] ['perm-count'] ['set-test']\n",
            "    (alias: --cmh)\n",
            "  --bd ['perm' | 'perm-bd' | 'mperm='<value>] ['perm-count'] ['set-test']\n",
            "  --mh2\n",
            "  --homog\n",
            "    Given a case/control phenotype and a set of clusters, --mh computes 2x2xK\n",
            "    Cochran-Mantel-Haenszel statistics for each variant, while --bd also\n",
            "    performs the Breslow-Day test for odds ratio homogeneity.  Permutation and\n",
            "    variant set testing based on the CMH (default) or Breslow-Day (when\n",
            "    'perm-bd' is present) statistic are supported.\n",
            "    The following similar analyses are also available:\n",
            "    * --mh2 swaps the roles of case/control status and cluster membership,\n",
            "      performing a phenotype-stratified IxJxK Cochran-Mantel-Haenszel test on\n",
            "      association between cluster assignments and genotypes.\n",
            "    * --homog executes an alternative to the Breslow-Day test, based on\n",
            "      partitioning of the chi-square statistic.\n",
            "\n",
            "  --gxe [covariate index]\n",
            "    Given both a quantitative phenotype and a case/control covariate loaded\n",
            "    with --covar defining two groups, --gxe compares the regression coefficient\n",
            "    derived from considering only members of one group to the regression\n",
            "    coefficient derived from considering only members of the other.  By\n",
            "    default, the first covariate in the --covar file defines the groups; use\n",
            "    e.g. \"--gxe 3\" to base them on the third covariate instead.\n",
            "\n",
            "  --linear ['perm' | 'mperm='<value>] ['perm-count'] ['set-test']\n",
            "           [{genotypic | hethom | dominant | recessive | no-snp}]\n",
            "           ['hide-covar'] [{sex | no-x-sex}] ['interaction'] ['beta']\n",
            "           ['standard-beta'] ['intercept']\n",
            "  --logistic ['perm' | 'mperm='<value>] ['perm-count'] ['set-test']\n",
            "             [{genotypic | hethom | dominant | recessive | no-snp}]\n",
            "             ['hide-covar'] [{sex | no-x-sex}] ['interaction'] ['beta']\n",
            "             ['intercept']\n",
            "    Multi-covariate association analysis on a quantitative (--linear) or\n",
            "    case/control (--logistic) phenotype.  Normally used with --covar.\n",
            "    * 'perm' normally causes an adaptive permutation test to be performed on\n",
            "      the main effect, while 'mperm='<value> starts a max(T) permutation test.\n",
            "    * 'perm-count' causes the permutation test report to include counts instead\n",
            "      of frequencies.\n",
            "    * 'set-test' tests the significance of variant sets.  Requires permutation;\n",
            "      can be customized with --set-p/--set-r2/--set-max.\n",
            "    * The 'genotypic' modifier adds an additive effect/dominance deviation 2df\n",
            "      joint test (0/1/2 and 0/1/0 coding), while 'hethom' uses 0/0/1 and 0/1/0\n",
            "      coding instead.  If permutation is also requested, these modifiers cause\n",
            "      permutation to be based on the joint test.\n",
            "    * 'dominant' and 'recessive' specify a model assuming full dominance or\n",
            "      recessiveness, respectively, for the A1 allele.\n",
            "    * 'no-snp' causes regression to be performed only on the phenotype and the\n",
            "      covariates, without reference to genomic data.  If permutation is also\n",
            "      requested, results are reported for all covariates.\n",
            "    * 'hide-covar' removes covariate-specific lines from the report.\n",
            "    * By default, sex (male = 1, female = 0) is automatically added as a\n",
            "      covariate on X chromosome variants, and nowhere else.  The 'sex' modifier\n",
            "      causes it to be added everywhere, while 'no-x-sex' excludes it.\n",
            "    * 'interaction' adds genotype x covariate interactions to the model.  This\n",
            "      cannot be used with the usual permutation tests; use --tests to define\n",
            "      the permutation test statistic instead.\n",
            "    * 'intercept' causes intercepts to be included in the main report.\n",
            "    * For logistic regressions, the 'beta' modifier causes regression\n",
            "      coefficients instead of odds ratios to be reported.\n",
            "    * With --linear, the 'standard-beta' modifier standardizes the phenotype\n",
            "      and all predictors to zero mean and unit variance before regression.\n",
            "\n",
            "  --dosage <allele dosage file> ['noheader'] ['skip0='<i>] ['skip1='<j>]\n",
            "           ['skip2='<k>] ['dose1'] ['format='<m>] ['Zout']\n",
            "           [{occur | standard-beta}] ['sex'] ['case-control-freqs']\n",
            "  --dosage <list file> list [{sepheader | noheader}] ['skip0='<i>]\n",
            "           ['skip1='<j>] ['skip2='<k>] ['dose1'] ['format='<m>] ['Zout']\n",
            "           [{occur | standard-beta}] ['sex'] ['case-control-freqs']\n",
            "  --write-dosage\n",
            "    Process (possibly gzipped) text files with variant-major allelic dosage\n",
            "    data.  This cannot be used with a regular input fileset; instead, you must\n",
            "    *only* specify a .fam and possibly a .map file, and you can't specify any\n",
            "    other commands.\n",
            "    * PLINK 2.0 will have first-class support for genotype probabilities.  An\n",
            "      equivalent data import flag will be provided then, and --dosage will be\n",
            "      retired.\n",
            "    * By default, --dosage assumes that only one allelic dosage file should be\n",
            "      loaded.  To specify multiple files,\n",
            "      1. create a master list with one entry per line.  There are normally two\n",
            "         supported formats for this list: just a filename per line, or variant\n",
            "         batch numbers in the first column and filenames in the second.\n",
            "      2. Provide the name of that list as the first --dosage parameter.\n",
            "      3. Add the 'list' modifier.\n",
            "    * By default, --dosage assumes the allelic dosage file(s) contain a header\n",
            "      line, which has 'SNP' in column i+1, 'A1' in column i+j+2, 'A2' in column\n",
            "      i+j+3, and sample FID/IIDs starting from column i+j+k+4.  (i/j/k are\n",
            "      normally zero, but can be changed with 'skip0', 'skip1', and 'skip2'\n",
            "      respectively.)  If such a header line is not present,\n",
            "      * when all samples appear in the same order as they do in the .fam file,\n",
            "        you can use the 'noheader' modifier.\n",
            "      * Otherwise, use the 'sepheader' modifier, and append sample ID filenames\n",
            "        to your 'list' file entries.\n",
            "    * The 'format=' modifier lets you specify the number of values used to\n",
            "      represent each dosage.  'format=1' normally indicates a single 0..2 A1\n",
            "      expected count; 'dose1' modifies this to a 0..1 frequency.  'format=2'\n",
            "      (the default) indicates a 0..1 homozygous A1 likelihood followed by a\n",
            "      0..1 het likelihood, while 'format=3' indicates 0..1 hom A1, 0..1 het,\n",
            "      0..1 hom A2.\n",
            "    * 'Zout' causes the output file to be gzipped.\n",
            "    * Normally, an association analysis is performed.  'standard-beta' and\n",
            "      'sex' behave as they are supposed to with --linear/--logistic.\n",
            "      'case-control-freqs' causes case and control allele frequencies to be\n",
            "      reported separately.\n",
            "    * There are three alternate modes which cause the association analysis to\n",
            "      be skipped.\n",
            "      * 'occur' requests a simple variant occurrence report.\n",
            "      * --write-dosage causes a simple merged file matching the 'format'\n",
            "        specification (not including 'dose1') to be generated.\n",
            "      * --score applies a linear scoring system to the dosages.\n",
            "\n",
            "  --lasso <h2 estimate> [min lambda] ['report-zeroes']\n",
            "    Estimate variant effect sizes via LASSO regression.  You must provide an\n",
            "    additive heritability estimate to calibrate the regression.\n",
            "    Note that this method may require a very large sample size (e.g. hundreds\n",
            "    of thousands) to be effective on complex polygenic traits.\n",
            "\n",
            "  --test-missing ['perm' | 'mperm='<value>] ['perm-count'] ['midp']\n",
            "    Check for association between missingness and case/control status, using\n",
            "    Fisher's exact test.  (Het. haploids are treated as missing.)\n",
            "    The 'midp' modifier causes Lancaster's mid-p adjustment to be applied.\n",
            "\n",
            "  --make-perm-pheno <ct>\n",
            "    Generate phenotype permutations and write them to disk, without invoking an\n",
            "    association test.\n",
            "\n",
            "  --tdt [{exact | exact-midp | poo}] ['perm' | 'mperm='<value>] ['perm-count']\n",
            "        [{parentdt1 | parentdt2 | pat | mat}] ['set-test']\n",
            "    Report transmission disequilibrium test statistics, given case/control\n",
            "    phenotypes and pedigree information.\n",
            "    * A Mendel error check is performed before the main tests; offending\n",
            "      genotypes are treated as missing by this analysis.\n",
            "    * By default, the basic TDT p-value is based on a chi-square test unless\n",
            "      you request the exact binomial test with 'exact' or 'exact-midp'.\n",
            "    * 'perm'/'mperm=' requests a family-based adaptive or max(T) permutation\n",
            "      test.  By default, the permutation test statistic is the basic TDT\n",
            "      p-value; 'parentdt1'/'parentdt2' cause parenTDT or combined test\n",
            "      p-values, respectively, to be considered instead.\n",
            "    * 'set-test' tests the significance of variant sets.  This cannot be used\n",
            "      with exact tests for now.\n",
            "    The 'poo' modifier causes a parent-of-origin analysis to be performed\n",
            "    instead, with transmissions from heterozygous fathers and heterozygous\n",
            "    mothers considered separately.\n",
            "    * The parent-of-origin analysis does not currently support exact tests.\n",
            "    * By default, the permutation test statistic is the absolute\n",
            "      parent-of-origin test Z score; 'pat'/'mat' cause paternal or maternal TDT\n",
            "      chi-square statistics, respectively, to be considered instead.\n",
            "\n",
            "  --qfam ['perm' | 'mperm='<value>] ['perm-count'] ['emp-se']\n",
            "  --qfam-parents ['perm' | 'mperm='<value>] ['perm-count'] ['emp-se']\n",
            "  --qfam-between ['perm' | 'mperm='<value>] ['perm-count'] ['emp-se']\n",
            "  --qfam-total ['perm' | 'mperm='<value>] ['perm-count'] ['emp-se']\n",
            "    QFAM family-based association test for quantitative traits.\n",
            "    * A Mendel error check is performed before the main tests; offending\n",
            "      genotypes are treated as missing by this analysis.\n",
            "    * This procedure requires permutation.  'perm' and 'perm-count' have the\n",
            "      usual meanings.  However, 'mperm='<value> just specifies a fixed number\n",
            "      of permutations; the method does not support a proper max(T) test.\n",
            "    * The 'emp-se' modifier adds BETA and EMP_SE (empirical standard error for\n",
            "      beta) fields to the .perm output file.\n",
            "\n",
            "  --annotate <PLINK report> ['attrib='<file>] ['ranges='<file>]\n",
            "             ['filter='<file>] ['snps='<file>] [{NA | prune}] ['block']\n",
            "             ['subset='<file>] ['minimal'] ['distance']\n",
            "    Add annotations to a variant-based PLINK report.  This requires an\n",
            "    annotation source:\n",
            "    * 'attrib='<file> specifies a (possibly gzipped) attribute file.\n",
            "    * 'ranges='<file> specifies a gene/range list file.\n",
            "    (Both source types can be specified simultaneously.)  The following options\n",
            "    are also supported:\n",
            "    * 'filter='<file> causes only variants within one of the ranges in the file\n",
            "      to be included in the new report.\n",
            "    * 'snps='<file> causes only variants named in the file to be included in\n",
            "      the new report.\n",
            "    * The 'NA' modifier causes unannotated variants to have 'NA' instead of '.'\n",
            "      in the new report's ANNOT column, while the 'prune' modifier excludes\n",
            "      them entirely.\n",
            "    * The 'block' modifier replaces the single ANNOT column with a 0/1-coded\n",
            "      column for each possible annotation.\n",
            "    * With 'ranges',\n",
            "      * 'subset='<file> causes only intervals named in the subset file to be\n",
            "        loaded from the ranges file.\n",
            "      * interval annotations normally come with a parenthesized signed distance\n",
            "        to the interval boundary (0 if the variant is located inside the\n",
            "        interval; this is always true without --border).  They can be excluded\n",
            "        with the 'minimal' modifier.\n",
            "      * the 'distance' modifier adds 'DIST' and 'SGN' columns describing signed\n",
            "        distance to the nearest interval.\n",
            "    * When --pfilter is present, high p-values are filtered out.\n",
            "\n",
            "  --clump <PLINK report filename(s)...>\n",
            "    Process association analysis report(s) with 'SNP' and p-value columns,\n",
            "    organizing results by LD-based clumps.  Multiple filenames can be separated\n",
            "    by spaces or commas.\n",
            "\n",
            "  --gene-report <PLINK report> <gene range file>\n",
            "    Generate a gene-based report from a variant-based report.\n",
            "    * When --pfilter is present, high p-values are filtered out.\n",
            "    * When --extract (without 'range') is present, only variants named in the\n",
            "      --extract file are considered.\n",
            "\n",
            "  --meta-analysis <PLINK report filenames...>\n",
            "  --meta-analysis <PLINK report filenames...> + [{logscale | qt}]\n",
            "                  [{no-map | no-allele}] ['study'] ['report-all']\n",
            "                  ['weighted-z']\n",
            "    Perform a meta-analysis on several variant-based reports with 'SNP' and\n",
            "    'SE' fields.\n",
            "    * Normally, an 'OR' odds ratio field must also be present in each input\n",
            "      file.  With 'logscale', 'BETA' log-odds values/regression coefficients\n",
            "      are expected instead, but the generated report will still contain odds\n",
            "      ratio estimates.  With 'qt', both input and output values are regression\n",
            "      betas.\n",
            "    * 'CHR', 'BP', and 'A1' fields are also normally required.  'no-map' causes\n",
            "      them to all be ignored, while 'no-allele' causes just 'A1' to be ignored.\n",
            "    * If 'A2' fields are present, and neither 'no-map' nor 'no-allele' was\n",
            "      specified, A1/A2 allele flips are handled properly.  Otherwise, A1\n",
            "      mismatches are thrown out.\n",
            "    * 'study' causes study-specific effect estimates to be collated in the\n",
            "      meta-analysis report.\n",
            "    * 'report-all' causes variants present in only a single input file to be\n",
            "      included in the meta-analysis report.\n",
            "    * 'weighted-z' requests weighted Z-score-based p-values (as computed by the\n",
            "      Abecasis Lab's METAL software) in addition to the usual inverse\n",
            "      variance-based analysis.  This requires P and effective sample size\n",
            "      fields.\n",
            "    * When --extract (without 'range') is present, only variants named in the\n",
            "      --extract file are considered.\n",
            "    * Unless 'no-map' is specified, chromosome filters are also respected.\n",
            "\n",
            "  --fast-epistasis [{boost | joint-effects | no-ueki}] ['case-only']\n",
            "                   [{set-by-set | set-by-all}] ['nop']\n",
            "  --epistasis [{set-by-set | set-by-all}]\n",
            "    Scan for epistatic interactions.  --fast-epistasis inspects 3x3 joint\n",
            "    genotype count tables and only applies to case/control phenotypes, while\n",
            "    --epistasis performs linear or logistic regression.\n",
            "    * By default, --fast-epistasis uses the PLINK 1.07 allele-based test.  Two\n",
            "      newer tests are now supported: 'boost' invokes the likelihood ratio test\n",
            "      introduced by Wan X et al. (2010) BOOST: A Fast Approach to Detecting\n",
            "      Gene-Gene Interactions in Genome-wide Case-Control Studies, while\n",
            "      'joint-effects' applies the joint effects test introduced in Ueki M,\n",
            "      Cordell HJ (2012) Improved statistics for genome-wide interaction\n",
            "      analysis.\n",
            "    * The original --fast-epistasis test normally applies the variance and\n",
            "      empty cell corrections suggested by Ueki and Cordell's paper.  To disable\n",
            "      them, use the 'no-ueki' modifier.\n",
            "    * 'case-only' requests a case-only instead of a case/control test.\n",
            "    * By default, all pairs of variants across the entire genome are tested.\n",
            "      To just test pairs of variants within a single set, add the 'set-by-set'\n",
            "      modifier and load exactly one set with --set/--make-set; with exactly two\n",
            "      sets loaded, all variants in one set are tested against all variants in\n",
            "      the other.  'set-by-all' tests all variants in one set against the entire\n",
            "      genome instead.\n",
            "    * 'nop' strips p-values from the main report.\n",
            "    * These computations can be subdivided with --parallel; however...\n",
            "  --epistasis-summary-merge <common file prefix> <ct>\n",
            "    When a --[fast-]epistasis job is subdivided with --parallel, the main\n",
            "    report can be assembled at the end by applying Unix 'cat' in the usual\n",
            "    manner, but the .summary.1, .summary.2, ... files may require a specialized\n",
            "    merge.  --epistasis-summary-merge takes care of the latter.\n",
            "\n",
            "  --twolocus <variant ID> <variant ID>\n",
            "    Two-locus joint genotype count report.\n",
            "\n",
            "  --score <filename> [i] [j] [k] ['header'] [{sum | no-sum}]\n",
            "          [{no-mean-imputation | center}] ['include-cnt'] ['double-dosage']\n",
            "    Apply a linear scoring system to each sample.\n",
            "    The input file should have one line per scored variant.  Variant IDs are\n",
            "    read from column #i, allele codes are read from column #j, and scores are\n",
            "    read from column #k, where i defaults to 1, j defaults to i+1, and k\n",
            "    defaults to j+1.\n",
            "    * The 'header' modifier causes the first nonempty line of the input file to\n",
            "      be ignored; otherwise, --score assumes there is no header line.\n",
            "    * By default, final scores are averages of the valid per-variant scores.\n",
            "      The 'sum' modifier causes sums to be reported instead.  (This cannot be\n",
            "      used with 'no-mean-imputation'.  And for backward compatibility, 'sum' is\n",
            "      automatically on with dosage data unless 'no-sum' is specified.)\n",
            "    * By default, copies of the unnamed allele contribute zero to score, while\n",
            "      missing genotypes contribute an amount proportional to the loaded (via\n",
            "      --read-freq) or imputed allele frequency.  To throw out missing\n",
            "      observations instead (decreasing the denominator in the final average\n",
            "      when this happens), use the 'no-mean-imputation' modifier.\n",
            "    * Alternatively, you can use the 'center' modifier to shift all scores to\n",
            "      mean zero.\n",
            "    * This command can be used with dosage data.  By default, the 'CNT' column\n",
            "      is omitted from the output file in this case; use 'include-cnt' to keep\n",
            "      it.  Also, note that scores are multiplied by 0..1 dosages, not 0..2\n",
            "      diploid allele counts, unless the 'double-dosage' modifier is present.\n",
            "\n",
            "  --R <R script file> ['debug']\n",
            "    Connect to a Rserve (preferably version 1.7 or later) background process,\n",
            "    and execute the Rplink function defined in the input file.  (Unless the\n",
            "    'debug' modifier is present; in that case, the R commands that PLINK would\n",
            "    have tried to execute are logged to a file.)\n",
            "\n",
            "  --write-var-ranges <block ct>\n",
            "    Divide the set of variants into equal-size blocks.  (Can be used with\n",
            "    --snps to split a job across multiple machines.)\n",
            "\n",
            "The following other flags are supported.  (Order of operations is described at\n",
            "https://www.cog-genomics.org/plink/1.9/order .)\n",
            "  --script <fname> : Include command-line options from file.\n",
            "  --rerun [log]    : Rerun commands in log (default 'plink.log').\n",
            "  --version        : Display only version number before exiting.\n",
            "  --silent         : Suppress output to console.\n",
            "  --gplink         : Reserved for interoperation with gPLINK.\n",
            "  --missing-genotype <char> : Set missing genotype code (normally '0').\n",
            "  --missing-genotype2 <c>   : Set additional missing genotype code in .bim\n",
            "                              input (normally '.').\n",
            "  --double-id          : Set both FIDs and IIDs to the VCF/BCF sample ID.\n",
            "  --const-fid [ID]     : Set all FIDs to the given constant (default '0').\n",
            "  --id-delim [d]       : Parse sample IDs as <FID><d><IID> (default delim '_').\n",
            "  --vcf-idspace-to <c> : Convert spaces in sample IDs to the given character.\n",
            "  --biallelic-only ['strict'] ['list'] : Skip VCF variants with 2+ ALT alleles.\n",
            "  --vcf-min-qual <val>           : Skip VCF variants with low/missing QUAL.\n",
            "  --vcf-filter [exception(s)...] : Skip variants which have FILTER failures.\n",
            "  --vcf-require-gt               : Skip variants with no GT field.\n",
            "  --vcf-min-gq <val>             : No-call a genotype when GQ is below the\n",
            "                                   given threshold.\n",
            "  --vcf-min-gp <val>             : No-call a genotype when 0-1 scaled GP is\n",
            "                                   below the given threshold.\n",
            "  --vcf-half-call <m>  : Specify how '0/.' and similar VCF GT values should be\n",
            "                         handled.  The following four modes are supported:\n",
            "                         * 'error'/'e' (default) errors out and reports line #.\n",
            "                         * 'haploid'/'h' treats them as haploid calls.\n",
            "                         * 'missing'/'m' treats them as missing.\n",
            "                         * 'reference'/'r' treats the missing value as 0.\n",
            "  --oxford-single-chr <chr nm> : Specify single-chromosome .gen file with\n",
            "                                 ignorable first column.\n",
            "  --oxford-pheno-name <col nm> : Import named phenotype from the .sample file.\n",
            "  --hard-call-threshold <val>  : When an Oxford-format fileset is loaded, calls\n",
            "  --hard-call-threshold random   with uncertainty level greater than 0.1 are\n",
            "                                 normally treated as missing.  You can adjust\n",
            "                                 this threshold by providing a numeric\n",
            "                                 parameter, or randomize all calls with\n",
            "                                 'random'.\n",
            "  --missing-code [string list] : Comma-delimited list of missing phenotype\n",
            "    (alias: --missing_code)      values for Oxford-format filesets (def. 'NA').\n",
            "  --simulate-ncases <num>   : Set --simulate case count (default 1000).\n",
            "  --simulate-ncontrols <n>  : Set --simulate control count (default 1000).\n",
            "  --simulate-prevalence <p> : Set --simulate disease prevalence (default 0.01).\n",
            "  --simulate-n <num>        : Set --simulate-qt sample count (default 1000).\n",
            "  --simulate-label <prefix> : Set --simulate[-qt] FID/IID name prefix.\n",
            "  --simulate-missing <freq> : Set --simulate[-qt] missing genotype frequency.\n",
            "  --allow-extra-chr ['0']   : Permit unrecognized chromosome codes.  The '0'\n",
            "    (alias: --aec)            modifier causes them to be treated as if they had\n",
            "                              been set to zero.\n",
            "  --chr-set <autosome ct> ['no-x'] ['no-y'] ['no-xy'] ['no-mt'] :\n",
            "    Specify a nonhuman chromosome set.  The first parameter sets the number of\n",
            "    diploid autosome pairs if positive, or haploid chromosomes if negative.\n",
            "    Given diploid autosomes, the remaining modifiers indicate the absence of\n",
            "    the named non-autosomal chromosomes.\n",
            "  --cow/--dog/--horse/--mouse/--rice/--sheep : Shortcuts for those species.\n",
            "  --autosome-num <value>    : Alias for \"--chr-set <value> no-y no-xy no-mt\".\n",
            "  --cm-map <fname pattern> [chr] : Use SHAPEIT-format recombination maps to set\n",
            "                                   centimorgan positions.  To process more than\n",
            "                                   one chromosome, include a '@' in the first\n",
            "                                   parameter where the chrom. number belongs,\n",
            "                                   e.g. 'genetic_map_chr@_combined_b37.txt'.\n",
            "  --zero-cms         : Zero out centimorgan positions.\n",
            "  --allow-no-samples : Allow the input fileset to contain no samples.\n",
            "  --allow-no-vars    : Allow the input fileset to contain no variants.\n",
            "  --pheno <fname>  : Load phenotype data from the specified file, instead of\n",
            "                     using the values in the main input fileset.\n",
            "  --all-pheno      : For basic association tests, loop through all phenotypes\n",
            "                     in --pheno file.\n",
            "  --mpheno <n>     : Load phenotype from column (n+2) in --pheno file.\n",
            "  --pheno-name <c> : If --pheno file has a header row, use column with the\n",
            "                     given name.\n",
            "  --pheno-merge    : When the main input fileset contains an phenotype value\n",
            "                     for a sample, but the --pheno file does not, use the\n",
            "                     original value instead of treating the phenotype as\n",
            "                     missing.\n",
            "  --missing-phenotype <v> : Set missing phenotype value (normally -9).\n",
            "  --1                     : Expect case/control phenotypes to be coded as\n",
            "                            0 = control, 1 = case, instead of the usual\n",
            "                            0 = missing, 1 = control, 2 = case.  This also\n",
            "                            forces phenotypes to be interpreted as case/ctrl.\n",
            "  --make-pheno <fn> <val> : Define a new case/control phenotype.  If the val\n",
            "                            parameter is '*', all samples listed in the given\n",
            "                            file are cases, and everyone else is a control.\n",
            "                            (Note that, in some shells, it is necessary to\n",
            "                            surround the * with quotes.)\n",
            "                            Otherwise, all samples with third column entry\n",
            "                            equal to the val parameter are cases, and all other\n",
            "                            samples mentioned in the file are controls.\n",
            "  --tail-pheno <Lt> [Hbt] : Downcode a scalar phenotype to a case/control\n",
            "                            phenotype.  All samples with phenotype values\n",
            "                            greater than Hbt are cases, and all with values\n",
            "                            less than or equal to Lt are controls.  If Hbt is\n",
            "                            unspecified, it is equal to Lt; otherwise,\n",
            "                            in-between phenotype values are set to missing.\n",
            "  --covar <filename> ['keep-pheno-on-missing-cov'] : Specify covariate file.\n",
            "  --covar-name <...>       : Specify covariate(s) in --covar file by name.\n",
            "                             Separate multiple names with spaces or commas, and\n",
            "                             use dashes to designate ranges.\n",
            "  --covar-number <...>     : Specify covariate(s) in --covar file by index.\n",
            "  --no-const-covar         : Exclude constant covariates.\n",
            "  --allow-no-covars        : Allow no covariates to be loaded from --covar\n",
            "                             file.\n",
            "  --within <f> ['keep-NA'] : Specify initial cluster assignments.\n",
            "  --mwithin <n>            : Load cluster assignments from column n+2.\n",
            "  --family                 : Create a cluster for each family ID.\n",
            "  --loop-assoc <f> ['keep-NA']  : Run specified case/control association\n",
            "                                  commands once for each cluster in the file,\n",
            "                                  using cluster membership as the phenotype.\n",
            "  --set <filename>              : Load sets from a .set file.\n",
            "  --set-names <name(s)...>      : Load only sets named on the command line.\n",
            "                                  Use spaces to separate multiple names.\n",
            "  --subset <filename>           : Load only sets named in the given text file.\n",
            "  --set-collapse-all <set name> : Merge all sets.\n",
            "  --complement-sets             : Invert all sets.  (Names gain 'C_' prefixes.)\n",
            "  --make-set-complement-all <s> : --set-collapse-all + inversion.\n",
            "  --make-set <filename>         : Define sets from a list of named bp ranges.\n",
            "  --make-set-border <kbs>       : Stretch regions in --make-set file.\n",
            "  --make-set-collapse-group     : Define sets from groups instead of sets in\n",
            "                                  --make-set file.\n",
            "  --keep <filename>       : Exclude all samples not named in the file.\n",
            "  --remove <filename>     : Exclude all samples named in the file.\n",
            "  --keep-fam <filename>   : Exclude all families not named in the file.\n",
            "  --remove-fam <filename> : Exclude all families named in the file.\n",
            "  --extract ['range'] <f> : Exclude all variants not named in the file.\n",
            "  --exclude ['range'] <f> : Exclude all variants named in the file.\n",
            "  --keep-clusters <filename>          : These can be used individually or in\n",
            "  --keep-cluster-names <name(s)...>     combination to define a list of\n",
            "                                        clusters to keep; all samples not in a\n",
            "                                        cluster in that list are then excluded.\n",
            "                                        Use spaces to separate cluster names\n",
            "                                        for --keep-cluster-names.\n",
            "  --remove-clusters <filename>        : Exclude all clusters named in the file.\n",
            "  --remove-cluster-names <name(s)...> : Exclude the named clusters.\n",
            "  --gene <sets...> : Exclude variants not in a set named on the command line.\n",
            "                     (Separate multiple set names with spaces.)\n",
            "  --gene-all       : Exclude variants which aren't a member of any set.  (PLINK\n",
            "                     1.07 automatically did this under some circumstances.)\n",
            "  --attrib <f> [att lst] : Given a file assigning attributes to variants, and a\n",
            "  --attrib-indiv <f> [a]   comma-delimited list (with no whitespace) of\n",
            "                           attribute names, remove variants/samples which are\n",
            "                           either missing from the file or don't have any of\n",
            "                           the listed attributes.  If some attribute names in\n",
            "                           the list are preceded by '-', they are treated as\n",
            "                           \"negative match conditions\" instead: variants with\n",
            "                           at least one negative match attribute are removed.\n",
            "                           The first character in the list cannot be a '-', due\n",
            "                           to how command-line parsing works; add a comma in\n",
            "                           front to get around this.\n",
            "  --chr <chrs...>  : Exclude all variants not on the given chromosome(s).\n",
            "                     Valid choices for humans are 0 (unplaced), 1-22, X, Y, XY,\n",
            "                     and MT.  Separate multiple chromosomes with spaces and/or\n",
            "                     commas, and use a dash (no adjacent spaces permitted) to\n",
            "                     denote a range, e.g. \"--chr 1-4, 22, xy\".\n",
            "  --not-chr <...>  : Reverse of --chr (exclude variants on listed chromosomes).\n",
            "  --autosome       : Exclude all non-autosomal variants.\n",
            "  --autosome-xy    : Exclude all non-autosomal variants, except those with\n",
            "                     chromosome code XY (pseudo-autosomal region of X).\n",
            "  --snps-only ['just-acgt'] : Exclude non-SNP variants.  By default, SNP = both\n",
            "                              allele codes are single-character; 'just-acgt'\n",
            "                              restricts codes to {A,C,G,T,a,c,g,t,<missing>}.\n",
            "  --from <var ID>  : Use ID(s) to specify a variant range to load.  When used\n",
            "  --to   <var ID>    together, both variants must be on the same chromosome.\n",
            "  --snp  <var ID>  : Specify a single variant to load.\n",
            "  --exclude-snp <> : Specify a single variant to exclude.\n",
            "  --window  <kbs>  : With --snp or --exclude-snp, loads/excludes all variants\n",
            "                     within half the specified kb distance of the named one.\n",
            "  --from-bp <pos>  : Use physical position(s) to define a variant range to\n",
            "  --to-bp   <pos>    load.  --from-kb/--to-kb/--from-mb/--to-mb allow decimal\n",
            "  --from-kb <pos>    values.  You must also specify a single chromosome (using\n",
            "  --to-kb   <pos>    e.g. --chr) when using these flags.\n",
            "  --from-mb <pos>\n",
            "  --to-mb   <pos>\n",
            "  --snps <var IDs...>  : Use IDs to specify variant range(s) to load or\n",
            "  --exclude-snps <...>   exclude.  E.g. \"--snps rs1111-rs2222, rs3333, rs4444\".\n",
            "  --thin <p>       : Randomly remove variants, retaining each with prob. p.\n",
            "  --thin-count <n> : Randomly remove variants until n of them remain.\n",
            "  --bp-space <bps> : Remove variants so that each pair is no closer than the\n",
            "                     given bp distance.  (Equivalent to VCFtools --thin.)\n",
            "  --thin-indiv <p>         : Randomly remove samples, retaining with prob. p.\n",
            "  --thin-indiv-count <n>   : Randomly remove samples until n of them remain.\n",
            "  --filter <f> <val(s)...> : Exclude all samples without a 3rd column entry in\n",
            "                             the given file matching one of the given\n",
            "                             space-separated value(s).\n",
            "  --mfilter <n>            : Match against (n+2)th column instead.\n",
            "  --geno [val]     : Exclude variants with missing call frequencies greater\n",
            "                     than a threshold (default 0.1).  (Note that the default\n",
            "                     threshold is only applied if --geno is invoked without a\n",
            "                     parameter; when --geno is not invoked, no per-variant\n",
            "                     missing call frequency ceiling is enforced at all.  Other\n",
            "                     inclusion/exclusion default thresholds work the same way.)\n",
            "  --mind [val]     : Exclude samples with missing call frequencies greater than\n",
            "                     a threshold (default 0.1).\n",
            "  --oblig-missing <f1> <f2> : Specify blocks of missing genotype calls for\n",
            "                              --geno/--mind to ignore.  The first file should\n",
            "                              have variant IDs in the first column and block\n",
            "                              IDs in the second, while the second file should\n",
            "                              have FIDs in the first column, IIDs in the\n",
            "                              second, and block IDs in the third.\n",
            "  --prune             : Remove samples with missing phenotypes.\n",
            "  --maf [freq]        : Exclude variants with minor allele frequency lower than\n",
            "                        a threshold (default 0.01).\n",
            "  --max-maf <freq>    : Exclude variants with MAF greater than the threshold.\n",
            "  --mac <ct>          : Exclude variants with minor allele count lower than the\n",
            "    (alias: --min-ac)   given threshold.\n",
            "  --max-mac <ct>      : Exclude variants with minor allele count greater than\n",
            "    (alias: --max-ac)   the given threshold.\n",
            "  --maf-succ       : Rule of succession MAF estimation (used in EIGENSOFT).\n",
            "                     Given j observations of one allele and k >= j observations\n",
            "                     of the other, infer a MAF of (j+1) / (j+k+2), rather than\n",
            "                     the default j / (j+k).\n",
            "  --read-freq <fn> : Estimate MAFs and heterozygote frequencies from the given\n",
            "                     --freq[x] report, instead of the input fileset.\n",
            "  --hwe <p> ['midp'] ['include-nonctrl'] : Exclude variants with Hardy-Weinberg\n",
            "                                           equilibrium exact test p-values\n",
            "                                           below a threshold.\n",
            "  --me <t> <v> ['var-first'] : Filter out trios and variants with Mendel error\n",
            "                               rates exceeding the given thresholds.\n",
            "  --me-exclude-one [ratio]   : Make --me exclude only one sample per trio.\n",
            "  --qual-scores <f> [qcol] [IDcol] [skip] : Filter out variants with\n",
            "                                            out-of-range quality scores.\n",
            "                                            Default range is now [0, \\infty ).\n",
            "  --qual-threshold <min qual score>       : Set --qual-scores range floor.\n",
            "  --qual-max-threshold <max qual score>   : Set --qual-scores range ceiling.\n",
            "  --allow-no-sex   : Do not treat ambiguous-sex samples as having missing\n",
            "                     phenotypes in analysis commands.  (Automatic /w --no-sex.)\n",
            "  --must-have-sex  : Force ambiguous-sex phenotypes to missing on\n",
            "                     --make-bed/--make-just-fam/--recode/--write-covar.\n",
            "  --filter-cases       : Include only cases in the current analysis.\n",
            "  --filter-controls    : Include only controls.\n",
            "  --filter-males       : Include only males.\n",
            "  --filter-females     : Include only females.\n",
            "  --filter-founders    : Include only founders.\n",
            "  --filter-nonfounders : Include only nonfounders.\n",
            "  --nonfounders        : Include nonfounders in allele freq/HWE calculations.\n",
            "  --make-founders ['require-2-missing'] ['first'] :\n",
            "    Clear parental IDs for those with 1+ missing parent(s).\n",
            "  --recode-allele <fn> : With --recode A/A-transpose/AD, count alleles named in\n",
            "                         the file (otherwise A1 alleles are always counted).\n",
            "  --output-chr <MT code> : Set chromosome coding scheme in output files by\n",
            "                           providing the desired human mitochondrial code.\n",
            "                           (Options are '26', 'M', 'MT', '0M', 'chr26', 'chrM',\n",
            "                           and 'chrMT'.)\n",
            "  --output-missing-genotype <ch> : Set the code used to represent missing\n",
            "                                   genotypes in output files (normally the\n",
            "                                   --missing-genotype value).\n",
            "  --output-missing-phenotype <s> : Set the string used to represent missing\n",
            "                                   phenotypes in output files (normally the\n",
            "                                   --missing-phenotype value).\n",
            "  --zero-cluster <f> : In combination with --within/--family, set blocks of\n",
            "                       genotype calls to missing.  The input file should have\n",
            "                       variant IDs in the first column and cluster IDs in the\n",
            "                       second.  This must now be used with --make-bed and no\n",
            "                       other output commands.\n",
            "  --set-hh-missing       : Cause --make-bed and --recode to set heterozygous\n",
            "                           haploid genotypes to missing.\n",
            "  --set-mixed-mt-missing : Cause --make-bed and --recode to set mixed MT\n",
            "                           genotypes to missing.\n",
            "  --split-x <bp1> <bp2> ['no-fail']\n",
            "  --split-x <build> ['no-fail'] :\n",
            "    Changes chromosome code of all chrX variants with bp position <= bp1 or >=\n",
            "    bp2 to XY.  The following build codes are supported as shorthand:\n",
            "    * 'b36'/'hg18' = NCBI 36, 2709521/154584237\n",
            "    * 'b37'/'hg19' = GRCh37, 2699520/154931044\n",
            "    * 'b38'/'hg38' = GRCh38, 2781479/155701383\n",
            "    By default, PLINK errors out when no variants would be affected by\n",
            "    --split-x; the 'no-fail' modifier (useful in scripts) overrides this.\n",
            "  --merge-x ['no-fail'] : Merge XY chromosome back with X.\n",
            "  --set-me-missing      : Cause --make-bed to set Mendel errors to missing.\n",
            "  --fill-missing-a2     : Cause --make-bed to replace all missing calls with\n",
            "                          homozygous A2 calls.\n",
            "  --set-missing-var-ids <t>   : Given a template string with a '@' where the\n",
            "                                chromosome code should go and '#' where the bp\n",
            "                                coordinate belongs, --set-missing-var-ids\n",
            "                                assigns chromosome-and-bp-based IDs to unnamed\n",
            "                                variants.\n",
            "                                You may also use '$1' and '$2' to refer to\n",
            "                                allele names in the template string, and in\n",
            "                                fact this becomes essential when multiple\n",
            "                                variants share the same coordinate.\n",
            "  --new-id-max-allele-len <n> : Specify maximum number of leading characters\n",
            "                                from allele names to include in new variant IDs\n",
            "                                (default 23).\n",
            "  --missing-var-code <string> : Change unnamed variant code (default '.').\n",
            "  --update-chr  <f> [chrcol] [IDcol]  [skip] : Update variant chromosome codes.\n",
            "  --update-cm   <f> [cmcol]  [IDcol]  [skip] : Update centimorgan positions.\n",
            "  --update-map  <f> [bpcol]  [IDcol]  [skip] : Update variant bp positions.\n",
            "  --update-name <f> [newcol] [oldcol] [skip] : Update variant IDs.\n",
            "  --update-alleles <fname> : Update variant allele codes.\n",
            "  --allele1234 ['multichar'] : Interpret/recode A/C/G/T alleles as 1/2/3/4.\n",
            "                               With 'multichar', converts all A/C/G/Ts in\n",
            "                               allele names to 1/2/3/4s.\n",
            "  --alleleACGT ['multichar'] : Reverse of --allele1234.\n",
            "  --update-ids <f>     : Update sample IDs.\n",
            "  --update-parents <f> : Update parental IDs.\n",
            "  --update-sex <f> [n] : Update sexes.  Sex (1 or M = male, 2 or F = female, 0\n",
            "                         = missing) is loaded from column n+2 (default n is 1).\n",
            "  --flip <filename>    : Flip alleles (A<->T, C<->G) for SNP IDs in the file.\n",
            "  --flip-subset <fn>   : Only apply --flip to samples in --flip-subset file.\n",
            "  --flip-scan-window <ct+1> : Set --flip-scan max variant ct dist. (def. 10).\n",
            "  --flip-scan-window-kb <x> : Set --flip-scan max kb distance (default 1000).\n",
            "  --flip-scan-threshold <x> : Set --flip-scan min correlation (default 0.5).\n",
            "  --keep-allele-order  : Keep the allele order defined in the .bim file,\n",
            "  --real-ref-alleles     instead of forcing A2 to be the major allele.\n",
            "                         --real-ref-alleles also removes 'PR' from the INFO\n",
            "                         values emitted by --recode vcf{,-fid,-iid}.\n",
            "  --a1-allele <f> [a1col] [IDcol] [skip] : Force alleles in the file to A1.\n",
            "  --a2-allele <filename> [a2col] [IDcol] [skip] :\n",
            "    Force alleles in the file to A2.  (\"--a2-allele <VCF filename> 4 3 '#'\",\n",
            "    which scrapes reference allele assignments from a VCF file, is especially\n",
            "    useful.)\n",
            "  --indiv-sort <m> [f] : Specify FID/IID sort order.  The following four modes\n",
            "                         are supported:\n",
            "                         * 'none'/'0' keeps samples in the order they were\n",
            "                           loaded.  Default for non-merge operations.\n",
            "                         * 'natural'/'n' invokes 'natural sort', e.g.\n",
            "                           'id2' < 'ID3' < 'id10'.  Default when merging.\n",
            "                         * 'ascii'/'a' sorts in ASCII order, e.g.\n",
            "                           'ID3' < 'id10' < 'id2'.\n",
            "                         * 'file'/'f' uses the order in the given file (named\n",
            "                           in the second parameter).\n",
            "                         For now, only --merge/--bmerge/--merge-list and\n",
            "                         --make-bed/--make-just-fam respect this flag.\n",
            "  --with-phenotype ['no-parents'] [{no-sex | female-2}] :\n",
            "    Include more sample info in new .cov file.\n",
            "  --dummy-coding [N] ['no-round'] : Split categorical variables (n categories,\n",
            "                                    2 < n <= N, default N is 49) into n-1\n",
            "                                    binary dummy variables when writing\n",
            "                                    covariate file.\n",
            "  --merge-mode <n>   : Adjust --[b]merge/--merge-list behavior based on a\n",
            "                       numeric code.\n",
            "                       1 (default) = ignore missing calls, otherwise difference\n",
            "                                     -> missing\n",
            "                       2 = only overwrite originally missing calls\n",
            "                       3 = only overwrite when nonmissing in new file\n",
            "                       4/5 = never overwrite and always overwrite, respectively\n",
            "                       6 = report all mismatching calls without merging\n",
            "                       7 = report mismatching nonmissing calls without merging\n",
            "  --merge-equal-pos  : With --merge/--bmerge/--merge-list, merge variants with\n",
            "                       different names but identical positions.  (Exception:\n",
            "                       same-position chromosome code 0 variants aren't merged.)\n",
            "  --mendel-duos      : Make Mendel error checks consider samples with only one\n",
            "                       parent in the dataset.\n",
            "  --mendel-multigen  : Make Mendel error checks consider (great-)grandparental\n",
            "                       genotypes when parental genotype data is missing.\n",
            "  --ld-window <ct+1> : Set --r/--r2 max variant ct pairwise distance (usu. 10).\n",
            "  --ld-window-kb <x> : Set --r/--r2 max kb pairwise distance (usually 1000).\n",
            "  --ld-window-cm <x> : Set --r/--r2 max centimorgan pairwise distance.\n",
            "  --ld-window-r2 <x> : Set threshold for --r2 report inclusion (usually 0.2).\n",
            "  --ld-snp <var ID>  : Set first variant in all --r/--r2 pairs.\n",
            "  --ld-snps <vID...> : Restrict first --r/--r2 variant to the given ranges.\n",
            "  --ld-snp-list <f>  : Restrict first --r/--r2 var. to those named in the file.\n",
            "  --list-all         : Generate the 'all' mode report when using --show-tags in\n",
            "                       file mode.\n",
            "  --tag-kb <kbs>     : Set --show-tags max tag kb distance (default 250).\n",
            "  --tag-r2 <val>     : Set --show-tags min tag r-squared (default 0.8)\n",
            "  --tag-mode2        : Use two-column --show-tags (file mode) I/O format.\n",
            "  --ld-xchr <code>   : Set chrX model for --indep[-pairwise], --r/--r2,\n",
            "                       --flip-scan, and --show-tags.\n",
            "                       1 (default) = males coded 0/1, females 0/1/2 (A1 dosage)\n",
            "                       2 = males coded 0/2\n",
            "                       3 = males coded 0/2, but females given double weighting\n",
            "  --blocks-max-kb <kbs>      : Set --blocks maximum haploblock span (def. 200).\n",
            "  --blocks-min-maf <cutoff>  : Adjust --blocks MAF minimum (default 0.05).\n",
            "  --blocks-strong-lowci <x>  : Set --blocks \"strong LD\" CI thresholds (defaults\n",
            "  --blocks-strong-highci <x>   0.70 and 0.98).\n",
            "  --blocks-recomb-highci <x> : Set 'recombination' CI threshold (default 0.90).\n",
            "  --blocks-inform-frac <x>   : Force haploblock <strong LD pairs>:<total\n",
            "                               informative pairs> ratios to be larger than this\n",
            "                               value (default 0.95).\n",
            "  --distance-wts exp=<x>        : When computing genomic distances, assign each\n",
            "                                  variant a weight of (2q(1-q))^{-x}, where q\n",
            "                                  is the loaded or inferred MAF.\n",
            "  --read-dists <dist file> [id file] : Load a triangular binary distance matrix\n",
            "                                       instead of recalculating from scratch.\n",
            "  --ppc-gap <val>    : Minimum number of base pairs, in thousands, between\n",
            "                       informative pairs of markers used in --genome PPC test.\n",
            "                       500 if unspecified.\n",
            "  --min <cutoff>     : Specify minimum PI_HAT for inclusion in --genome report.\n",
            "  --max <cutoff>     : Specify maximum PI_HAT for inclusion in --genome report.\n",
            "  --homozyg-match <> : Set minimum concordance across jointly homozygous\n",
            "                       variants for a pairwise allelic match to be declared.\n",
            "  --pool-size <ct>   : Set minimum size of pools in \"--homozyg group\" report.\n",
            "  --read-genome <fn> : Load --genome report for --cluster/--neighbour, instead\n",
            "                       of recalculating IBS and PPC test p-values from scratch.\n",
            "  --ppc <p-val>    : Specify minimum PPC test p-value within a cluster.\n",
            "  --mc <max size>  : Specify maximum cluster size.\n",
            "  --mcc <c1> <c2>  : Specify maximum case and control counts per cluster.\n",
            "  --K <min count>  : Specify minimum cluster count.\n",
            "  --ibm <val>      : Specify minimum identity-by-missingness.\n",
            "  --match <f> [mv] : Use covariate values to restrict clustering.  Without\n",
            "                     --match-type, two samples can only be in the same cluster\n",
            "                     if all covariates match.  The optional second parameter\n",
            "                     specifies a covariate value to treat as missing.\n",
            "  --match-type <f> : Refine interpretation of --match file.  The --match-type\n",
            "                     file is expected to be a single line with as many entries\n",
            "                     as the --match file has covariates; '0' entries specify\n",
            "                     \"negative matches\" (i.e. samples with equal covariate\n",
            "                     values cannot be in the same cluster), '1' entries specify\n",
            "                     \"positive matches\" (default), and '-1' causes the\n",
            "                     corresponding covariate to be ignored.\n",
            "  --qmatch <f> [m] : Force all members of a cluster to have similar\n",
            "  --qt <fname>       quantitative covariate values.  The --qmatch file contains\n",
            "                     the covariate values, while the --qt file is a list of\n",
            "                     nonnegative tolerances (and '-1's marking covariates to\n",
            "                     skip).\n",
            "  --pca-cluster-names <...> : These can be used individually or in combination\n",
            "  --pca-clusters <fname>      to define a list of clusters to use in the basic\n",
            "                              --pca computation.  (--pca-cluster-names expects\n",
            "                              a space-delimited sequence of cluster names,\n",
            "                              while --pca-clusters expects a file with one\n",
            "                              cluster name per line.)  All samples outside\n",
            "                              those clusters will then be projected on to the\n",
            "                              calculated PCs.\n",
            "  --mds-plot <dims> ['by-cluster'] ['eigendecomp'] ['eigvals'] :\n",
            "    Multidimensional scaling analysis.  Requires --cluster.\n",
            "  --cell <thresh>  : Skip some --model tests when a contingency table entry is\n",
            "                     smaller than the given threshold.\n",
            "  --condition <var ID> [{dominant | recessive}] : Add one variant as a --linear\n",
            "                                                  or --logistic covariate.\n",
            "  --condition-list <f> [{dominant | recessive}] : Add variants named in the\n",
            "                                                  file as --linear/--logistic\n",
            "                                                  covariates.\n",
            "  --parameters <...>  : Include only the given covariates/interactions in the\n",
            "                        --linear/--logistic models, identified by a list of\n",
            "                        1-based indices and/or ranges of them.\n",
            "  --tests <all> [...] : Perform a (joint) test on the specified term(s) in the\n",
            "                        --linear/--logistic model, identified by 1-based\n",
            "                        indices and/or ranges of them.  If permutation was\n",
            "                        requested, it is based on this test.\n",
            "                        * Note that, when --parameters is also present, the\n",
            "                          indices refer to the terms remaining AFTER pruning by\n",
            "                          --parameters.\n",
            "                        * You can use \"--tests all\" to include all terms.\n",
            "  --vif <max VIF>     : Set VIF threshold for --linear multicollinearity check\n",
            "                        (default 50).\n",
            "  --xchr-model <code> : Set the X chromosome --linear/--logistic model.\n",
            "                        0 = skip sex and haploid chromosomes\n",
            "                        1 (default) = add sex as a covariate on X chromosome\n",
            "                        2 = code male genotypes 0/2 instead of 0/1\n",
            "                        3 = test for interaction between genotype and sex\n",
            "  --lasso-select-covars [cov(s)...] : Subject some or all covariates to LASSO\n",
            "                                      model selection.\n",
            "  --adjust ['gc'] ['log10'] ['qq-plot'] : Report some multiple-testing\n",
            "                                          corrections.\n",
            "  --lambda <val>   : Set genomic control lambda for --adjust.\n",
            "  --ci <size>      : Report confidence intervals for odds ratios.\n",
            "  --pfilter <val>  : Filter out association test results with higher p-values.\n",
            "  --aperm <min perms - 1> [max perms] [alpha] [beta] [init interval] [slope] :\n",
            "    Set up to six parameters controlling adaptive permutation tests.\n",
            "    * The first two control the minimum and maximum number of permutations that\n",
            "      may be run for each variant; default values are 5 and 1000000.\n",
            "    * The next two control the early termination condition.  A\n",
            "      100% * (1 - beta/2T) confidence interval is calculated for each empirical\n",
            "      p-value, where T is the total number of variants; whenever this\n",
            "      confidence interval doesn't contain alpha, the variant is exempted from\n",
            "      further permutation testing.  Default values are 0 and 1e-4.\n",
            "    * The last two control when the early termination condition is checked.  If\n",
            "      a check occurs at permutation #p, the next check occurs after\n",
            "      <slope>p + <init interval> more permutations (rounded down).  Default\n",
            "      initial interval is 1, and default slope is 0.001.\n",
            "  --mperm-save     : Save best max(T) permutation test statistics.\n",
            "  --mperm-save-all : Save all max(T) permutation test statistics.\n",
            "  --set-p <p-val>        : Adjust set test significant variant p-value ceiling\n",
            "                           (default 0.05).\n",
            "  --set-r2 [v] ['write'] : Adjust set test significant variant pairwise r^2\n",
            "                           ceiling (default 0.5).  'write' causes violating\n",
            "                          pairs to be dumped to <output prefix>.ldset.\n",
            "  --set-max <ct>         : Adjust set test maximum # of significant variants\n",
            "                           considered per set (default 5).\n",
            "  --set-test-lambda <v>  : Specify genomic control correction for set test.\n",
            "  --border <kbs>            : Extend --annotate range intervals by given # kbs.\n",
            "  --annotate-snp-field <nm> : Set --annotate variant ID field name.\n",
            "  --clump-p1 <pval> : Set --clump index var. p-value ceiling (default 1e-4).\n",
            "  --clump-p2 <pval> : Set --clump secondary p-value threshold (default 0.01).\n",
            "  --clump-r2 <r^2>  : Set --clump r^2 threshold (default 0.5).\n",
            "  --clump-kb <kbs>  : Set --clump kb radius (default 250).\n",
            "  --clump-snp-field <n...>  : Set --clump variant ID field name (default\n",
            "                              'SNP').  With multiple field names, earlier names\n",
            "                              take precedence over later ones.\n",
            "  --clump-field <name...>   : Set --clump p-value field name (default 'P').\n",
            "  --clump-allow-overlap     : Let --clump non-index vars. join multiple clumps.\n",
            "  --clump-verbose           : Request extended --clump report.\n",
            "  --clump-annotate <hdr...> : Include named extra fields in --clump-verbose and\n",
            "                              --clump-best reports.  (Field names can be\n",
            "                              separated with spaces or commas.)\n",
            "  --clump-range <filename>  : Report overlaps between clumps and regions.\n",
            "  --clump-range-border <kb> : Stretch regions in --clump-range file.\n",
            "  --clump-index-first       : Extract --clump index vars. from only first file.\n",
            "  --clump-replicate         : Exclude clumps which contain secondary results\n",
            "                              from only one file.\n",
            "  --clump-best              : Report best proxy for each --clump index var.\n",
            "  --meta-analysis-chr-field <n...> : Set --meta-analysis chromosome, variant\n",
            "  --meta-analysis-snp-field <n...>   ID, position, A1/A2 allele, p-value,\n",
            "  --meta-analysis-bp-field <n...>    standard error, and/or effective sample\n",
            "  --meta-analysis-a1-field <n...>    size field names.\n",
            "  --meta-analysis-a2-field <n...>    Defaults are 'CHR', 'SNP', 'BP', 'A1',\n",
            "  --meta-analysis-p-field <n...>     'A2', 'P', 'SE', and 'NMISS',\n",
            "  --meta-analysis-se-field <n...>    respectively.  When multiple parameters\n",
            "  --meta-analysis-ess-field <n...>   are given to these flags, earlier names\n",
            "                                     take precedence over later ones.\n",
            "                                     Note that, if the numbers of cases and\n",
            "                                     controls are unequal, effective sample\n",
            "                                     size should be\n",
            "                                       4 / (1/<# cases> + 1/<# controls>).\n",
            "  --meta-analysis-report-dups      : When a variant appears multiple times in\n",
            "                                     in the same file, report that.\n",
            "  --gene-list-border <kbs>   : Extend --gene-report regions by given # of kbs.\n",
            "  --gene-subset <filename>   : Specify gene name subset for --gene-report.\n",
            "  --gene-report-snp-field <> : Set --gene-report variant ID field name (default\n",
            "                               'SNP').  Only relevant with --extract.\n",
            "  --gap <kbs>      : Set \"--fast-epistasis case-only\" min. gap (default 1000).\n",
            "  --epi1 <p-value> : Set --[fast-]epistasis reporting threshold (default\n",
            "                     5e-6 for 'boost', 1e-4 otherwise).\n",
            "  --epi2 <p-value> : Set threshold for contributing to SIG_E count (def. 0.01).\n",
            "  --je-cellmin <n> : Set required number of observations per 3x3x2 contingency\n",
            "                     table cell for joint-effects test (default 5).\n",
            "  --q-score-range <range file> <data file> [i] [j] ['header'] :\n",
            "    Apply --score to subset(s) of variants in the primary score list based\n",
            "    on e.g. p-value ranges.\n",
            "    * The first file should have range labels in the first column, p-value\n",
            "      lower bounds in the second column, and upper bounds in the third column.\n",
            "      Lines with too few entries, or nonnumeric values in the second or third\n",
            "      column, are ignored.\n",
            "    * The second file should contain a variant ID and a p-value on each\n",
            "      nonempty line (except possibly the first).  Variant IDs are read from\n",
            "      column #i and p-values are read from column #j, where i defaults to 1 and\n",
            "      j defaults to i+1.  The 'header' modifier causes the first nonempty line\n",
            "      of this file to be skipped.\n",
            "  --R-port <port #>  : Connect to Rserve on a port other than 6311.\n",
            "  --R-host <host>    : Connect to Rserve host.\n",
            "  --R-socket <sock>  : Connect to Rserve socket.\n",
            "  --parallel <k> <n> : Divide the output matrix into n pieces, and only compute\n",
            "                       the kth piece.  The primary output file will have the\n",
            "                       piece number included in its name, e.g. plink.rel.13 or\n",
            "                       plink.rel.13.gz if k is 13.  Concatenating these files\n",
            "                       in order will yield the full matrix of interest.  (Yes,\n",
            "                       this can be done before unzipping.)\n",
            "                       N.B. This generally cannot be used to directly write a\n",
            "                       symmetric square matrix.  Choose square0 or triangle\n",
            "                       shape instead, and postprocess as necessary.\n",
            "  --memory <val>     : Set size, in MB, of initial workspace malloc attempt.\n",
            "                       (Practically mandatory when using GNU parallel.)\n",
            "  --threads <val>    : Set maximum number of concurrent threads.\n",
            "                       This has one known limitation: some BLAS/LAPACK linear\n",
            "                       algebra operations are multithreaded in a way that PLINK\n",
            "                       cannot control.  If this is problematic, you should\n",
            "                       recompile against single-threaded BLAS/LAPACK.\n",
            "  --d <char>         : Change variant/covariate range delimiter (normally '-').\n",
            "  --seed <val...>    : Set random number seed(s).  Each value must be an\n",
            "                       integer between 0 and 4294967295 inclusive.\n",
            "  --perm-batch-size <val> : Set number of permutations per batch for some\n",
            "                            permutation tests.\n",
            "  --native           : Allow Intel MKL to use processor-dependent code paths.\n",
            "  --output-min-p <p> : Specify minimum p-value to write to reports.\n",
            "  --debug            : Use slower, more crash-resistant logging method.\n",
            "\n",
            "Primary methods paper:\n",
            "Chang CC, Chow CC, Tellier LCAM, Vattikuti S, Purcell SM, Lee JJ (2015)\n",
            "Second-generation PLINK: rising to the challenge of larger and richer datasets.\n",
            "GigaScience, 4.\n",
            "\n",
            "For further documentation and support, consult the main webpage\n",
            "(https://www.cog-genomics.org/plink/1.9 ) and/or the mailing list\n",
            "(https://groups.google.com/d/forum/plink2-users ).\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#set variables\n",
        "input_dir=\"drive/MyDrive/statgen_workshop/data/workshop1/inputs/penncath_withPheno\"\n",
        "analysis_dir=os.getcwd() + \"/analysis/\"\n",
        "os.environ['input_dir']=input_dir\n",
        "os.environ['analysis_dir']=analysis_dir"
      ],
      "metadata": {
        "id": "8OZDaIr9e3t0"
      },
      "execution_count": 39,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#create analysis dir, this is going to be ephemeral and sits on the hosted runtime environment\n",
        "mkdir -p ${analysis_dir}"
      ],
      "metadata": {
        "id": "K5EonQFmtdbu"
      },
      "execution_count": 43,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "load R libraries"
      ],
      "metadata": {
        "id": "RxHN_mS3rh3z"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%R -i input_dir -i analysis_dir\n",
        "library(tidyverse)\n",
        "library(data.table)\n",
        "library(glue)\n",
        "\n",
        "print(input_dir)\n",
        "print(analysis_dir)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pP64D7rjilwL",
        "outputId": "85f98c95-1540-4a08-f759-69e5ea2a3fb6"
      },
      "execution_count": 30,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            " Attaching core tidyverse packages  tidyverse 2.0.0 \n",
            " dplyr     1.1.2      readr     2.1.4\n",
            " forcats   1.0.0      stringr   1.5.0\n",
            " ggplot2   3.4.2      tibble    3.2.1\n",
            " lubridate 1.9.2      tidyr     1.3.0\n",
            " purrr     1.0.1     \n",
            " Conflicts  tidyverse_conflicts() \n",
            " dplyr::filter() masks stats::filter()\n",
            " dplyr::lag()    masks stats::lag()\n",
            " Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: data.table 1.14.8 using 1 threads (see ?getDTthreads).  Latest news: r-datatable.com\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
            "Attaching package: data.table\n",
            "\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from package:lubridate:\n",
            "\n",
            "    hour, isoweek, mday, minute, month, quarter, second, wday, week,\n",
            "    yday, year\n",
            "\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from package:dplyr:\n",
            "\n",
            "    between, first, last\n",
            "\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following object is masked from package:purrr:\n",
            "\n",
            "    transpose\n",
            "\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[1] \"drive/MyDrive/statgen_workshop/data/workshop1/inputs/penncath_withPheno\"\n",
            "[1] \"/content/analysis/\"\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# QC steps"
      ],
      "metadata": {
        "id": "KhyuNNg1rJ_F"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Calculate call rates for flagging poorly called SNPs and individuals\n",
        "the penncath.lmss file shows the per variant missingness and the penncath.imiss file shows the per individual missingness"
      ],
      "metadata": {
        "id": "8nKM1SoKr5xT"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash -s \"$analysis_dir\" \"$input_dir\"\n",
        "plink --bfile ${input_dir}/penncath --missing --out ${analysis_dir}/penncath"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rjiXbONrr-XQ",
        "outputId": "ab628437-cd98-4ddf-a19a-d7010508397e"
      },
      "execution_count": 44,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "Logging to /content/analysis//penncath.log.\n",
            "Options in effect:\n",
            "  --bfile drive/MyDrive/statgen_workshop/data/workshop1/inputs/penncath_withPheno/penncath\n",
            "  --missing\n",
            "  --out /content/analysis//penncath\n",
            "\n",
            "12982 MB RAM detected; reserving 6491 MB for main workspace.\n",
            "861473 variants loaded from .bim file.\n",
            "1401 people (937 males, 464 females) loaded from .fam.\n",
            "933 phenotype values loaded from .fam.\n",
            "Using 1 thread (no multithreaded calculations invoked).\n",
            "Before main variant filters, 1401 founders and 0 nonfounders present.\n",
            "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
            "Total genotyping rate is 0.977517.\n",
            "--missing: Sample missing data report written to\n",
            "/content/analysis//penncath.imiss, and variant-based missing data report\n",
            "written to /content/analysis//penncath.lmiss.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "let's look at the distribution of per variant call rates   "
      ],
      "metadata": {
        "id": "eJyjiIZ01GUE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "variant_miss <- fread(glue(analysis_dir, \"penncath.lmiss\"))\n",
        "\n",
        "variant_miss %>% ggplot( aes(x = F_MISS)) +\n",
        "  geom_histogram() +\n",
        "  labs(title = \"Variant Missing Rate\") +\n",
        "  theme_classic()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 514
        },
        "id": "pmvb4c5L1IJ2",
        "outputId": "0b41b9df-4a6f-4087-85ee-e6efa438a326"
      },
      "execution_count": 45,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAMAAABKCk6nAAACx1BMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgaGhobGxsdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dZWVlaWlpcXFxdXV1eXl5fX19hYWFiYmJjY2NkZGRlZWVnZ2doaGhpaWlqampsbGxtbW1ubm5vb29wcHBxcXFycnJ0dHR1dXV2dnZ3d3d5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5udnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWnp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vNzc3Ozs7Pz8/Q0NDR0dHT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7///8M6cZRAAANg0lEQVR4nO3b+3sU5RnG8VSKVKkUgwqlnAqiRSgFi1ilYqktShG0QKGE0gKl1YIWCkk2hw3oGklCKIVKiOAaoiGQwEoCgSLBlBQLRQolStqQYIw5gDk8f0RnZmdykEmWJDPszr3394eZnZl3X97kc+0mG66JEAZdRLAXwOyNwOARGDwCg9dt4Om/13bT/tB2qizi8y8Neut//n3k0GZ1NyryxjE3Pkl7Qt9+/e4Y67lxJtbDug2cFXlN2X5027/bTjWXt3xp0NjT/n3k0H3K9vCQyBvH3Pgk7QnZIk1FA7JvmIn1sG4DNw7JVLarZsiBiSOGeaSsz6Z7fMqL0X/08Vd2zBo/69qciBHbtMGRS2cr20VLtFdw8+9Gfnt8vug75YQ+WjLGjP7NnCT/EzTaH/1R9Am1mT54dMzIP1v3JYdX3f8ZvO5xkS/ue/f6gLeltM+ZioiXmxWr1qM4aRqdKRHGK7iw/3+lZqBPAz4wvF6OLRR9p5zQR1/u+zfJ7LfB/wT1Fezrf1D0CdWZPr1vh3wy+JCVX3UY1X3gT/qek7eHNkm18uN16LtXIv6p/Tg1ji6IzIxvAz79nFu2zjmlAf/j62mXlXP6Tjmhj84cqxyO1oHvHDDgq+N2KY/8E6ozvTlYOXzhV5Z8ueFXD36LfvZFmRErkjZ18sP9sq9EVGjAxtEVkVmx7YAPjJZH9vmB5fBzAyfuF32nAWujX5+qjJze9gp+4TH1kX9Cdaa024cNGzZ4rlVfcZjVA2DfvZf6lcvBuy+IDNZIFat2Rx2BW4bv/GbTKeO36MZtd9TpuzbgHd9RTo1tA/7svu1iTK/OlDPakq80TOsBcMuYp54XeWNMk2y6K0MHbnekAvc94h8aeVqiB78kfuDtUV9I6e31+q4N+EKfEvHe2QYs2wZVGdOrM9Xc845cW8GfwT2rJ3/oeDWiUKT2qeGTtq8e4PUDtztSgRfduU4bqQCX9flIB675xdCRD+0RfdcGLMnDxq36yUb/E1TglslRxvSF6kzFU0eNWnbNuq85rAqFv2SpfwyZsi3YqwAtBIAbBmVLydf+FexlgBYCwJL7wLfufyPYi0AtFICZjREYPAKDR2DwCAxed4Gvmf03PQvdugv8l5/ZsgxmVwQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgWQI8r11WLIpZF4HBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8EyAW9ITouv1x0UidfHujc3Fy12uCvUEgR2WCXBJplyqqU1I2dQo4hHJPizeYwU+/SKBHZYJ8K6tKWnXvUdkz3FfUlRSlqdMSjL2rt+Qrnhnr37+aZNJCBy6mQDvzJF38lNjPHE+7RWsAu+urJSsfJFTWS88YzIJgUM3E+Cj74ov13tCquo14JxC2XWytFzyctWLfIt2WCbATZ6k2Loatye2WjtscCckt1xcmxiv/eJFYIfFj0ngERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8U+CCF41HRSJ18e6NzcXLXa4K9QSBHZYZ8KevrpHahJRNjSIekezD4j1W4NOvEdhhmQG/dmWNeI/InuO+pKikLE+ZlGTsXb8hXfF+/6/LnzF5AoFDNxPgo/tljaTGeOJ82itYBd5dWSlZ+SJ5cQsJ7KxMgD0ez8J87wmpqteAcwpl18nScsnLVS/yLdphmf8WvUZq3J7Yau1xgzshueXi2sT4evWIwA6LH5PAIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg2cAL1Q3jwYeT2CH5QfO/N5dEyZMGPtg4PEEdlj6K7h2wdmzZ881BB5PYIdlvEV//lb65s2bA48nsMMygKfNW6kUeDyBHZYBPP8mxxPYYRnAy+pvbjyBHZYB/OS906ZPnx54PIEdlgFccFQt8HgCOywDOENtR+DxBHZYBnBUVNSiB+YGHk9gh9X+b9FNSwOPJ7DD6vCfDXMCjyewwzKAZylN+Wng8QR2WAZwXl5e/vtNgccT2GEZwNe2Lv31GwTGywBePC8t7dkVgccT2GEZwD9s3XQdgR2WAfyI8vbc9Fjg8QR2WAbwnx5evXpyTODxBHZYrZ+DC+ITj9zEeAI7rFbgWpGrNzGewA7LAN78hMjULYHHE9hhGcCTakTqJwUeT2CHZQBPbBZpnBx4PIEdlgH80uNrXp60LvB4Ajus1l+yDsbEFd7EeAI7LN6bBB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDJ4JcHV0QnSD/rhIpC7evbG5eLnLVaGeILDDMgEuPS07i2sTUjY1inhEsg+L91iBT79IYIdl+hbdGH3Ve0T2HPclRSVlecqkJGPv+g3pivd/iqNnmYwncOhmBlydeF5SYzxxPu0VrALvrqyUrHyR7Qt+PNPkCQQO3UyAa2KrRLwnpKpeA84plF0nS8slL1e9yLdoh2UCnLnS5TpR4/bEVmuHDe6E5JaLaxPj69UjAjssfkwCj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgWQ48j9ohFYHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDx7gakd9AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBu4XA1A5GBAaPwOARGLxgAVP7FkVg8EICmNr2ZQJcF+/e2Kw/LtIPW8/ZD0xtSzMBzj4s3mO1CSmbGkU8+qG2US/eYmBq9zYTYE+ZlGR4j8ie476kqKQs7VDbiDR8lhZEYNL3oE6Ad6fGeOJ82itYO9Q2IgmTJs285UtkvckEOKdQdp30npCqeg1YO9Q26kXTt2gWupkAN7gTkltq3J7Y6rZDbaMeEdhhWfIxiYVuBAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHB6y7wWw8910lz5nR2xYpmz7Vx8rmzbZw8aN+YN3sELLu3dNLiJZ1dsaJpLhsnd02zcfItSxbbOfsTsZ1dOdkz4E5LTrFsKpOePmPj5GeetnFySUm2c/bZp7u+bh3woQLLpjIptdzGyctTbZxcCg7ZOXvaJ11ftw6YhWQEBq/3wPptSx1vYbIsfcbq6ITohuLlLleFHbNrE9u29vdcrt/us3ztzVkLWv+Nrpbee2D9tqWOtzBZlj5j6WnZWVzgs3Tqttm1iW1bu0hTXIPla7/64arWf6OrpfceWL9tqf0tTBbWOmNj9NW96zekN9oxuzaxjWvfXyTWr11Wtf4bXS3dGuDdHW9hsjBjxurE81JZKVn5dsyuTWzf2ltebrFh7W3AXX7bew+s37bU8RYmy9JnrImtUt6nyyUv147ZtYltW7ucS7dj7X7ggN/23gNrty2df6XjLUyWpc+eudLlOnFxbWJ8vR2zaxPbtnZ5L0fE8rWfdc13Hb+Jbzs/JoFHYPAIDB6BwSMweGEIXH7bKKUi/SDiNWVbGOGTPp9LzpTvj5vfusMoHIEj2x+MmKhsfzlcBW4YdFlalqbou2Atz+LCHvi7P/i71I2eqwJX9v9YPaXvQAp74AnpK2Tnyp9rb9GJ35jpuSTGDqNwBO5zv9JF/WDC1SHXn/zADyzV2cvuzmjdQRSOwB1fwfJs6oOiATd/qpzY+4i+C9LqrI7Akj0wxg+cO75KJHaBvgva+qyNwHI98oIfWF4fP2H8vApjh1EYAodXYQt8aYbWh8Feh92FLXC4RGDwCAwegcEjMHj/Bx5fNY9ILvLeAAAAAElFTkSuQmCC\n"
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "If we want to filter out the variants with missing rate >0.01, how many are there?"
      ],
      "metadata": {
        "id": "MGJq2nQB1WRz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "variant_miss %>% mutate(missing_rate_gt0.01 = if_else(F_MISS > 0.01, \"Yes\", \"No\")) %>%\n",
        "                 group_by(missing_rate_gt0.01) %>%\n",
        "                 tally()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Taxfw0SJ1YvW",
        "outputId": "d6e35491-745b-409d-f440-c43c07375c55"
      },
      "execution_count": 46,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# A tibble: 2  2\n",
            "  missing_rate_gt0.01      n\n",
            "  <chr>                <int>\n",
            "1 No                  569723\n",
            "2 Yes                 291750\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "We will use plink to filter out the 291750 SNPs and"
      ],
      "metadata": {
        "id": "0bNzOhrh1dvL"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "plink --bfile ${input_dir}/penncath --geno 0.01 --make-bed --out ${analysis_dir}/penncath.geno0.01"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ULEEfQO_1hKb",
        "outputId": "e9eebf47-d1f2-496f-932f-58400c5c17b2"
      },
      "execution_count": 47,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "Logging to /content/analysis//penncath.geno0.01.log.\n",
            "Options in effect:\n",
            "  --bfile drive/MyDrive/statgen_workshop/data/workshop1/inputs/penncath_withPheno/penncath\n",
            "  --geno 0.01\n",
            "  --make-bed\n",
            "  --out /content/analysis//penncath.geno0.01\n",
            "\n",
            "12982 MB RAM detected; reserving 6491 MB for main workspace.\n",
            "861473 variants loaded from .bim file.\n",
            "1401 people (937 males, 464 females) loaded from .fam.\n",
            "933 phenotype values loaded from .fam.\n",
            "Using 1 thread (no multithreaded calculations invoked).\n",
            "Before main variant filters, 1401 founders and 0 nonfounders present.\n",
            "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
            "Total genotyping rate is 0.977517.\n",
            "291750 variants removed due to missing genotype data (--geno).\n",
            "569723 variants and 1401 people pass filters and QC.\n",
            "Among remaining phenotypes, 463 are cases and 470 are controls.  (468\n",
            "phenotypes are missing.)\n",
            "--make-bed to /content/analysis//penncath.geno0.01.bed +\n",
            "/content/analysis//penncath.geno0.01.bim +\n",
            "/content/analysis//penncath.geno0.01.fam ... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\bdone.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "After removing the 291750 SNPs, we will recalculate the missing rate for the samples"
      ],
      "metadata": {
        "id": "trg2P5yc1z6g"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "plink --bfile ${analysis_dir}/penncath.geno0.01 --missing --out ${analysis_dir}/penncath.geno0.01"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "IdMaAjsc123D",
        "outputId": "09a6760f-ca50-4174-8faa-624bb211d2f4"
      },
      "execution_count": 48,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "Logging to /content/analysis//penncath.geno0.01.log.\n",
            "Options in effect:\n",
            "  --bfile /content/analysis//penncath.geno0.01\n",
            "  --missing\n",
            "  --out /content/analysis//penncath.geno0.01\n",
            "\n",
            "12982 MB RAM detected; reserving 6491 MB for main workspace.\n",
            "569723 variants loaded from .bim file.\n",
            "1401 people (937 males, 464 females) loaded from .fam.\n",
            "933 phenotype values loaded from .fam.\n",
            "Using 1 thread (no multithreaded calculations invoked).\n",
            "Before main variant filters, 1401 founders and 0 nonfounders present.\n",
            "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
            "Total genotyping rate is 0.997631.\n",
            "--missing: Sample missing data report written to\n",
            "/content/analysis//penncath.geno0.01.imiss, and variant-based missing data\n",
            "report written to /content/analysis//penncath.geno0.01.lmiss.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "take a look at per sample missing rate"
      ],
      "metadata": {
        "id": "3DiYFmWR2An0"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "sample_miss <- fread(glue(analysis_dir, \"penncath.geno0.01.imiss\"))\n",
        "sample_miss %>% mutate(missing_rate_gt0.01 = if_else(F_MISS > 0.01, \"Yes\", \"No\")) %>%\n",
        "                 group_by(missing_rate_gt0.01) %>%\n",
        "                 tally()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ashEy9WA2Cpi",
        "outputId": "489d11a5-e2d9-4066-d584-2e032a277472"
      },
      "execution_count": 49,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# A tibble: 2  2\n",
            "  missing_rate_gt0.01     n\n",
            "  <chr>               <int>\n",
            "1 No                   1380\n",
            "2 Yes                    21\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "we will remove the 21 samples with >0.01 missing rate"
      ],
      "metadata": {
        "id": "3fLFkOmy2GZ5"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "plink --bfile ${analysis_dir}/penncath.geno0.01 --mind 0.01 --make-bed --out ${analysis_dir}/penncath.geno0.01.mind0.01"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "TrntuJSk2IFd",
        "outputId": "9eaf3375-df2c-467e-9e5f-dfa656933271"
      },
      "execution_count": 50,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "Logging to /content/analysis//penncath.geno0.01.mind0.01.log.\n",
            "Options in effect:\n",
            "  --bfile /content/analysis//penncath.geno0.01\n",
            "  --make-bed\n",
            "  --mind 0.01\n",
            "  --out /content/analysis//penncath.geno0.01.mind0.01\n",
            "\n",
            "12982 MB RAM detected; reserving 6491 MB for main workspace.\n",
            "569723 variants loaded from .bim file.\n",
            "1401 people (937 males, 464 females) loaded from .fam.\n",
            "933 phenotype values loaded from .fam.\n",
            "21 people removed due to missing genotype data (--mind).\n",
            "IDs written to /content/analysis//penncath.geno0.01.mind0.01.irem .\n",
            "Using 1 thread (no multithreaded calculations invoked).\n",
            "Before main variant filters, 1380 founders and 0 nonfounders present.\n",
            "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
            "Total genotyping rate in remaining samples is 0.997808.\n",
            "569723 variants and 1380 people pass filters and QC.\n",
            "Among remaining phenotypes, 455 are cases and 463 are controls.  (462\n",
            "phenotypes are missing.)\n",
            "--make-bed to /content/analysis//penncath.geno0.01.mind0.01.bed +\n",
            "/content/analysis//penncath.geno0.01.mind0.01.bim +\n",
            "/content/analysis//penncath.geno0.01.mind0.01.fam ... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\bdone.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Next, we look at the HWE statistics"
      ],
      "metadata": {
        "id": "on2V22Mn2SI-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "plink --bfile ${analysis_dir}/penncath.geno0.01.mind0.01 --hardy --out ${analysis_dir}/penncath.geno0.01.mind0.01"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "o00uoDcs2UXK",
        "outputId": "5d83ffec-9962-4ca5-fa07-1bc7e618944b"
      },
      "execution_count": 52,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "Logging to /content/analysis//penncath.geno0.01.mind0.01.log.\n",
            "Options in effect:\n",
            "  --bfile /content/analysis//penncath.geno0.01.mind0.01\n",
            "  --hardy\n",
            "  --out /content/analysis//penncath.geno0.01.mind0.01\n",
            "\n",
            "12982 MB RAM detected; reserving 6491 MB for main workspace.\n",
            "569723 variants loaded from .bim file.\n",
            "1380 people (922 males, 458 females) loaded from .fam.\n",
            "918 phenotype values loaded from .fam.\n",
            "Using 1 thread (no multithreaded calculations invoked).\n",
            "Before main variant filters, 1380 founders and 0 nonfounders present.\n",
            "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
            "Total genotyping rate is 0.997808.\n",
            "--hardy: Writing Hardy-Weinberg report (founders only) to\n",
            "/content/analysis//penncath.geno0.01.mind0.01.hwe ... 0%\b\b0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\bdone.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Examine HWE statistics"
      ],
      "metadata": {
        "id": "N75SOdNy2cnk"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "hwe_stat <- fread(glue(analysis_dir, \"penncath.geno0.01.mind0.01.hwe\"))\n",
        "\n",
        "\n",
        "hwe_stat %>% ggplot( aes(x = P)) +\n",
        "  geom_histogram() +\n",
        "  labs(title = \"Hardy-Weinberg equilibrium exact test p-values\") +\n",
        "  theme_classic()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 514
        },
        "id": "eWPRoGwT2kzC",
        "outputId": "5e519c00-f9c6-4f54-a2f1-e9940433243e"
      },
      "execution_count": 53,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAMAAABKCk6nAAAC61BMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVXV1dZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7///9rZcGfAAAQr0lEQVR4nO3cf3xVdR3H8UuEmYZmlJEiig60zIUU/sASI0tTWISlGZqznzP7oSSygO3u3t1d8jb3A9QCZZKM21iNxZZX9gMXMmotNUgprK55GdsYY7ubjrvPn51zdu52GdvuPffe487evF8PH3f3fs89n/vdfXovmz4uNmHQ2cZ7A8zcCAwegcEjMHjGgaeVq5cL3KccuC1y6dN25SJr6jsiwTOfH1jy247L8Guxppwx+E/EmsEpatsOn/z11CNj7CGOxxvvzAJes0C5uH56rcjOqW8PLIUC/fpB40+Vcq5y0sBlxJrBKWqXv3Ly11OPjNbpB/zHuZfM9Ih/csFHDmyeNfv+L7mvzVOOfOJX6vG/TG6Xo1NXrhD54TL58w2zZz2pPkFvTtqcdlVar9/25OwPfzMYXlcHlMxO+f6yPO0BBlZl8yUp6V9xHLT1iCwqiHgFa6eq5/hsx/Wjb0761eKLVq9ffKn+r9jAhKeueFsqLmjXtyn1qefP3SXLbJdsVO+ifR24X+hHsy67qnrwyBuTHv9S6j1B9ep1v1Au5jwd/kYHH+/kE61dAsBvn/tbaZ68v8W2MvTWGQ1S9173U5eLvHzWMfV4/8dKZdsXXkwVmb3p6Ec3y/+m71KeoBZbjpxI2eK3fSfUnpKvr2sDpvxJtrxP89FX35pSL7XvdZ8KrJ2qnjP0hLfYlOf8Peuk+Yy3IybI4tWdM54Pb/PoOX+Qyg8cE5v+OlW+6vf748VBaVgu4SN+W6a8c/Vj6tXi+SJNZx/XJww93rATLV0cwB/8qNIUBaMjJDLj90dsf5ffpCgHPuXuPLtBHvnGwN3uv1fuc/ed89ah9xx+brpy+yffVp6gI7Z/idzu8NteFVmxVF9XB2xR/tWQFA1YX/3NpcqXq0YA1k5Vzxl6wo/Y/HLMdlB6bf+NmCCHL1jygIS3uU2ddzgUCazf728fKH5LX1Hz294QyVyiXm0749/y0+XhCUOPN+xESxcH8KaA0nwFo3jB/GveV37E1iKPX6McuNkt935LLvHJ09OmLZTfTe+/4GVZsrH4Wik+Y+bMmdPv0ICPiKRl+21tInk36uvaAPWP7EUasL76uPLqkS+OAKydqp4TCXxUjtsC0qfaDE4Q+bHtnxLepvpq1L7hIeDw/eq+dt7cnRHAR0Xcn9W+ibS8/gtrwhOGHm/YiZYugbfo5z+kvBynl6toz16mLF3plt3nvDAjJMFAoFW6z/R+TKTwrmU5UpGinRgJ/JrII8v0dXVx8yeUK5drwPrqswOv4EO2bpHPRAJrp6rnKLf0o8OA9Qny9+kP3iLhbZbNUFb2ByOBw/cT6dv4/u4h4H8o7xFp2jfh/Uz9zP7whKHHG3aipUsA+JnZJ6Rgaon6XP9ncp1Uq+/as2c9Gr7fLanfEnn9wvP/Jp0f+Z30PrArEvin0jlnvb6uLv5rcpN4z9KA9dX/TK6Xyinu4JS/yMtnRgJrp+rA+tFhwPqEd67e2De3OLzNY+dulRfOPirKH+1aylf9fpvS31H+9A6Gj/htP5Hjcx7Xrvee9/VMCU8YerxhJ1q6BIC7br143qYV53pVtCdnXHTPUuWHYKft9fD9imxblcuLL1IuGhdceukPeoeA35i06VMXpvfq69pi/swrH/7yOu3EgVV15L23uuWxlJsz0jyDwPqpOrB+dBiwPuHRm0X+eu7r+jZrG66cdqXyG/k9Z63W7qN+Hbhf5zdnzPrk9sEjflvx3Avu7R34Jr496aCEv9GSwccbdqKlS/J/ydpyY5wnKj/GyHUbh6/edupv26Y3MX/dHbXkAh+eHefPHD0fLpemM/8xfJnACZdU4LXnr4n31MorLprzzCmrBE44/s8G8AgMHoHBIzB4BAbPKHAv1I+Yp0FGgX/9FVO2wcyKwOARGDwCg0dg8AgMHoHBGwG4McNub9Gv7xbpdrjWhQbXCDzBGgG4xqdcdDkLC/pEPCLldeJt0NbUCDzBGgF4xxr3hj5vvWzf48tLzyvz+KWpRFsTeenpjCXv/h5ZAo0A3NoqZdVFWZ4cn/YKVoFLtTWRqpzlBJ5YjQDcHJCqSu9eaQtqwBW1snWftqYe5Fv0BGsE4EOZuY5gp8uT3aHd7HE58/u1NfUWgSdY/DUJPAKDR2DwCAwegcEjMHgEnrDdGdmo9yLwhI3A4BEYPAKDR2DwCAwegcEjMHgEBo/A4BEYPAKDR2DwCAwegcEjMHgEBo/A4BEYPAKDR2DwCAwegcEjMHgEBo/A4BEYPAKDR2DwCAwegcEjMHgEBo/A4BEYPAKDR2DwCAwegcEjMHgEBo/A4BEYPAKDR2DwCAwegcEjMHgEBo/A4BEYPAKDR2DwCAwegcEjMHgEBo/A4BEYPAKDR2DwCAwegcEjMHgEBo/A4CUAXPNQ+NpukW6Ha12oMcNub1EXCGyZ4gc++tgq6XIWFvSJeETK68TbUOPTjxHYMsUP/Msjq8RbL9v3+PLS88o8fmkq2bHGvUHxDuzPTTNvx8xQcQO/uFNWSVGWJ8envYJV4NLWVimrFlm/5KbbTdwyM1LcwB6PZ3m1d6+0BTXgilrZuq85IFWV6kG+RVumRH6KXiWdLk92h3a9x+XM7z+UmesIqrcIbJn4axJ4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAYvbuA3VzsyO/Tru0W6Ha51ocYMu71FXSCwZYob+N9H5MnmLmdhQZ+IR6S8TrwNNT79IIEtU/xv0W/+zNHvrZfte3x56XllHr80lexY496get90/e3m7ZgZKpE/g7fUFmV5cnzaK1gFLm1tlbJqkcD+3DSz9ssMFjfwtpdlZ6V3r7QFNeCKWtm6rzkgVZXqQb5FW6a4gQ+vdq7t7nR5sgd+0upxOfP7D2XmOoLqLQJbJv6aBB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweAQGj8DgERg8AoNHYPAIDB6BwSMweMaAl6sXN0SfSmDLZAR4y9VTU1NTL/949KkEtkyGXsFddx84cOC1nuhTCWyZjL1FH9+2Yf369dGnEtgyGQNeeOeDStGnEtgyGQO+K8apBLZMxoB/EIxtKoEtkzHgm89fuGjRouhTCWyZjAHXvKgWfSqBLZMx4BK1zdGnEtgyGQNOT0+/54o7ok8lsGUyBqx24rvRpxLYMhkHlmXRpxLYMhkDTlO67rboUwlsmYwBV1VVVb90IvpUAlsmY8C9T333e88QeCJlDPi+O4uLv/pA9KkEtkzGgG8avBg7AlsmY8DXK2/PJz4XfSqBLZMx4J9fs2LF/KzoUwlsmQz+HlzjyK2PYSqBLZNB4C6R9himEtgyGQNe/3mRBU9En0pgy2QMeF6nSHBe9KkEtkzGgOeGRPrmR59KYMtkDPiRG1etnLc6+lQCWyaDP2Q9n5VTG8NUAlsmg8AxRmDLRGDwCAwegcEjMHgEBo/A4BEYPAKDR2DwCAwegcEjMHgEBo/A4BEYPAKDR2DwCAxe3MAda51rw39r5W6RbodrXagxw25vURcIbJniBm5+RZ5t7HIWFvSJeETK68TbUOPTDxLYMiXwFt23tt1bL9v3+PLS88o8fmkq2bHGvUHxrspZvsS0DTNjxQ/ckXtQirI8OT7tFawCl7a2Slm1yEtPZxDYKsUN3JndJuLdK21BDbiiVrbuaw5IVaV6kG/Rlilu4C0P2u17O12e7A7tZo/Lmd9/KDPXof11pQS2TPw1CTwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGLy4gUNldw9e3y3S7XCtCzVm2O0t6gKBLVPcwO2vPizS5Sws6BPxiJTXibehxqcfJLBlSuAtWgH21sv2Pb689Lwyj1+aSnascW9QvA/4VqaZtmFmrMSAi7I8OT7tFawCl7a2Slm1yHPfT1ts2oaZsRJ8Be+VtqAGXFErW/c1B6SqUj3Et2jLFDfwAftd9j2dLk92h3azx+XM7z+UmesIqrcIbJn4axJ4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAaPwEkvtqfUWrshsIEIDB6BwSMweAROejE+pe/SMz/6w8RzxLzdRHY6ASf8XBN4zIc18em503gJP8y79b3FsZvIzAU2sUQfNMbzR7+biUdiK7bzJyzwuBfjk53o5Bif6VHvRWDrFSNFbOcQ2HrFSBHbOQS2eAQGj8DgEfh0isDgERg8AoNHYPAIDN54AY/Tt3v6RWDwCAwegcEjMHgEBo/A4BEYPAKDR2DwCAwegcEjMHgEBo/A4BH4tI3A4BEYPCPA3Q7XupB+fbd+c3CNwNbMCHB5nXgbupyFBX0iHv2mdkFg62YE2OOXphJvvWzf48tLzyvTbmoXIj3HiglsyYwClxZleXJ82itYu6ldiDjnzbt91FHMio0AXFErW/d590pbUAPWbmoX6sHT+q9RmoiNANzjcub3d7o82R1DN7UL9RaBJ1gW/2uUWKIRGDwCg0dg8AgMHoHBIzB4BAaPwOARGDwCg0dg8AgMHoHBIzB4BAbPKPC2T35tlJYtG+1IMlp6h4nD71hq4vBxe2KeiwtYSp8YpfvuH+1IMlpoN3G4faGJw5+4/z4zp38+e7Qj++IDHrX8wqSNGqHF+00cvn+xicOlMN/M6UtfGft48oB31SRt1AgVBUwcHigycbjU7DJzevH/xj6ePGBmyQgMXuLA+seWTv4IU9LSJ3asda7tacyw21vMmK4NNm3vL9jtP/xD0vceKrt78DHG2nriwPrHlk7+CFPS0ic2vyLPNtb4kjp6aLo22LS9i5zI6Un63ttffXjwMcbaeuLA+seWIj/ClMQGJ/atbd+xxr2hz4zp2mAT975ztyR/7/Lw4GOMtfXkAJee/BGmJBae2JF7UFpbpazajOnaYPP23r+y34S9DwGP+bQnDqx/bOnkjzAlLX1iZ3ab8j4dkKpKM6Zrg03bu7y2wYy9DwBHfdoTB9Y+tnTwFyd/hClp6dO3PGi37z2UmesImjFdG2za3uWFCpGk7/2A/S77nhiedv6aBB6BwSMweAQGj8DgEVgCk+fMSfl6+3hvw6QILIFpIqHvPDTe2zApAmvAUnbbeG/DpAisAfd9I2e8t2FSBFb+DE5NverHveO9DZMi8MBbNGwEJjB6BGYTOAKDR2DwCAwegcEjMHgEBu//2Th64jrrxl4AAAAASUVORK5CYII=\n"
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "How many markers are with P<1e-8?\n",
        "TODO: what threshold to use????"
      ],
      "metadata": {
        "id": "Ozpf00nI2yQi"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "hwe_stat %>% mutate(hwe_stat_lessThan1e_8 = if_else(P < 1e-8, TRUE, FALSE)) %>%\n",
        "                 group_by(hwe_stat_lessThan1e_8) %>%\n",
        "                 tally()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "IFxI-xKW21bK",
        "outputId": "385deeec-b100-4095-dc65-56e9cfd60bf5"
      },
      "execution_count": 54,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# A tibble: 2  2\n",
            "  hwe_stat_lessThan1e_8       n\n",
            "  <lgl>                   <int>\n",
            "1 FALSE                 1708443\n",
            "2 TRUE                      726\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "LD pruning"
      ],
      "metadata": {
        "id": "S_7gXrns242b"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "plink --bfile ${analysis_dir}/penncath.geno0.01.mind0.01 --indep-pairwise 50 5 0.3 --out ${analysis_dir}/penncath.geno0.01.mind0.01"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Zcky7pTE6MZs",
        "outputId": "8c69f89a-f2d2-471d-c701-6d19b2ea1f7a"
      },
      "execution_count": 55,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "Logging to /content/analysis//penncath.geno0.01.mind0.01.log.\n",
            "Options in effect:\n",
            "  --bfile /content/analysis//penncath.geno0.01.mind0.01\n",
            "  --indep-pairwise 50 5 0.3\n",
            "  --out /content/analysis//penncath.geno0.01.mind0.01\n",
            "\n",
            "12982 MB RAM detected; reserving 6491 MB for main workspace.\n",
            "569723 variants loaded from .bim file.\n",
            "1380 people (922 males, 458 females) loaded from .fam.\n",
            "918 phenotype values loaded from .fam.\n",
            "Using 1 thread (no multithreaded calculations invoked).\n",
            "Before main variant filters, 1380 founders and 0 nonfounders present.\n",
            "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
            "Total genotyping rate is 0.997808.\n",
            "569723 variants and 1380 people pass filters and QC.\n",
            "Among remaining phenotypes, 455 are cases and 463 are controls.  (462\n",
            "phenotypes are missing.)\n",
            "\r1%\r2%\r3%\r4%\r5%\r6%\r7%\r7%\r8%\r8%\r9%\r10%\r11%\r12%\r13%\r14%\r15%\r16%\r17%\r17%\r18%\r18%\r19%\r20%\r21%\r22%\r23%\r24%\r25%\r26%\r27%\r27%\r28%\r28%\r29%\r30%\r31%\r32%\r33%\r34%\r35%\r36%\r37%\r37%\r38%\r38%\r39%\r40%\r41%\r42%\r43%\r44%\r45%\r46%\r47%\r47%\r48%\r48%\r49%\r50%\r51%\r52%\r53%\r54%\r55%\r56%\r56%\r57%\r57%\r58%\r59%\r60%\r61%\r62%\r63%\r64%\r65%\r66%\r66%\r67%\r67%\r68%\r69%\r70%\r71%\r72%\r73%\r74%\r75%\r76%\r76%\r77%\r77%\r78%\r79%\r80%\r81%\r82%\r83%\r84%\r85%\r86%\r86%\r87%\r87%\r88%\r89%\r90%\r91%\r92%\r93%\r94%\r95%\r96%\r96%\r97%\r97%\r98%\r99%\rPruned 32605 variants from chromosome 1, leaving 14344.\n",
            "\r1%\r1%\r2%\r3%\r4%\r5%\r6%\r6%\r7%\r8%\r9%\r10%\r11%\r11%\r12%\r13%\r14%\r15%\r16%\r16%\r17%\r18%\r19%\r20%\r21%\r22%\r23%\r23%\r24%\r25%\r26%\r27%\r28%\r28%\r29%\r30%\r31%\r32%\r33%\r33%\r34%\r35%\r36%\r37%\r38%\r38%\r39%\r40%\r40%\r41%\r42%\r43%\r44%\r45%\r45%\r46%\r47%\r48%\r49%\r50%\r50%\r51%\r52%\r53%\r54%\r55%\r55%\r56%\r57%\r58%\r59%\r60%\r61%\r62%\r62%\r63%\r64%\r65%\r66%\r67%\r67%\r68%\r69%\r70%\r71%\r72%\r72%\r73%\r74%\r75%\r76%\r77%\r77%\r78%\r79%\r80%\r81%\r82%\r83%\r84%\r84%\r85%\r86%\r87%\r88%\r89%\r89%\r90%\r91%\r92%\r93%\r94%\r94%\r95%\r96%\r97%\r98%\r99%\rPruned 35010 variants from chromosome 2, leaving 14285.\n",
            "\r1%\r2%\r3%\r4%\r5%\r6%\r6%\r7%\r8%\r9%\r9%\r10%\r11%\r12%\r12%\r13%\r14%\r15%\r16%\r17%\r18%\r19%\r19%\r20%\r21%\r22%\r22%\r23%\r24%\r25%\r25%\r26%\r27%\r28%\r29%\r30%\r31%\r32%\r32%\r33%\r34%\r35%\r35%\r36%\r37%\r38%\r38%\r39%\r40%\r41%\r42%\r43%\r44%\r45%\r45%\r46%\r47%\r48%\r48%\r49%\r50%\r51%\r51%\r52%\r53%\r54%\r55%\r56%\r57%\r58%\r58%\r59%\r60%\r61%\r61%\r62%\r63%\r64%\r64%\r65%\r66%\r67%\r68%\r69%\r70%\r71%\r71%\r72%\r73%\r74%\r74%\r75%\r76%\r77%\r77%\r78%\r79%\r80%\r81%\r82%\r83%\r84%\r84%\r85%\r86%\r87%\r87%\r88%\r89%\r90%\r90%\r91%\r92%\r93%\r94%\r95%\r96%\r97%\r97%\r98%\r99%\rPruned 28095 variants from chromosome 3, leaving 12059.\n",
            "\r0%\r1%\r1%\r2%\r2%\r3%\r3%\r4%\r4%\r5%\r5%\r6%\r6%\r7%\r7%\r8%\r8%\r9%\r9%\r10%\r10%\r11%\r11%\r12%\r12%\r13%\r13%\r14%\r14%\r15%\r15%\r16%\r16%\r17%\r17%\r18%\r18%\r19%\r20%\r21%\r22%\r23%\r24%\r25%\r26%\r27%\r28%\r29%\r30%\r31%\r32%\r33%\r34%\r35%\r36%\r37%\r38%\r39%\r40%\r41%\r42%\r43%\r44%\r45%\r46%\r47%\r48%\r49%\r50%\r51%\r52%\r53%\r54%\r55%\r56%\r57%\r58%\r59%\r60%\r61%\r62%\r63%\r64%\r65%\r66%\r67%\r68%\r69%\r70%\r71%\r72%\r73%\r74%\r75%\r76%\r77%\r78%\r79%\r80%\r81%\r82%\r83%\r84%\r85%\r86%\r87%\r88%\r89%\r90%\r91%\r92%\r93%\r94%\r95%\r96%\r97%\r98%\r99%\rPruned 25434 variants from chromosome 4, leaving 11071.\n",
            "\r1%\r2%\r2%\r3%\r4%\r5%\r5%\r6%\r7%\r8%\r8%\r9%\r10%\r11%\r11%\r12%\r13%\r14%\r14%\r15%\r16%\r17%\r17%\r18%\r19%\r20%\r20%\r21%\r22%\r23%\r24%\r25%\r26%\r27%\r28%\r29%\r30%\r31%\r32%\r33%\r34%\r35%\r36%\r37%\r38%\r39%\r40%\r40%\r41%\r42%\r43%\r43%\r44%\r45%\r46%\r46%\r47%\r48%\r49%\r49%\r50%\r51%\r52%\r52%\r53%\r54%\r55%\r55%\r56%\r57%\r58%\r58%\r59%\r60%\r61%\r62%\r63%\r64%\r65%\r66%\r67%\r68%\r69%\r70%\r71%\r72%\r73%\r74%\r75%\r76%\r77%\r78%\r78%\r79%\r80%\r81%\r81%\r82%\r83%\r84%\r84%\r85%\r86%\r87%\r87%\r88%\r89%\r90%\r90%\r91%\r92%\r93%\r93%\r94%\r95%\r96%\r96%\r97%\r98%\r99%\rPruned 26021 variants from chromosome 5, leaving 11150.\n",
            "\r1%\r2%\r3%\r4%\r5%\r6%\r6%\r7%\r8%\r9%\r10%\r11%\r12%\r13%\r14%\r15%\r15%\r16%\r17%\r18%\r19%\r20%\r21%\r22%\r23%\r24%\r24%\r25%\r26%\r27%\r28%\r29%\r30%\r31%\r32%\r33%\r33%\r34%\r35%\r36%\r37%\r38%\r39%\r40%\r41%\r42%\r42%\r43%\r44%\r45%\r46%\r47%\r48%\r49%\r50%\r51%\r51%\r52%\r53%\r54%\r55%\r56%\r57%\r58%\r58%\r59%\r60%\r60%\r61%\r62%\r63%\r64%\r65%\r66%\r67%\r67%\r68%\r69%\r69%\r70%\r71%\r72%\r73%\r74%\r75%\r76%\r76%\r77%\r78%\r78%\r79%\r80%\r81%\r82%\r83%\r84%\r85%\r85%\r86%\r87%\r87%\r88%\r89%\r90%\r91%\r92%\r93%\r94%\r94%\r95%\r96%\r96%\r97%\r98%\r99%\rPruned 24912 variants from chromosome 6, leaving 10810.\n",
            "\r1%\r2%\r3%\r4%\r5%\r6%\r7%\r8%\r8%\r9%\r10%\r11%\r12%\r12%\r13%\r14%\r15%\r16%\r16%\r17%\r18%\r19%\r20%\r20%\r21%\r22%\r23%\r24%\r25%\r26%\r27%\r28%\r29%\r30%\r31%\r32%\r33%\r33%\r34%\r35%\r36%\r37%\r37%\r38%\r39%\r40%\r41%\r41%\r42%\r43%\r44%\r45%\r45%\r46%\r47%\r48%\r49%\r50%\r51%\r52%\r53%\r54%\r55%\r56%\r57%\r58%\r58%\r59%\r60%\r61%\r62%\r62%\r63%\r64%\r65%\r66%\r66%\r67%\r68%\r69%\r70%\r70%\r71%\r72%\r73%\r74%\r75%\r76%\r77%\r78%\r79%\r80%\r81%\r82%\r83%\r83%\r84%\r85%\r86%\r87%\r87%\r88%\r89%\r90%\r91%\r91%\r92%\r93%\r94%\r95%\r95%\r96%\r97%\r98%\r99%\rPruned 21109 variants from chromosome 7, leaving 9511.\n",
            "\r1%\r2%\r3%\r4%\r5%\r6%\r6%\r7%\r8%\r9%\r10%\r10%\r11%\r12%\r13%\r14%\r14%\r15%\r16%\r17%\r18%\r18%\r19%\r20%\r21%\r22%\r22%\r23%\r24%\r25%\r26%\r27%\r28%\r29%\r30%\r31%\r32%\r33%\r33%\r34%\r35%\r36%\r37%\r37%\r38%\r39%\r40%\r41%\r41%\r42%\r43%\r44%\r45%\r45%\r46%\r47%\r48%\r49%\r50%\r51%\r52%\r53%\r54%\r55%\r56%\r56%\r57%\r58%\r59%\r60%\r60%\r61%\r62%\r63%\r64%\r64%\r65%\r66%\r67%\r68%\r68%\r69%\r70%\r71%\r72%\r72%\r73%\r74%\r75%\r76%\r77%\r78%\r79%\r80%\r81%\r82%\r83%\r83%\r84%\r85%\r86%\r87%\r87%\r88%\r89%\r90%\r91%\r91%\r92%\r93%\r94%\r95%\r95%\r96%\r97%\r98%\r99%\rPruned 23093 variants from chromosome 8, leaving 9277.\n",
            "\r1%\r2%\r3%\r4%\r5%\r6%\r7%\r8%\r9%\r10%\r11%\r12%\r13%\r14%\r15%\r16%\r17%\r18%\r19%\r20%\r21%\r22%\r23%\r24%\r25%\r26%\r27%\r28%\r29%\r30%\r31%\r32%\r33%\r34%\r35%\r36%\r37%\r38%\r39%\r40%\r41%\r42%\r43%\r44%\r45%\r46%\r47%\r48%\r49%\r50%\r51%\r52%\r52%\r53%\r54%\r55%\r55%\r56%\r57%\r58%\r58%\r59%\r60%\r61%\r61%\r62%\r63%\r64%\r64%\r65%\r66%\r67%\r67%\r68%\r69%\r70%\r70%\r71%\r72%\r73%\r73%\r74%\r75%\r76%\r76%\r77%\r78%\r79%\r79%\r80%\r81%\r82%\r82%\r83%\r84%\r85%\r85%\r86%\r87%\r88%\r88%\r89%\r90%\r91%\r91%\r92%\r93%\r94%\r94%\r95%\r96%\r97%\r97%\r98%\r99%\rPruned 18654 variants from chromosome 9, leaving 8178.\n",
            "\r0%\r1%\r1%\r2%\r2%\r3%\r3%\r4%\r4%\r5%\r5%\r6%\r6%\r7%\r7%\r8%\r8%\r9%\r9%\r10%\r10%\r11%\r11%\r12%\r12%\r13%\r13%\r14%\r14%\r15%\r15%\r16%\r16%\r17%\r17%\r18%\r18%\r19%\r20%\r21%\r22%\r23%\r24%\r25%\r26%\r27%\r28%\r29%\r30%\r31%\r32%\r33%\r34%\r35%\r36%\r37%\r38%\r39%\r40%\r41%\r42%\r43%\r44%\r45%\r46%\r47%\r48%\r49%\r50%\r51%\r52%\r53%\r54%\r55%\r56%\r57%\r58%\r59%\r60%\r61%\r62%\r63%\r64%\r65%\r66%\r67%\r68%\r69%\r70%\r71%\r72%\r73%\r74%\r75%\r76%\r77%\r78%\r79%\r80%\r81%\r82%\r83%\r84%\r85%\r86%\r87%\r88%\r89%\r90%\r91%\r92%\r93%\r94%\r95%\r96%\r97%\r98%\r99%\rPruned 22399 variants from chromosome 10, leaving 9606.\n",
            "\r1%\r2%\r3%\r4%\r5%\r6%\r6%\r7%\r8%\r8%\r9%\r10%\r11%\r12%\r13%\r14%\r15%\r16%\r17%\r17%\r18%\r19%\r19%\r20%\r21%\r22%\r23%\r24%\r25%\r26%\r27%\r28%\r28%\r29%\r30%\r30%\r31%\r32%\r33%\r34%\r35%\r36%\r37%\r38%\r39%\r39%\r40%\r41%\r41%\r42%\r43%\r44%\r45%\r46%\r47%\r48%\r49%\r50%\r50%\r51%\r52%\r52%\r53%\r54%\r55%\r56%\r57%\r58%\r59%\r60%\r61%\r61%\r62%\r63%\r63%\r64%\r65%\r66%\r67%\r68%\r69%\r70%\r71%\r72%\r72%\r73%\r74%\r74%\r75%\r76%\r77%\r78%\r79%\r80%\r81%\r82%\r83%\r83%\r84%\r85%\r85%\r86%\r87%\r88%\r89%\r90%\r91%\r92%\r93%\r94%\r94%\r95%\r96%\r96%\r97%\r98%\r99%\rPruned 20614 variants from chromosome 11, leaving 8613.\n",
            "\r1%\r1%\r2%\r3%\r3%\r4%\r5%\r5%\r6%\r7%\r7%\r8%\r9%\r9%\r10%\r11%\r11%\r12%\r13%\r13%\r14%\r15%\r15%\r16%\r17%\r17%\r18%\r19%\r19%\r20%\r21%\r21%\r22%\r23%\r23%\r24%\r25%\r26%\r27%\r28%\r29%\r30%\r31%\r32%\r33%\r34%\r35%\r36%\r37%\r38%\r39%\r40%\r41%\r42%\r43%\r44%\r45%\r46%\r47%\r48%\r49%\r50%\r51%\r52%\r53%\r54%\r55%\r56%\r57%\r58%\r59%\r60%\r61%\r62%\r62%\r63%\r64%\r64%\r65%\r66%\r66%\r67%\r68%\r68%\r69%\r70%\r70%\r71%\r72%\r72%\r73%\r74%\r74%\r75%\r76%\r76%\r77%\r78%\r78%\r79%\r80%\r80%\r81%\r82%\r82%\r83%\r84%\r84%\r85%\r86%\r86%\r87%\r88%\r89%\r90%\r91%\r92%\r93%\r94%\r95%\r96%\r97%\r98%\r99%\rPruned 19014 variants from chromosome 12, leaving 8740.\n",
            "\r1%\r2%\r3%\r4%\r5%\r6%\r6%\r7%\r8%\r9%\r10%\r11%\r12%\r13%\r13%\r14%\r14%\r15%\r16%\r17%\r18%\r19%\r20%\r21%\r21%\r22%\r23%\r24%\r25%\r26%\r27%\r28%\r28%\r29%\r29%\r30%\r31%\r32%\r33%\r34%\r35%\r36%\r36%\r37%\r38%\r39%\r40%\r41%\r42%\r43%\r43%\r44%\r44%\r45%\r46%\r47%\r48%\r49%\r50%\r51%\r51%\r52%\r52%\r53%\r54%\r55%\r56%\r57%\r58%\r59%\r59%\r60%\r61%\r62%\r63%\r64%\r65%\r66%\r66%\r67%\r67%\r68%\r69%\r70%\r71%\r72%\r73%\r74%\r74%\r75%\r76%\r77%\r78%\r79%\r80%\r81%\r81%\r82%\r82%\r83%\r84%\r85%\r86%\r87%\r88%\r89%\r89%\r90%\r91%\r92%\r93%\r94%\r95%\r96%\r96%\r97%\r97%\r98%\r99%\rPruned 15428 variants from chromosome 13, leaving 7006.\n",
            "\r1%\r2%\r2%\r3%\r4%\r5%\r6%\r7%\r7%\r8%\r9%\r10%\r11%\r12%\r13%\r14%\r15%\r15%\r16%\r17%\r18%\r19%\r20%\r20%\r21%\r22%\r23%\r23%\r24%\r25%\r26%\r27%\r28%\r28%\r29%\r30%\r31%\r32%\r33%\r34%\r35%\r36%\r36%\r37%\r38%\r39%\r40%\r41%\r41%\r42%\r43%\r44%\r44%\r45%\r46%\r47%\r48%\r49%\r49%\r50%\r51%\r52%\r53%\r54%\r54%\r55%\r56%\r57%\r57%\r58%\r59%\r60%\r61%\r62%\r62%\r63%\r64%\r65%\r66%\r67%\r68%\r69%\r70%\r70%\r71%\r72%\r73%\r74%\r75%\r75%\r76%\r77%\r78%\r78%\r79%\r80%\r81%\r82%\r83%\r83%\r84%\r85%\r86%\r87%\r88%\r89%\r90%\r91%\r91%\r92%\r93%\r94%\r95%\r96%\r96%\r97%\r98%\r99%\rPruned 12725 variants from chromosome 14, leaving 5966.\n",
            "\r1%\r1%\r2%\r3%\r3%\r4%\r5%\r6%\r7%\r8%\r9%\r10%\r11%\r12%\r12%\r13%\r14%\r14%\r15%\r16%\r17%\r18%\r19%\r20%\r21%\r22%\r23%\r23%\r24%\r25%\r25%\r26%\r27%\r28%\r29%\r30%\r31%\r32%\r33%\r34%\r34%\r35%\r36%\r36%\r37%\r38%\r39%\r40%\r41%\r42%\r43%\r44%\r45%\r45%\r46%\r47%\r47%\r48%\r49%\r50%\r51%\r52%\r53%\r54%\r55%\r56%\r56%\r57%\r58%\r58%\r59%\r60%\r60%\r61%\r62%\r63%\r64%\r65%\r66%\r67%\r68%\r69%\r69%\r70%\r71%\r71%\r72%\r73%\r74%\r75%\r76%\r77%\r78%\r79%\r80%\r80%\r81%\r82%\r82%\r83%\r84%\r85%\r86%\r87%\r88%\r89%\r90%\r91%\r91%\r92%\r93%\r93%\r94%\r95%\r96%\r97%\r98%\r99%\rPruned 11658 variants from chromosome 15, leaving 5614.\n",
            "\r0%\r1%\r1%\r2%\r2%\r3%\r3%\r4%\r4%\r5%\r5%\r6%\r6%\r7%\r7%\r8%\r8%\r9%\r9%\r10%\r10%\r11%\r12%\r13%\r14%\r15%\r16%\r17%\r18%\r19%\r20%\r21%\r22%\r23%\r24%\r25%\r26%\r27%\r28%\r29%\r30%\r31%\r32%\r33%\r34%\r35%\r36%\r37%\r38%\r39%\r40%\r41%\r42%\r43%\r44%\r45%\r46%\r47%\r48%\r49%\r50%\r51%\r52%\r53%\r54%\r55%\r55%\r56%\r56%\r57%\r57%\r58%\r58%\r59%\r59%\r60%\r60%\r61%\r61%\r62%\r62%\r63%\r63%\r64%\r64%\r65%\r65%\r66%\r67%\r68%\r69%\r70%\r71%\r72%\r73%\r74%\r75%\r76%\r77%\r78%\r79%\r80%\r81%\r82%\r83%\r84%\r85%\r86%\r87%\r88%\r89%\r90%\r91%\r92%\r93%\r94%\r95%\r96%\r97%\r98%\r99%\rPruned 12467 variants from chromosome 16, leaving 6042.\n",
            "\r1%\r2%\r3%\r4%\r5%\r6%\r6%\r7%\r7%\r8%\r9%\r10%\r11%\r12%\r13%\r14%\r15%\r15%\r16%\r16%\r17%\r18%\r19%\r20%\r21%\r22%\r23%\r24%\r24%\r25%\r26%\r27%\r28%\r29%\r30%\r31%\r32%\r32%\r33%\r33%\r34%\r35%\r36%\r37%\r38%\r39%\r40%\r41%\r41%\r42%\r42%\r43%\r44%\r45%\r46%\r47%\r48%\r49%\r50%\r50%\r51%\r52%\r53%\r54%\r55%\r56%\r57%\r58%\r58%\r59%\r59%\r60%\r61%\r62%\r63%\r64%\r65%\r66%\r67%\r67%\r68%\r69%\r70%\r71%\r72%\r73%\r74%\r75%\r75%\r76%\r76%\r77%\r78%\r79%\r80%\r81%\r82%\r83%\r84%\r84%\r85%\r85%\r86%\r87%\r88%\r89%\r90%\r91%\r92%\r93%\r93%\r94%\r95%\r96%\r97%\r98%\r99%\rPruned 8123 variants from chromosome 17, leaving 4819.\n",
            "\r1%\r1%\r2%\r3%\r4%\r5%\r6%\r6%\r7%\r8%\r9%\r10%\r11%\r11%\r12%\r13%\r14%\r15%\r16%\r16%\r17%\r18%\r19%\r20%\r21%\r22%\r23%\r23%\r24%\r25%\r26%\r27%\r28%\r28%\r29%\r30%\r31%\r32%\r33%\r33%\r34%\r35%\r36%\r37%\r38%\r38%\r39%\r40%\r40%\r41%\r42%\r43%\r44%\r45%\r45%\r46%\r47%\r48%\r49%\r50%\r50%\r51%\r52%\r53%\r54%\r55%\r55%\r56%\r57%\r58%\r59%\r60%\r61%\r62%\r62%\r63%\r64%\r65%\r66%\r67%\r67%\r68%\r69%\r70%\r71%\r72%\r72%\r73%\r74%\r75%\r76%\r77%\r77%\r78%\r79%\r80%\r81%\r82%\r83%\r84%\r84%\r85%\r86%\r87%\r88%\r89%\r89%\r90%\r91%\r92%\r93%\r94%\r94%\r95%\r96%\r97%\r98%\r99%\rPruned 11646 variants from chromosome 18, leaving 5649.\n",
            "\r1%\r2%\r2%\r3%\r4%\r5%\r6%\r7%\r7%\r8%\r9%\r10%\r11%\r12%\r12%\r13%\r14%\r15%\r16%\r17%\r17%\r18%\r19%\r20%\r20%\r21%\r22%\r23%\r24%\r25%\r25%\r26%\r27%\r28%\r29%\r30%\r30%\r31%\r32%\r33%\r34%\r35%\r35%\r36%\r37%\r38%\r39%\r40%\r41%\r42%\r43%\r43%\r44%\r45%\r46%\r47%\r48%\r48%\r49%\r50%\r51%\r52%\r53%\r53%\r54%\r55%\r56%\r57%\r58%\r58%\r59%\r60%\r61%\r61%\r62%\r63%\r64%\r65%\r66%\r66%\r67%\r68%\r69%\r70%\r71%\r71%\r72%\r73%\r74%\r75%\r76%\r76%\r77%\r78%\r79%\r80%\r81%\r82%\r83%\r84%\r84%\r85%\r86%\r87%\r88%\r89%\r89%\r90%\r91%\r92%\r93%\r94%\r94%\r95%\r96%\r97%\r98%\r99%\rPruned 4199 variants from chromosome 19, leaving 2996.\n",
            "\r1%\r2%\r3%\r3%\r4%\r5%\r6%\r7%\r7%\r8%\r9%\r10%\r11%\r12%\r12%\r13%\r14%\r15%\r16%\r16%\r17%\r18%\r19%\r20%\r21%\r22%\r23%\r24%\r25%\r25%\r26%\r27%\r28%\r29%\r29%\r30%\r31%\r32%\r33%\r34%\r34%\r35%\r36%\r37%\r38%\r38%\r39%\r40%\r41%\r42%\r42%\r43%\r44%\r45%\r46%\r47%\r47%\r48%\r49%\r50%\r51%\r51%\r52%\r53%\r54%\r55%\r56%\r56%\r57%\r58%\r59%\r60%\r60%\r61%\r62%\r63%\r64%\r64%\r65%\r66%\r67%\r68%\r69%\r69%\r70%\r71%\r72%\r73%\r73%\r74%\r75%\r76%\r77%\r78%\r79%\r80%\r81%\r82%\r82%\r83%\r84%\r85%\r86%\r86%\r87%\r88%\r89%\r90%\r91%\r91%\r92%\r93%\r94%\r95%\r95%\r96%\r97%\r98%\r99%\rPruned 10345 variants from chromosome 20, leaving 5041.\n",
            "\r0%\r1%\r2%\r3%\r4%\r5%\r6%\r6%\r7%\r8%\r9%\r10%\r11%\r12%\r12%\r13%\r13%\r14%\r15%\r16%\r17%\r18%\r19%\r19%\r20%\r21%\r22%\r23%\r24%\r25%\r25%\r26%\r26%\r27%\r28%\r29%\r30%\r31%\r32%\r32%\r33%\r34%\r35%\r36%\r37%\r38%\r38%\r39%\r39%\r40%\r41%\r42%\r43%\r44%\r45%\r45%\r46%\r47%\r48%\r49%\r50%\r51%\r51%\r52%\r52%\r53%\r54%\r55%\r56%\r57%\r58%\r58%\r59%\r60%\r61%\r62%\r63%\r64%\r64%\r65%\r65%\r66%\r67%\r68%\r69%\r70%\r71%\r71%\r72%\r73%\r74%\r75%\r76%\r77%\r77%\r78%\r78%\r79%\r80%\r81%\r82%\r83%\r84%\r84%\r85%\r86%\r87%\r88%\r89%\r90%\r90%\r91%\r91%\r92%\r93%\r94%\r95%\r96%\r97%\r97%\r98%\r99%\rPruned 5378 variants from chromosome 21, leaving 2699.\n",
            "\r1%\r2%\r3%\r4%\r4%\r5%\r6%\r7%\r7%\r8%\r9%\r10%\r11%\r12%\r13%\r14%\r15%\r15%\r16%\r17%\r18%\r18%\r19%\r20%\r21%\r22%\r23%\r24%\r25%\r26%\r26%\r27%\r28%\r29%\r29%\r30%\r31%\r32%\r33%\r34%\r35%\r36%\r37%\r37%\r38%\r39%\r40%\r40%\r41%\r42%\r43%\r44%\r45%\r46%\r47%\r48%\r48%\r49%\r50%\r51%\r51%\r52%\r53%\r54%\r55%\r56%\r57%\r58%\r59%\r59%\r60%\r61%\r62%\r62%\r63%\r64%\r65%\r66%\r67%\r68%\r69%\r70%\r70%\r71%\r72%\r73%\r73%\r74%\r75%\r76%\r77%\r78%\r79%\r80%\r81%\r81%\r82%\r83%\r84%\r84%\r85%\r86%\r87%\r88%\r89%\r90%\r91%\r92%\r92%\r93%\r94%\r95%\r95%\r96%\r97%\r98%\r99%\rPruned 4588 variants from chromosome 22, leaving 2730.\n",
            "Pruning complete.  393517 of 569723 variants removed.\n",
            "Writing...\rMarker lists written to /content/analysis//penncath.geno0.01.mind0.01.prune.in\n",
            "and /content/analysis//penncath.geno0.01.mind0.01.prune.out .\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Relatedness check"
      ],
      "metadata": {
        "id": "mWaMo9ZL6cpz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "plink --bfile ${analysis_dir}/penncath.geno0.01.mind0.01 \\\n",
        "      --extract ${analysis_dir}/penncath.geno0.01.mind0.01.prune.in \\\n",
        "      --genome --min 0.05 --out ${analysis_dir}/penncath.geno0.01.mind0.01.LD0.3"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tU3ww9-56ffC",
        "outputId": "9cd96b45-3f04-4e32-90f1-c4ec6f3af7b5"
      },
      "execution_count": 56,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "Logging to /content/analysis//penncath.geno0.01.mind0.01.LD0.3.log.\n",
            "Options in effect:\n",
            "  --bfile /content/analysis///penncath.geno0.01.mind0.01\n",
            "  --extract /content/analysis//penncath.geno0.01.mind0.01.prune.in\n",
            "  --genome\n",
            "  --min 0.05\n",
            "  --out /content/analysis//penncath.geno0.01.mind0.01.LD0.3\n",
            "\n",
            "12982 MB RAM detected; reserving 6491 MB for main workspace.\n",
            "569723 variants loaded from .bim file.\n",
            "1380 people (922 males, 458 females) loaded from .fam.\n",
            "918 phenotype values loaded from .fam.\n",
            "--extract: 176206 variants remaining.\n",
            "Using up to 2 threads (change this with --threads).\n",
            "Before main variant filters, 1380 founders and 0 nonfounders present.\n",
            "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
            "Total genotyping rate is 0.997683.\n",
            "176206 variants and 1380 people pass filters and QC.\n",
            "Among remaining phenotypes, 455 are cases and 463 are controls.  (462\n",
            "phenotypes are missing.)\n",
            "\r1152 markers complete.\r2304 markers complete.\r3456 markers complete.\r4608 markers complete.\r5760 markers complete.\r6912 markers complete.\r8064 markers complete.\r9216 markers complete.\r10368 markers complete.\r11520 markers complete.\r12672 markers complete.\r13824 markers complete.\r14976 markers complete.\r16128 markers complete.\r17280 markers complete.\r18432 markers complete.\r19584 markers complete.\r20736 markers complete.\r21888 markers complete.\r23040 markers complete.\r24192 markers complete.\r25344 markers complete.\r26496 markers complete.\r27648 markers complete.\r28800 markers complete.\r29952 markers complete.\r31104 markers complete.\r32256 markers complete.\r33408 markers complete.\r34560 markers complete.\r35712 markers complete.\r36864 markers complete.\r38016 markers complete.\r39168 markers complete.\r40320 markers complete.\r41472 markers complete.\r42624 markers complete.\r43776 markers complete.\r44928 markers complete.\r46080 markers complete.\r47232 markers complete.\r48384 markers complete.\r49536 markers complete.\r50688 markers complete.\r51840 markers complete.\r52992 markers complete.\r54144 markers complete.\r55296 markers complete.\r56448 markers complete.\r57600 markers complete.\r58752 markers complete.\r59904 markers complete.\r61056 markers complete.\r62208 markers complete.\r63360 markers complete.\r64512 markers complete.\r65664 markers complete.\r66816 markers complete.\r67968 markers complete.\r69120 markers complete.\r70272 markers complete.\r71424 markers complete.\r72576 markers complete.\r73728 markers complete.\r74880 markers complete.\r76032 markers complete.\r77184 markers complete.\r78336 markers complete.\r79488 markers complete.\r80640 markers complete.\r81792 markers complete.\r82944 markers complete.\r84096 markers complete.\r85248 markers complete.\r86400 markers complete.\r87552 markers complete.\r88704 markers complete.\r89856 markers complete.\r91008 markers complete.\r92160 markers complete.\r93312 markers complete.\r94464 markers complete.\r95616 markers complete.\r96768 markers complete.\r97920 markers complete.\r99072 markers complete.\r100224 markers complete.\r101376 markers complete.\r102528 markers complete.\r103680 markers complete.\r104832 markers complete.\r105984 markers complete.\r107136 markers complete.\r108288 markers complete.\r109440 markers complete.\r110592 markers complete.\r111744 markers complete.\r112896 markers complete.\r114048 markers complete.\r115200 markers complete.\r116352 markers complete.\r117504 markers complete.\r118656 markers complete.\r119808 markers complete.\r120960 markers complete.\r122112 markers complete.\r123264 markers complete.\r124416 markers complete.\r125568 markers complete.\r126720 markers complete.\r127872 markers complete.\r129024 markers complete.\r130176 markers complete.\r131328 markers complete.\r132480 markers complete.\r133632 markers complete.\r134784 markers complete.\r135936 markers complete.\r137088 markers complete.\r138240 markers complete.\r139392 markers complete.\r140544 markers complete.\r141696 markers complete.\r142848 markers complete.\r144000 markers complete.\r145152 markers complete.\r146304 markers complete.\r147456 markers complete.\r148608 markers complete.\r149760 markers complete.\r150912 markers complete.\r152064 markers complete.\r153216 markers complete.\r154368 markers complete.\r155520 markers complete.\r156672 markers complete.\r157824 markers complete.\r158976 markers complete.\r160128 markers complete.\r161280 markers complete.\r162432 markers complete.\r163584 markers complete.\r164736 markers complete.\r165888 markers complete.\r167040 markers complete.\r168192 markers complete.\r169344 markers complete.\r170496 markers complete.\r171648 markers complete.\r172800 markers complete.\r173952 markers complete.\r175104 markers complete.\r176206 markers complete.\rIBD calculations complete.  \n",
            "\rWriting... 99%\rFinished writing /content/analysis//penncath.geno0.01.mind0.01.LD0.3.genome .\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Check pairs of samples with PI_HAT>0.05, maximum PI_HAT is 0.1745, no duplicates nor first degree relatives\n",
        "TODO: do we want to remove any samples?"
      ],
      "metadata": {
        "id": "nWFd5vEE6tdG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "ibd <- fread(glue(analysis_dir, \"penncath.geno0.01.mind0.01.LD0.3.genome\"))\n",
        "summary(ibd$PI_HAT)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rJGO1hal6xG-",
        "outputId": "4d700169-3f0c-4573-f93f-911405469e8a"
      },
      "execution_count": 57,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
            "0.05000 0.05225 0.05395 0.06848 0.05962 0.17450 \n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "run Heterozygocity"
      ],
      "metadata": {
        "id": "9wKt7vcb60tu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "plink --bfile ${analysis_dir}/penncath.geno0.01.mind0.01 --het \\\n",
        "      --extract ${analysis_dir}/penncath.geno0.01.mind0.01.prune.in \\\n",
        "      --out ${analysis_dir}/penncath.geno0.01.mind0.01.LD0.3"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hy1tL0HA64lp",
        "outputId": "b684bbb3-e5bc-4699-9dfd-b9019def6c2b"
      },
      "execution_count": 59,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "Logging to /content/analysis//penncath.geno0.01.mind0.01.LD0.3.log.\n",
            "Options in effect:\n",
            "  --bfile /content/analysis//penncath.geno0.01.mind0.01\n",
            "  --extract /content/analysis//penncath.geno0.01.mind0.01.prune.in\n",
            "  --het\n",
            "  --out /content/analysis//penncath.geno0.01.mind0.01.LD0.3\n",
            "\n",
            "12982 MB RAM detected; reserving 6491 MB for main workspace.\n",
            "569723 variants loaded from .bim file.\n",
            "1380 people (922 males, 458 females) loaded from .fam.\n",
            "918 phenotype values loaded from .fam.\n",
            "--extract: 176206 variants remaining.\n",
            "Using 1 thread (no multithreaded calculations invoked).\n",
            "Before main variant filters, 1380 founders and 0 nonfounders present.\n",
            "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
            "Total genotyping rate is 0.997683.\n",
            "176206 variants and 1380 people pass filters and QC.\n",
            "Among remaining phenotypes, 455 are cases and 463 are controls.  (462\n",
            "phenotypes are missing.)\n",
            "--het: 176206 variants scanned, report written to\n",
            "/content/analysis//penncath.geno0.01.mind0.01.LD0.3.het .\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "look at the heterozygocity of individuals, find individuals with heterozygocity +- 3 s.d.\n",
        "TODO: remove any one?"
      ],
      "metadata": {
        "id": "dzIm53Y17NlM"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "heterozygocity <- fread(glue(analysis_dir, \"penncath.geno0.01.mind0.01.LD0.3.het\"))\n",
        "\n",
        "heterozygocity %>% mutate (het_rate = (`N(NM)` - `O(HOM)`) / `N(NM)` ) %>%\n",
        "                   ggplot( aes(x = het_rate)) +\n",
        "                   geom_histogram() +\n",
        "                   labs(title = \"Heterozygocity\") +\n",
        "                   theme_classic()\n",
        "\n",
        "heterozygocity %>% mutate (het_rate = (`N(NM)` - `O(HOM)`) / `N(NM)` ) %>%\n",
        "                   filter (het_rate < mean(het_rate)-3*sd(het_rate) |\n",
        "                          het_rate > mean(het_rate)+3*sd(het_rate))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "GExedNVh7On3",
        "outputId": "2afb1ea4-d2c5-4878-9eaf-5f205a17304f"
      },
      "execution_count": 60,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "      FID IID O(HOM) E(HOM)  N(NM)        F  het_rate\n",
            " 1: 10063   1 135070 136300 175538 -0.03105 0.2305370\n",
            " 2: 10312   1 134247 135600 174626 -0.03475 0.2312313\n",
            " 3: 10386   1 133933 135400 174490 -0.03681 0.2324317\n",
            " 4: 10420   1 133492 135300 174257 -0.04522 0.2339361\n",
            " 5: 10675   1 133848 135300 174286 -0.03779 0.2320209\n",
            " 6: 10986   1 137896 136400 175638  0.03909 0.2148852\n",
            " 7: 11232   1 135362 136600 175908 -0.03114 0.2304955\n",
            " 8: 11310   1 134242 135500 174565 -0.03295 0.2309913\n",
            " 9: 11315   1 135452 136700 176066 -0.03182 0.2306749\n",
            "10: 11455   1 138068 136600 175997  0.03639 0.2155094\n",
            "11: 11594   1 133948 135300 174422 -0.03472 0.2320464\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "use plink for simple pca"
      ],
      "metadata": {
        "id": "A3pobi5v7WZj"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "plink --bfile ${analysis_dir}/penncath.geno0.01.mind0.01  \\\n",
        "      --extract ${analysis_dir}/penncath.geno0.01.mind0.01.prune.in \\\n",
        "      --maf 0.05 --pca  --out ${analysis_dir}/penncath.geno0.01.mind0.01.LD0.3"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "T_s2lafm7XUI",
        "outputId": "fa3be585-2202-47ab-c419-ea7202065c57"
      },
      "execution_count": 62,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "Logging to /content/analysis//penncath.geno0.01.mind0.01.LD0.3.log.\n",
            "Options in effect:\n",
            "  --bfile /content/analysis//penncath.geno0.01.mind0.01\n",
            "  --extract /content/analysis//penncath.geno0.01.mind0.01.prune.in\n",
            "  --maf 0.05\n",
            "  --out /content/analysis//penncath.geno0.01.mind0.01.LD0.3\n",
            "  --pca\n",
            "\n",
            "12982 MB RAM detected; reserving 6491 MB for main workspace.\n",
            "569723 variants loaded from .bim file.\n",
            "1380 people (922 males, 458 females) loaded from .fam.\n",
            "918 phenotype values loaded from .fam.\n",
            "--extract: 176206 variants remaining.\n",
            "Using up to 2 threads (change this with --threads).\n",
            "Before main variant filters, 1380 founders and 0 nonfounders present.\n",
            "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
            "Total genotyping rate is 0.997683.\n",
            "63537 variants removed due to minor allele threshold(s)\n",
            "(--maf/--max-maf/--mac/--max-mac).\n",
            "112669 variants and 1380 people pass filters and QC.\n",
            "Among remaining phenotypes, 455 are cases and 463 are controls.  (462\n",
            "phenotypes are missing.)\n",
            "\r60 markers complete.\r120 markers complete.\r180 markers complete.\r240 markers complete.\r300 markers complete.\r360 markers complete.\r420 markers complete.\r480 markers complete.\r540 markers complete.\r600 markers complete.\r660 markers complete.\r720 markers complete.\r780 markers complete.\r840 markers complete.\r900 markers complete.\r960 markers complete.\r1020 markers complete.\r1080 markers complete.\r1140 markers complete.\r1200 markers complete.\r1260 markers complete.\r1320 markers complete.\r1380 markers complete.\r1440 markers complete.\r1500 markers complete.\r1560 markers complete.\r1620 markers complete.\r1680 markers complete.\r1740 markers complete.\r1800 markers complete.\r1860 markers complete.\r1920 markers complete.\r1980 markers complete.\r2040 markers complete.\r2100 markers complete.\r2160 markers complete.\r2220 markers complete.\r2280 markers complete.\r2340 markers complete.\r2400 markers complete.\r2460 markers complete.\r2520 markers complete.\r2580 markers complete.\r2640 markers complete.\r2700 markers complete.\r2760 markers complete.\r2820 markers complete.\r2880 markers complete.\r2940 markers complete.\r3000 markers complete.\r3060 markers complete.\r3120 markers complete.\r3180 markers complete.\r3240 markers complete.\r3300 markers complete.\r3360 markers complete.\r3420 markers complete.\r3480 markers complete.\r3540 markers complete.\r3600 markers complete.\r3660 markers complete.\r3720 markers complete.\r3780 markers complete.\r3840 markers complete.\r3900 markers complete.\r3960 markers complete.\r4020 markers complete.\r4080 markers complete.\r4140 markers complete.\r4200 markers complete.\r4260 markers complete.\r4320 markers complete.\r4380 markers complete.\r4440 markers complete.\r4500 markers complete.\r4560 markers complete.\r4620 markers complete.\r4680 markers complete.\r4740 markers complete.\r4800 markers complete.\r4860 markers complete.\r4920 markers complete.\r4980 markers complete.\r5040 markers complete.\r5100 markers complete.\r5160 markers complete.\r5220 markers complete.\r5280 markers complete.\r5340 markers complete.\r5400 markers complete.\r5460 markers complete.\r5520 markers complete.\r5580 markers complete.\r5640 markers complete.\r5700 markers complete.\r5760 markers complete.\r5820 markers complete.\r5880 markers complete.\r5940 markers complete.\r6000 markers complete.\r6060 markers complete.\r6120 markers complete.\r6180 markers complete.\r6240 markers complete.\r6300 markers complete.\r6360 markers complete.\r6420 markers complete.\r6480 markers complete.\r6540 markers complete.\r6600 markers complete.\r6660 markers complete.\r6720 markers complete.\r6780 markers complete.\r6840 markers complete.\r6900 markers complete.\r6960 markers complete.\r7020 markers complete.\r7080 markers complete.\r7140 markers complete.\r7200 markers complete.\r7260 markers complete.\r7320 markers complete.\r7380 markers complete.\r7440 markers complete.\r7500 markers complete.\r7560 markers complete.\r7620 markers complete.\r7680 markers complete.\r7740 markers complete.\r7800 markers complete.\r7860 markers complete.\r7920 markers complete.\r7980 markers complete.\r8040 markers complete.\r8100 markers complete.\r8160 markers complete.\r8220 markers complete.\r8280 markers complete.\r8340 markers complete.\r8400 markers complete.\r8460 markers complete.\r8520 markers complete.\r8580 markers complete.\r8640 markers complete.\r8700 markers complete.\r8760 markers complete.\r8820 markers complete.\r8880 markers complete.\r8940 markers complete.\r9000 markers complete.\r9060 markers complete.\r9120 markers complete.\r9180 markers complete.\r9240 markers complete.\r9300 markers complete.\r9360 markers complete.\r9420 markers complete.\r9480 markers complete.\r9540 markers complete.\r9600 markers complete.\r9660 markers complete.\r9720 markers complete.\r9780 markers complete.\r9840 markers complete.\r9900 markers complete.\r9960 markers complete.\r10020 markers complete.\r10080 markers complete.\r10140 markers complete.\r10200 markers complete.\r10260 markers complete.\r10320 markers complete.\r10380 markers complete.\r10440 markers complete.\r10500 markers complete.\r10560 markers complete.\r10620 markers complete.\r10680 markers complete.\r10740 markers complete.\r10800 markers complete.\r10860 markers complete.\r10920 markers complete.\r10980 markers complete.\r11040 markers complete.\r11100 markers complete.\r11160 markers complete.\r11220 markers complete.\r11280 markers complete.\r11340 markers complete.\r11400 markers complete.\r11460 markers complete.\r11520 markers complete.\r11580 markers complete.\r11640 markers complete.\r11700 markers complete.\r11760 markers complete.\r11820 markers complete.\r11880 markers complete.\r11940 markers complete.\r12000 markers complete.\r12060 markers complete.\r12120 markers complete.\r12180 markers complete.\r12240 markers complete.\r12300 markers complete.\r12360 markers complete.\r12420 markers complete.\r12480 markers complete.\r12540 markers complete.\r12600 markers complete.\r12660 markers complete.\r12720 markers complete.\r12780 markers complete.\r12840 markers complete.\r12900 markers complete.\r12960 markers complete.\r13020 markers complete.\r13080 markers complete.\r13140 markers complete.\r13200 markers complete.\r13260 markers complete.\r13320 markers complete.\r13380 markers complete.\r13440 markers complete.\r13500 markers complete.\r13560 markers complete.\r13620 markers complete.\r13680 markers complete.\r13740 markers complete.\r13800 markers complete.\r13860 markers complete.\r13920 markers complete.\r13980 markers complete.\r14040 markers complete.\r14100 markers complete.\r14160 markers complete.\r14220 markers complete.\r14280 markers complete.\r14340 markers complete.\r14400 markers complete.\r14460 markers complete.\r14520 markers complete.\r14580 markers complete.\r14640 markers complete.\r14700 markers complete.\r14760 markers complete.\r14820 markers complete.\r14880 markers complete.\r14940 markers complete.\r15000 markers complete.\r15060 markers complete.\r15120 markers complete.\r15180 markers complete.\r15240 markers complete.\r15300 markers complete.\r15360 markers complete.\r15420 markers complete.\r15480 markers complete.\r15540 markers complete.\r15600 markers complete.\r15660 markers complete.\r15720 markers complete.\r15780 markers complete.\r15840 markers complete.\r15900 markers complete.\r15960 markers complete.\r16020 markers complete.\r16080 markers complete.\r16140 markers complete.\r16200 markers complete.\r16260 markers complete.\r16320 markers complete.\r16380 markers complete.\r16440 markers complete.\r16500 markers complete.\r16560 markers complete.\r16620 markers complete.\r16680 markers complete.\r16740 markers complete.\r16800 markers complete.\r16860 markers complete.\r16920 markers complete.\r16980 markers complete.\r17040 markers complete.\r17100 markers complete.\r17160 markers complete.\r17220 markers complete.\r17280 markers complete.\r17340 markers complete.\r17400 markers complete.\r17460 markers complete.\r17520 markers complete.\r17580 markers complete.\r17640 markers complete.\r17700 markers complete.\r17760 markers complete.\r17820 markers complete.\r17880 markers complete.\r17940 markers complete.\r18000 markers complete.\r18060 markers complete.\r18120 markers complete.\r18180 markers complete.\r18240 markers complete.\r18300 markers complete.\r18360 markers complete.\r18420 markers complete.\r18480 markers complete.\r18540 markers complete.\r18600 markers complete.\r18660 markers complete.\r18720 markers complete.\r18780 markers complete.\r18840 markers complete.\r18900 markers complete.\r18960 markers complete.\r19020 markers complete.\r19080 markers complete.\r19140 markers complete.\r19200 markers complete.\r19260 markers complete.\r19320 markers complete.\r19380 markers complete.\r19440 markers complete.\r19500 markers complete.\r19560 markers complete.\r19620 markers complete.\r19680 markers complete.\r19740 markers complete.\r19800 markers complete.\r19860 markers complete.\r19920 markers complete.\r19980 markers complete.\r20040 markers complete.\r20100 markers complete.\r20160 markers complete.\r20220 markers complete.\r20280 markers complete.\r20340 markers complete.\r20400 markers complete.\r20460 markers complete.\r20520 markers complete.\r20580 markers complete.\r20640 markers complete.\r20700 markers complete.\r20760 markers complete.\r20820 markers complete.\r20880 markers complete.\r20940 markers complete.\r21000 markers complete.\r21060 markers complete.\r21120 markers complete.\r21180 markers complete.\r21240 markers complete.\r21300 markers complete.\r21360 markers complete.\r21420 markers complete.\r21480 markers complete.\r21540 markers complete.\r21600 markers complete.\r21660 markers complete.\r21720 markers complete.\r21780 markers complete.\r21840 markers complete.\r21900 markers complete.\r21960 markers complete.\r22020 markers complete.\r22080 markers complete.\r22140 markers complete.\r22200 markers complete.\r22260 markers complete.\r22320 markers complete.\r22380 markers complete.\r22440 markers complete.\r22500 markers complete.\r22560 markers complete.\r22620 markers complete.\r22680 markers complete.\r22740 markers complete.\r22800 markers complete.\r22860 markers complete.\r22920 markers complete.\r22980 markers complete.\r23040 markers complete.\r23100 markers complete.\r23160 markers complete.\r23220 markers complete.\r23280 markers complete.\r23340 markers complete.\r23400 markers complete.\r23460 markers complete.\r23520 markers complete.\r23580 markers complete.\r23640 markers complete.\r23700 markers complete.\r23760 markers complete.\r23820 markers complete.\r23880 markers complete.\r23940 markers complete.\r24000 markers complete.\r24060 markers complete.\r24120 markers complete.\r24180 markers complete.\r24240 markers complete.\r24300 markers complete.\r24360 markers complete.\r24420 markers complete.\r24480 markers complete.\r24540 markers complete.\r24600 markers complete.\r24660 markers complete.\r24720 markers complete.\r24780 markers complete.\r24840 markers complete.\r24900 markers complete.\r24960 markers complete.\r25020 markers complete.\r25080 markers complete.\r25140 markers complete.\r25200 markers complete.\r25260 markers complete.\r25320 markers complete.\r25380 markers complete.\r25440 markers complete.\r25500 markers complete.\r25560 markers complete.\r25620 markers complete.\r25680 markers complete.\r25740 markers complete.\r25800 markers complete.\r25860 markers complete.\r25920 markers complete.\r25980 markers complete.\r26040 markers complete.\r26100 markers complete.\r26160 markers complete.\r26220 markers complete.\r26280 markers complete.\r26340 markers complete.\r26400 markers complete.\r26460 markers complete.\r26520 markers complete.\r26580 markers complete.\r26640 markers complete.\r26700 markers complete.\r26760 markers complete.\r26820 markers complete.\r26880 markers complete.\r26940 markers complete.\r27000 markers complete.\r27060 markers complete.\r27120 markers complete.\r27180 markers complete.\r27240 markers complete.\r27300 markers complete.\r27360 markers complete.\r27420 markers complete.\r27480 markers complete.\r27540 markers complete.\r27600 markers complete.\r27660 markers complete.\r27720 markers complete.\r27780 markers complete.\r27840 markers complete.\r27900 markers complete.\r27960 markers complete.\r28020 markers complete.\r28080 markers complete.\r28140 markers complete.\r28200 markers complete.\r28260 markers complete.\r28320 markers complete.\r28380 markers complete.\r28440 markers complete.\r28500 markers complete.\r28560 markers complete.\r28620 markers complete.\r28680 markers complete.\r28740 markers complete.\r28800 markers complete.\r28860 markers complete.\r28920 markers complete.\r28980 markers complete.\r29040 markers complete.\r29100 markers complete.\r29160 markers complete.\r29220 markers complete.\r29280 markers complete.\r29340 markers complete.\r29400 markers complete.\r29460 markers complete.\r29520 markers complete.\r29580 markers complete.\r29640 markers complete.\r29700 markers complete.\r29760 markers complete.\r29820 markers complete.\r29880 markers complete.\r29940 markers complete.\r30000 markers complete.\r30060 markers complete.\r30120 markers complete.\r30180 markers complete.\r30240 markers complete.\r30300 markers complete.\r30360 markers complete.\r30420 markers complete.\r30480 markers complete.\r30540 markers complete.\r30600 markers complete.\r30660 markers complete.\r30720 markers complete.\r30780 markers complete.\r30840 markers complete.\r30900 markers complete.\r30960 markers complete.\r31020 markers complete.\r31080 markers complete.\r31140 markers complete.\r31200 markers complete.\r31260 markers complete.\r31320 markers complete.\r31380 markers complete.\r31440 markers complete.\r31500 markers complete.\r31560 markers complete.\r31620 markers complete.\r31680 markers complete.\r31740 markers complete.\r31800 markers complete.\r31860 markers complete.\r31920 markers complete.\r31980 markers complete.\r32040 markers complete.\r32100 markers complete.\r32160 markers complete.\r32220 markers complete.\r32280 markers complete.\r32340 markers complete.\r32400 markers complete.\r32460 markers complete.\r32520 markers complete.\r32580 markers complete.\r32640 markers complete.\r32700 markers complete.\r32760 markers complete.\r32820 markers complete.\r32880 markers complete.\r32940 markers complete.\r33000 markers complete.\r33060 markers complete.\r33120 markers complete.\r33180 markers complete.\r33240 markers complete.\r33300 markers complete.\r33360 markers complete.\r33420 markers complete.\r33480 markers complete.\r33540 markers complete.\r33600 markers complete.\r33660 markers complete.\r33720 markers complete.\r33780 markers complete.\r33840 markers complete.\r33900 markers complete.\r33960 markers complete.\r34020 markers complete.\r34080 markers complete.\r34140 markers complete.\r34200 markers complete.\r34260 markers complete.\r34320 markers complete.\r34380 markers complete.\r34440 markers complete.\r34500 markers complete.\r34560 markers complete.\r34620 markers complete.\r34680 markers complete.\r34740 markers complete.\r34800 markers complete.\r34860 markers complete.\r34920 markers complete.\r34980 markers complete.\r35040 markers complete.\r35100 markers complete.\r35160 markers complete.\r35220 markers complete.\r35280 markers complete.\r35340 markers complete.\r35400 markers complete.\r35460 markers complete.\r35520 markers complete.\r35580 markers complete.\r35640 markers complete.\r35700 markers complete.\r35760 markers complete.\r35820 markers complete.\r35880 markers complete.\r35940 markers complete.\r36000 markers complete.\r36060 markers complete.\r36120 markers complete.\r36180 markers complete.\r36240 markers complete.\r36300 markers complete.\r36360 markers complete.\r36420 markers complete.\r36480 markers complete.\r36540 markers complete.\r36600 markers complete.\r36660 markers complete.\r36720 markers complete.\r36780 markers complete.\r36840 markers complete.\r36900 markers complete.\r36960 markers complete.\r37020 markers complete.\r37080 markers complete.\r37140 markers complete.\r37200 markers complete.\r37260 markers complete.\r37320 markers complete.\r37380 markers complete.\r37440 markers complete.\r37500 markers complete.\r37560 markers complete.\r37620 markers complete.\r37680 markers complete.\r37740 markers complete.\r37800 markers complete.\r37860 markers complete.\r37920 markers complete.\r37980 markers complete.\r38040 markers complete.\r38100 markers complete.\r38160 markers complete.\r38220 markers complete.\r38280 markers complete.\r38340 markers complete.\r38400 markers complete.\r38460 markers complete.\r38520 markers complete.\r38580 markers complete.\r38640 markers complete.\r38700 markers complete.\r38760 markers complete.\r38820 markers complete.\r38880 markers complete.\r38940 markers complete.\r39000 markers complete.\r39060 markers complete.\r39120 markers complete.\r39180 markers complete.\r39240 markers complete.\r39300 markers complete.\r39360 markers complete.\r39420 markers complete.\r39480 markers complete.\r39540 markers complete.\r39600 markers complete.\r39660 markers complete.\r39720 markers complete.\r39780 markers complete.\r39840 markers complete.\r39900 markers complete.\r39960 markers complete.\r40020 markers complete.\r40080 markers complete.\r40140 markers complete.\r40200 markers complete.\r40260 markers complete.\r40320 markers complete.\r40380 markers complete.\r40440 markers complete.\r40500 markers complete.\r40560 markers complete.\r40620 markers complete.\r40680 markers complete.\r40740 markers complete.\r40800 markers complete.\r40860 markers complete.\r40920 markers complete.\r40980 markers complete.\r41040 markers complete.\r41100 markers complete.\r41160 markers complete.\r41220 markers complete.\r41280 markers complete.\r41340 markers complete.\r41400 markers complete.\r41460 markers complete.\r41520 markers complete.\r41580 markers complete.\r41640 markers complete.\r41700 markers complete.\r41760 markers complete.\r41820 markers complete.\r41880 markers complete.\r41940 markers complete.\r42000 markers complete.\r42060 markers complete.\r42120 markers complete.\r42180 markers complete.\r42240 markers complete.\r42300 markers complete.\r42360 markers complete.\r42420 markers complete.\r42480 markers complete.\r42540 markers complete.\r42600 markers complete.\r42660 markers complete.\r42720 markers complete.\r42780 markers complete.\r42840 markers complete.\r42900 markers complete.\r42960 markers complete.\r43020 markers complete.\r43080 markers complete.\r43140 markers complete.\r43200 markers complete.\r43260 markers complete.\r43320 markers complete.\r43380 markers complete.\r43440 markers complete.\r43500 markers complete.\r43560 markers complete.\r43620 markers complete.\r43680 markers complete.\r43740 markers complete.\r43800 markers complete.\r43860 markers complete.\r43920 markers complete.\r43980 markers complete.\r44040 markers complete.\r44100 markers complete.\r44160 markers complete.\r44220 markers complete.\r44280 markers complete.\r44340 markers complete.\r44400 markers complete.\r44460 markers complete.\r44520 markers complete.\r44580 markers complete.\r44640 markers complete.\r44700 markers complete.\r44760 markers complete.\r44820 markers complete.\r44880 markers complete.\r44940 markers complete.\r45000 markers complete.\r45060 markers complete.\r45120 markers complete.\r45180 markers complete.\r45240 markers complete.\r45300 markers complete.\r45360 markers complete.\r45420 markers complete.\r45480 markers complete.\r45540 markers complete.\r45600 markers complete.\r45660 markers complete.\r45720 markers complete.\r45780 markers complete.\r45840 markers complete.\r45900 markers complete.\r45960 markers complete.\r46020 markers complete.\r46080 markers complete.\r46140 markers complete.\r46200 markers complete.\r46260 markers complete.\r46320 markers complete.\r46380 markers complete.\r46440 markers complete.\r46500 markers complete.\r46560 markers complete.\r46620 markers complete.\r46680 markers complete.\r46740 markers complete.\r46800 markers complete.\r46860 markers complete.\r46920 markers complete.\r46980 markers complete.\r47040 markers complete.\r47100 markers complete.\r47160 markers complete.\r47220 markers complete.\r47280 markers complete.\r47340 markers complete.\r47400 markers complete.\r47460 markers complete.\r47520 markers complete.\r47580 markers complete.\r47640 markers complete.\r47700 markers complete.\r47760 markers complete.\r47820 markers complete.\r47880 markers complete.\r47940 markers complete.\r48000 markers complete.\r48060 markers complete.\r48120 markers complete.\r48180 markers complete.\r48240 markers complete.\r48300 markers complete.\r48360 markers complete.\r48420 markers complete.\r48480 markers complete.\r48540 markers complete.\r48600 markers complete.\r48660 markers complete.\r48720 markers complete.\r48780 markers complete.\r48840 markers complete.\r48900 markers complete.\r48960 markers complete.\r49020 markers complete.\r49080 markers complete.\r49140 markers complete.\r49200 markers complete.\r49260 markers complete.\r49320 markers complete.\r49380 markers complete.\r49440 markers complete.\r49500 markers complete.\r49560 markers complete.\r49620 markers complete.\r49680 markers complete.\r49740 markers complete.\r49800 markers complete.\r49860 markers complete.\r49920 markers complete.\r49980 markers complete.\r50040 markers complete.\r50100 markers complete.\r50160 markers complete.\r50220 markers complete.\r50280 markers complete.\r50340 markers complete.\r50400 markers complete.\r50460 markers complete.\r50520 markers complete.\r50580 markers complete.\r50640 markers complete.\r50700 markers complete.\r50760 markers complete.\r50820 markers complete.\r50880 markers complete.\r50940 markers complete.\r51000 markers complete.\r51060 markers complete.\r51120 markers complete.\r51180 markers complete.\r51240 markers complete.\r51300 markers complete.\r51360 markers complete.\r51420 markers complete.\r51480 markers complete.\r51540 markers complete.\r51600 markers complete.\r51660 markers complete.\r51720 markers complete.\r51780 markers complete.\r51840 markers complete.\r51900 markers complete.\r51960 markers complete.\r52020 markers complete.\r52080 markers complete.\r52140 markers complete.\r52200 markers complete.\r52260 markers complete.\r52320 markers complete.\r52380 markers complete.\r52440 markers complete.\r52500 markers complete.\r52560 markers complete.\r52620 markers complete.\r52680 markers complete.\r52740 markers complete.\r52800 markers complete.\r52860 markers complete.\r52920 markers complete.\r52980 markers complete.\r53040 markers complete.\r53100 markers complete.\r53160 markers complete.\r53220 markers complete.\r53280 markers complete.\r53340 markers complete.\r53400 markers complete.\r53460 markers complete.\r53520 markers complete.\r53580 markers complete.\r53640 markers complete.\r53700 markers complete.\r53760 markers complete.\r53820 markers complete.\r53880 markers complete.\r53940 markers complete.\r54000 markers complete.\r54060 markers complete.\r54120 markers complete.\r54180 markers complete.\r54240 markers complete.\r54300 markers complete.\r54360 markers complete.\r54420 markers complete.\r54480 markers complete.\r54540 markers complete.\r54600 markers complete.\r54660 markers complete.\r54720 markers complete.\r54780 markers complete.\r54840 markers complete.\r54900 markers complete.\r54960 markers complete.\r55020 markers complete.\r55080 markers complete.\r55140 markers complete.\r55200 markers complete.\r55260 markers complete.\r55320 markers complete.\r55380 markers complete.\r55440 markers complete.\r55500 markers complete.\r55560 markers complete.\r55620 markers complete.\r55680 markers complete.\r55740 markers complete.\r55800 markers complete.\r55860 markers complete.\r55920 markers complete.\r55980 markers complete.\r56040 markers complete.\r56100 markers complete.\r56160 markers complete.\r56220 markers complete.\r56280 markers complete.\r56340 markers complete.\r56400 markers complete.\r56460 markers complete.\r56520 markers complete.\r56580 markers complete.\r56640 markers complete.\r56700 markers complete.\r56760 markers complete.\r56820 markers complete.\r56880 markers complete.\r56940 markers complete.\r57000 markers complete.\r57060 markers complete.\r57120 markers complete.\r57180 markers complete.\r57240 markers complete.\r57300 markers complete.\r57360 markers complete.\r57420 markers complete.\r57480 markers complete.\r57540 markers complete.\r57600 markers complete.\r57660 markers complete.\r57720 markers complete.\r57780 markers complete.\r57840 markers complete.\r57900 markers complete.\r57960 markers complete.\r58020 markers complete.\r58080 markers complete.\r58140 markers complete.\r58200 markers complete.\r58260 markers complete.\r58320 markers complete.\r58380 markers complete.\r58440 markers complete.\r58500 markers complete.\r58560 markers complete.\r58620 markers complete.\r58680 markers complete.\r58740 markers complete.\r58800 markers complete.\r58860 markers complete.\r58920 markers complete.\r58980 markers complete.\r59040 markers complete.\r59100 markers complete.\r59160 markers complete.\r59220 markers complete.\r59280 markers complete.\r59340 markers complete.\r59400 markers complete.\r59460 markers complete.\r59520 markers complete.\r59580 markers complete.\r59640 markers complete.\r59700 markers complete.\r59760 markers complete.\r59820 markers complete.\r59880 markers complete.\r59940 markers complete.\r60000 markers complete.\r60060 markers complete.\r60120 markers complete.\r60180 markers complete.\r60240 markers complete.\r60300 markers complete.\r60360 markers complete.\r60420 markers complete.\r60480 markers complete.\r60540 markers complete.\r60600 markers complete.\r60660 markers complete.\r60720 markers complete.\r60780 markers complete.\r60840 markers complete.\r60900 markers complete.\r60960 markers complete.\r61020 markers complete.\r61080 markers complete.\r61140 markers complete.\r61200 markers complete.\r61260 markers complete.\r61320 markers complete.\r61380 markers complete.\r61440 markers complete.\r61500 markers complete.\r61560 markers complete.\r61620 markers complete.\r61680 markers complete.\r61740 markers complete.\r61800 markers complete.\r61860 markers complete.\r61920 markers complete.\r61980 markers complete.\r62040 markers complete.\r62100 markers complete.\r62160 markers complete.\r62220 markers complete.\r62280 markers complete.\r62340 markers complete.\r62400 markers complete.\r62460 markers complete.\r62520 markers complete.\r62580 markers complete.\r62640 markers complete.\r62700 markers complete.\r62760 markers complete.\r62820 markers complete.\r62880 markers complete.\r62940 markers complete.\r63000 markers complete.\r63060 markers complete.\r63120 markers complete.\r63180 markers complete.\r63240 markers complete.\r63300 markers complete.\r63360 markers complete.\r63420 markers complete.\r63480 markers complete.\r63540 markers complete.\r63600 markers complete.\r63660 markers complete.\r63720 markers complete.\r63780 markers complete.\r63840 markers complete.\r63900 markers complete.\r63960 markers complete.\r64020 markers complete.\r64080 markers complete.\r64140 markers complete.\r64200 markers complete.\r64260 markers complete.\r64320 markers complete.\r64380 markers complete.\r64440 markers complete.\r64500 markers complete.\r64560 markers complete.\r64620 markers complete.\r64680 markers complete.\r64740 markers complete.\r64800 markers complete.\r64860 markers complete.\r64920 markers complete.\r64980 markers complete.\r65040 markers complete.\r65100 markers complete.\r65160 markers complete.\r65220 markers complete.\r65280 markers complete.\r65340 markers complete.\r65400 markers complete.\r65460 markers complete.\r65520 markers complete.\r65580 markers complete.\r65640 markers complete.\r65700 markers complete.\r65760 markers complete.\r65820 markers complete.\r65880 markers complete.\r65940 markers complete.\r66000 markers complete.\r66060 markers complete.\r66120 markers complete.\r66180 markers complete.\r66240 markers complete.\r66300 markers complete.\r66360 markers complete.\r66420 markers complete.\r66480 markers complete.\r66540 markers complete.\r66600 markers complete.\r66660 markers complete.\r66720 markers complete.\r66780 markers complete.\r66840 markers complete.\r66900 markers complete.\r66960 markers complete.\r67020 markers complete.\r67080 markers complete.\r67140 markers complete.\r67200 markers complete.\r67260 markers complete.\r67320 markers complete.\r67380 markers complete.\r67440 markers complete.\r67500 markers complete.\r67560 markers complete.\r67620 markers complete.\r67680 markers complete.\r67740 markers complete.\r67800 markers complete.\r67860 markers complete.\r67920 markers complete.\r67980 markers complete.\r68040 markers complete.\r68100 markers complete.\r68160 markers complete.\r68220 markers complete.\r68280 markers complete.\r68340 markers complete.\r68400 markers complete.\r68460 markers complete.\r68520 markers complete.\r68580 markers complete.\r68640 markers complete.\r68700 markers complete.\r68760 markers complete.\r68820 markers complete.\r68880 markers complete.\r68940 markers complete.\r69000 markers complete.\r69060 markers complete.\r69120 markers complete.\r69180 markers complete.\r69240 markers complete.\r69300 markers complete.\r69360 markers complete.\r69420 markers complete.\r69480 markers complete.\r69540 markers complete.\r69600 markers complete.\r69660 markers complete.\r69720 markers complete.\r69780 markers complete.\r69840 markers complete.\r69900 markers complete.\r69960 markers complete.\r70020 markers complete.\r70080 markers complete.\r70140 markers complete.\r70200 markers complete.\r70260 markers complete.\r70320 markers complete.\r70380 markers complete.\r70440 markers complete.\r70500 markers complete.\r70560 markers complete.\r70620 markers complete.\r70680 markers complete.\r70740 markers complete.\r70800 markers complete.\r70860 markers complete.\r70920 markers complete.\r70980 markers complete.\r71040 markers complete.\r71100 markers complete.\r71160 markers complete.\r71220 markers complete.\r71280 markers complete.\r71340 markers complete.\r71400 markers complete.\r71460 markers complete.\r71520 markers complete.\r71580 markers complete.\r71640 markers complete.\r71700 markers complete.\r71760 markers complete.\r71820 markers complete.\r71880 markers complete.\r71940 markers complete.\r72000 markers complete.\r72060 markers complete.\r72120 markers complete.\r72180 markers complete.\r72240 markers complete.\r72300 markers complete.\r72360 markers complete.\r72420 markers complete.\r72480 markers complete.\r72540 markers complete.\r72600 markers complete.\r72660 markers complete.\r72720 markers complete.\r72780 markers complete.\r72840 markers complete.\r72900 markers complete.\r72960 markers complete.\r73020 markers complete.\r73080 markers complete.\r73140 markers complete.\r73200 markers complete.\r73260 markers complete.\r73320 markers complete.\r73380 markers complete.\r73440 markers complete.\r73500 markers complete.\r73560 markers complete.\r73620 markers complete.\r73680 markers complete.\r73740 markers complete.\r73800 markers complete.\r73860 markers complete.\r73920 markers complete.\r73980 markers complete.\r74040 markers complete.\r74100 markers complete.\r74160 markers complete.\r74220 markers complete.\r74280 markers complete.\r74340 markers complete.\r74400 markers complete.\r74460 markers complete.\r74520 markers complete.\r74580 markers complete.\r74640 markers complete.\r74700 markers complete.\r74760 markers complete.\r74820 markers complete.\r74880 markers complete.\r74940 markers complete.\r75000 markers complete.\r75060 markers complete.\r75120 markers complete.\r75180 markers complete.\r75240 markers complete.\r75300 markers complete.\r75360 markers complete.\r75420 markers complete.\r75480 markers complete.\r75540 markers complete.\r75600 markers complete.\r75660 markers complete.\r75720 markers complete.\r75780 markers complete.\r75840 markers complete.\r75900 markers complete.\r75960 markers complete.\r76020 markers complete.\r76080 markers complete.\r76140 markers complete.\r76200 markers complete.\r76260 markers complete.\r76320 markers complete.\r76380 markers complete.\r76440 markers complete.\r76500 markers complete.\r76560 markers complete.\r76620 markers complete.\r76680 markers complete.\r76740 markers complete.\r76800 markers complete.\r76860 markers complete.\r76920 markers complete.\r76980 markers complete.\r77040 markers complete.\r77100 markers complete.\r77160 markers complete.\r77220 markers complete.\r77280 markers complete.\r77340 markers complete.\r77400 markers complete.\r77460 markers complete.\r77520 markers complete.\r77580 markers complete.\r77640 markers complete.\r77700 markers complete.\r77760 markers complete.\r77820 markers complete.\r77880 markers complete.\r77940 markers complete.\r78000 markers complete.\r78060 markers complete.\r78120 markers complete.\r78180 markers complete.\r78240 markers complete.\r78300 markers complete.\r78360 markers complete.\r78420 markers complete.\r78480 markers complete.\r78540 markers complete.\r78600 markers complete.\r78660 markers complete.\r78720 markers complete.\r78780 markers complete.\r78840 markers complete.\r78900 markers complete.\r78960 markers complete.\r79020 markers complete.\r79080 markers complete.\r79140 markers complete.\r79200 markers complete.\r79260 markers complete.\r79320 markers complete.\r79380 markers complete.\r79440 markers complete.\r79500 markers complete.\r79560 markers complete.\r79620 markers complete.\r79680 markers complete.\r79740 markers complete.\r79800 markers complete.\r79860 markers complete.\r79920 markers complete.\r79980 markers complete.\r80040 markers complete.\r80100 markers complete.\r80160 markers complete.\r80220 markers complete.\r80280 markers complete.\r80340 markers complete.\r80400 markers complete.\r80460 markers complete.\r80520 markers complete.\r80580 markers complete.\r80640 markers complete.\r80700 markers complete.\r80760 markers complete.\r80820 markers complete.\r80880 markers complete.\r80940 markers complete.\r81000 markers complete.\r81060 markers complete.\r81120 markers complete.\r81180 markers complete.\r81240 markers complete.\r81300 markers complete.\r81360 markers complete.\r81420 markers complete.\r81480 markers complete.\r81540 markers complete.\r81600 markers complete.\r81660 markers complete.\r81720 markers complete.\r81780 markers complete.\r81840 markers complete.\r81900 markers complete.\r81960 markers complete.\r82020 markers complete.\r82080 markers complete.\r82140 markers complete.\r82200 markers complete.\r82260 markers complete.\r82320 markers complete.\r82380 markers complete.\r82440 markers complete.\r82500 markers complete.\r82560 markers complete.\r82620 markers complete.\r82680 markers complete.\r82740 markers complete.\r82800 markers complete.\r82860 markers complete.\r82920 markers complete.\r82980 markers complete.\r83040 markers complete.\r83100 markers complete.\r83160 markers complete.\r83220 markers complete.\r83280 markers complete.\r83340 markers complete.\r83400 markers complete.\r83460 markers complete.\r83520 markers complete.\r83580 markers complete.\r83640 markers complete.\r83700 markers complete.\r83760 markers complete.\r83820 markers complete.\r83880 markers complete.\r83940 markers complete.\r84000 markers complete.\r84060 markers complete.\r84120 markers complete.\r84180 markers complete.\r84240 markers complete.\r84300 markers complete.\r84360 markers complete.\r84420 markers complete.\r84480 markers complete.\r84540 markers complete.\r84600 markers complete.\r84660 markers complete.\r84720 markers complete.\r84780 markers complete.\r84840 markers complete.\r84900 markers complete.\r84960 markers complete.\r85020 markers complete.\r85080 markers complete.\r85140 markers complete.\r85200 markers complete.\r85260 markers complete.\r85320 markers complete.\r85380 markers complete.\r85440 markers complete.\r85500 markers complete.\r85560 markers complete.\r85620 markers complete.\r85680 markers complete.\r85740 markers complete.\r85800 markers complete.\r85860 markers complete.\r85920 markers complete.\r85980 markers complete.\r86040 markers complete.\r86100 markers complete.\r86160 markers complete.\r86220 markers complete.\r86280 markers complete.\r86340 markers complete.\r86400 markers complete.\r86460 markers complete.\r86520 markers complete.\r86580 markers complete.\r86640 markers complete.\r86700 markers complete.\r86760 markers complete.\r86820 markers complete.\r86880 markers complete.\r86940 markers complete.\r87000 markers complete.\r87060 markers complete.\r87120 markers complete.\r87180 markers complete.\r87240 markers complete.\r87300 markers complete.\r87360 markers complete.\r87420 markers complete.\r87480 markers complete.\r87540 markers complete.\r87600 markers complete.\r87660 markers complete.\r87720 markers complete.\r87780 markers complete.\r87840 markers complete.\r87900 markers complete.\r87960 markers complete.\r88020 markers complete.\r88080 markers complete.\r88140 markers complete.\r88200 markers complete.\r88260 markers complete.\r88320 markers complete.\r88380 markers complete.\r88440 markers complete.\r88500 markers complete.\r88560 markers complete.\r88620 markers complete.\r88680 markers complete.\r88740 markers complete.\r88800 markers complete.\r88860 markers complete.\r88920 markers complete.\r88980 markers complete.\r89040 markers complete.\r89100 markers complete.\r89160 markers complete.\r89220 markers complete.\r89280 markers complete.\r89340 markers complete.\r89400 markers complete.\r89460 markers complete.\r89520 markers complete.\r89580 markers complete.\r89640 markers complete.\r89700 markers complete.\r89760 markers complete.\r89820 markers complete.\r89880 markers complete.\r89940 markers complete.\r90000 markers complete.\r90060 markers complete.\r90120 markers complete.\r90180 markers complete.\r90240 markers complete.\r90300 markers complete.\r90360 markers complete.\r90420 markers complete.\r90480 markers complete.\r90540 markers complete.\r90600 markers complete.\r90660 markers complete.\r90720 markers complete.\r90780 markers complete.\r90840 markers complete.\r90900 markers complete.\r90960 markers complete.\r91020 markers complete.\r91080 markers complete.\r91140 markers complete.\r91200 markers complete.\r91260 markers complete.\r91320 markers complete.\r91380 markers complete.\r91440 markers complete.\r91500 markers complete.\r91560 markers complete.\r91620 markers complete.\r91680 markers complete.\r91740 markers complete.\r91800 markers complete.\r91860 markers complete.\r91920 markers complete.\r91980 markers complete.\r92040 markers complete.\r92100 markers complete.\r92160 markers complete.\r92220 markers complete.\r92280 markers complete.\r92340 markers complete.\r92400 markers complete.\r92460 markers complete.\r92520 markers complete.\r92580 markers complete.\r92640 markers complete.\r92700 markers complete.\r92760 markers complete.\r92820 markers complete.\r92880 markers complete.\r92940 markers complete.\r93000 markers complete.\r93060 markers complete.\r93120 markers complete.\r93180 markers complete.\r93240 markers complete.\r93300 markers complete.\r93360 markers complete.\r93420 markers complete.\r93480 markers complete.\r93540 markers complete.\r93600 markers complete.\r93660 markers complete.\r93720 markers complete.\r93780 markers complete.\r93840 markers complete.\r93900 markers complete.\r93960 markers complete.\r94020 markers complete.\r94080 markers complete.\r94140 markers complete.\r94200 markers complete.\r94260 markers complete.\r94320 markers complete.\r94380 markers complete.\r94440 markers complete.\r94500 markers complete.\r94560 markers complete.\r94620 markers complete.\r94680 markers complete.\r94740 markers complete.\r94800 markers complete.\r94860 markers complete.\r94920 markers complete.\r94980 markers complete.\r95040 markers complete.\r95100 markers complete.\r95160 markers complete.\r95220 markers complete.\r95280 markers complete.\r95340 markers complete.\r95400 markers complete.\r95460 markers complete.\r95520 markers complete.\r95580 markers complete.\r95640 markers complete.\r95700 markers complete.\r95760 markers complete.\r95820 markers complete.\r95880 markers complete.\r95940 markers complete.\r96000 markers complete.\r96060 markers complete.\r96120 markers complete.\r96180 markers complete.\r96240 markers complete.\r96300 markers complete.\r96360 markers complete.\r96420 markers complete.\r96480 markers complete.\r96540 markers complete.\r96600 markers complete.\r96660 markers complete.\r96720 markers complete.\r96780 markers complete.\r96840 markers complete.\r96900 markers complete.\r96960 markers complete.\r97020 markers complete.\r97080 markers complete.\r97140 markers complete.\r97200 markers complete.\r97260 markers complete.\r97320 markers complete.\r97380 markers complete.\r97440 markers complete.\r97500 markers complete.\r97560 markers complete.\r97620 markers complete.\r97680 markers complete.\r97740 markers complete.\r97800 markers complete.\r97860 markers complete.\r97920 markers complete.\r97980 markers complete.\r98040 markers complete.\r98100 markers complete.\r98160 markers complete.\r98220 markers complete.\r98280 markers complete.\r98340 markers complete.\r98400 markers complete.\r98460 markers complete.\r98520 markers complete.\r98580 markers complete.\r98640 markers complete.\r98700 markers complete.\r98760 markers complete.\r98820 markers complete.\r98880 markers complete.\r98940 markers complete.\r99000 markers complete.\r99060 markers complete.\r99120 markers complete.\r99180 markers complete.\r99240 markers complete.\r99300 markers complete.\r99360 markers complete.\r99420 markers complete.\r99480 markers complete.\r99540 markers complete.\r99600 markers complete.\r99660 markers complete.\r99720 markers complete.\r99780 markers complete.\r99840 markers complete.\r99900 markers complete.\r99960 markers complete.\r100020 markers complete.\r100080 markers complete.\r100140 markers complete.\r100200 markers complete.\r100260 markers complete.\r100320 markers complete.\r100380 markers complete.\r100440 markers complete.\r100500 markers complete.\r100560 markers complete.\r100620 markers complete.\r100680 markers complete.\r100740 markers complete.\r100800 markers complete.\r100860 markers complete.\r100920 markers complete.\r100980 markers complete.\r101040 markers complete.\r101100 markers complete.\r101160 markers complete.\r101220 markers complete.\r101280 markers complete.\r101340 markers complete.\r101400 markers complete.\r101460 markers complete.\r101520 markers complete.\r101580 markers complete.\r101640 markers complete.\r101700 markers complete.\r101760 markers complete.\r101820 markers complete.\r101880 markers complete.\r101940 markers complete.\r102000 markers complete.\r102060 markers complete.\r102120 markers complete.\r102180 markers complete.\r102240 markers complete.\r102300 markers complete.\r102360 markers complete.\r102420 markers complete.\r102480 markers complete.\r102540 markers complete.\r102600 markers complete.\r102660 markers complete.\r102720 markers complete.\r102780 markers complete.\r102840 markers complete.\r102900 markers complete.\r102960 markers complete.\r103020 markers complete.\r103080 markers complete.\r103140 markers complete.\r103200 markers complete.\r103260 markers complete.\r103320 markers complete.\r103380 markers complete.\r103440 markers complete.\r103500 markers complete.\r103560 markers complete.\r103620 markers complete.\r103680 markers complete.\r103740 markers complete.\r103800 markers complete.\r103860 markers complete.\r103920 markers complete.\r103980 markers complete.\r104040 markers complete.\r104100 markers complete.\r104160 markers complete.\r104220 markers complete.\r104280 markers complete.\r104340 markers complete.\r104400 markers complete.\r104460 markers complete.\r104520 markers complete.\r104580 markers complete.\r104640 markers complete.\r104700 markers complete.\r104760 markers complete.\r104820 markers complete.\r104880 markers complete.\r104940 markers complete.\r105000 markers complete.\r105060 markers complete.\r105120 markers complete.\r105180 markers complete.\r105240 markers complete.\r105300 markers complete.\r105360 markers complete.\r105420 markers complete.\r105480 markers complete.\r105540 markers complete.\r105600 markers complete.\r105660 markers complete.\r105720 markers complete.\r105780 markers complete.\r105840 markers complete.\r105900 markers complete.\r105960 markers complete.\r106020 markers complete.\r106080 markers complete.\r106140 markers complete.\r106200 markers complete.\r106260 markers complete.\r106320 markers complete.\r106380 markers complete.\r106440 markers complete.\r106500 markers complete.\r106560 markers complete.\r106620 markers complete.\r106680 markers complete.\r106740 markers complete.\r106800 markers complete.\r106860 markers complete.\r106920 markers complete.\r106980 markers complete.\r107040 markers complete.\r107100 markers complete.\r107160 markers complete.\r107220 markers complete.\r107280 markers complete.\r107340 markers complete.\r107400 markers complete.\r107460 markers complete.\r107520 markers complete.\r107580 markers complete.\r107640 markers complete.\r107700 markers complete.\r107760 markers complete.\r107820 markers complete.\r107880 markers complete.\r107940 markers complete.\r108000 markers complete.\r108060 markers complete.\r108120 markers complete.\r108180 markers complete.\r108240 markers complete.\r108300 markers complete.\r108360 markers complete.\r108420 markers complete.\r108480 markers complete.\r108540 markers complete.\r108600 markers complete.\r108660 markers complete.\r108720 markers complete.\r108780 markers complete.\r108840 markers complete.\r108900 markers complete.\r108960 markers complete.\r109020 markers complete.\r109080 markers complete.\r109140 markers complete.\r109200 markers complete.\r109260 markers complete.\r109320 markers complete.\r109380 markers complete.\r109440 markers complete.\r109500 markers complete.\r109560 markers complete.\r109620 markers complete.\r109680 markers complete.\r109740 markers complete.\r109800 markers complete.\r109860 markers complete.\r109920 markers complete.\r109980 markers complete.\r110040 markers complete.\r110100 markers complete.\r110160 markers complete.\r110220 markers complete.\r110280 markers complete.\r110340 markers complete.\r110400 markers complete.\r110460 markers complete.\r110520 markers complete.\r110580 markers complete.\r110640 markers complete.\r110700 markers complete.\r110760 markers complete.\r110820 markers complete.\r110880 markers complete.\r110940 markers complete.\r111000 markers complete.\r111060 markers complete.\r111120 markers complete.\r111180 markers complete.\r111240 markers complete.\r111300 markers complete.\r111360 markers complete.\r111420 markers complete.\r111480 markers complete.\r111540 markers complete.\r111600 markers complete.\r111660 markers complete.\r111720 markers complete.\r111780 markers complete.\r111840 markers complete.\r111900 markers complete.\r111960 markers complete.\r112020 markers complete.\r112080 markers complete.\r112140 markers complete.\r112200 markers complete.\r112260 markers complete.\r112320 markers complete.\r112380 markers complete.\r112440 markers complete.\r112500 markers complete.\r112560 markers complete.\r112620 markers complete.\r112669 markers complete.\rRelationship matrix calculation complete.\n",
            "[extracting eigenvalues and eigenvectors]\r--pca: Results saved to\n",
            "/content/analysis//penncath.geno0.01.mind0.01.LD0.3.eigenval and\n",
            "/content/analysis//penncath.geno0.01.mind0.01.LD0.3.eigenvec .\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "plot PCA"
      ],
      "metadata": {
        "id": "E_hEaF5D7tBS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "pca <- fread(glue(analysis_dir, \"penncath.geno0.01.mind0.01.LD0.3.eigenvec\"))\n",
        "\n",
        "pca %>% rename (PC1 = V3, PC2 = V4) %>%\n",
        "        ggplot(aes(x = PC1, y = PC2)) +\n",
        "        geom_point() +\n",
        "        labs(title = \"PC1 vs. PC2\") + theme_classic()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 497
        },
        "id": "oZk8JX3b7t2r",
        "outputId": "b4b46ce1-0794-4b7a-be01-96b056ba4a12"
      },
      "execution_count": 64,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAMAAABKCk6nAAADAFBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////isF19AAAgAElEQVR4nO2dB3wUx9n/n7x53ziJ83/zT1zikuoSJ3ZsJ3HixLGNTVVDEkKidzDNdAMG06uE6M30ZorpvVdhqimiG0xvQoCokhDqN+88M7t3e/32bvdub29+n49uZ+d2Z+f4srtTnnkeIEKmFoS6AkL6SgA2uQRgk0sANrlUAX7qf5544tk62TS1+sOfPfnXyZg35+eD1Z5z4P1fvNjfv+oKqZU6wCsJuZVQmZApT07MfrTm+T6EdEj5yAtgp3NyfznFcuap5QFVW8hXqQZMMn9QUPDzCbh7eA0hB0miBPi9UfTjtTnlnV965a1Nns65M5Nu44doUXshr1IP+OAPi3bBQ1umDHjSuxTkk/mbf/+Y7Gns5RyS88x+v6sspEaqAefEJ5FVP1VkyoDv/egK6dqYHP/ZpGyv5+T8q6+/FRZSJ3WAf/rzn/+25X2yB+7YMmXAJHG45dc7CNlZ6xd/2+D5nGMvjwykzkIqpP4RTVX4S/YGPdwFP62Al/7zm99ZMFE66ycFns7JfGF9IFUWUiO/AJNZP0q9+mjVr0Zj2gq46Bd16IN3dosScuxHj8mq4+7OefzSOg0qLuSb/ANM1n/45M/+vYwmnnjiv/77iSSe2fIH5wnJa/Sbl95cQcg7ae7OWQpPUNXToPJC3iVGskwuAdjkEoBNLgHY5BKATS4B2OQSgE0uAdjkUgG4KF+/agjpJRWAZ9bQrxpCekkANrkEYJNLADa5BGCTSwA2uQRgk0sANrkEYJNLADa5BGCTSwA2uQRgk0sANrkEYJNLADa5BGCTSwA2onYDtC3TpigB2IgCqq80Ksr3QwXgYKkEAffXpiwB2IiqTQEf1KYoAdiIyk/vsFujogRgk0sANrkEYFOo2G2nSgA2g7oCTHHzlQBsRF3blKPm8F3Yrbrr+jsB2IBaRXl9q+L4xQj4jOvvBGADKo7y8uaFqHhav2Ny+g49PsrNW1gANqCqU2ANvBzTiB5zRN65njY+z81xArABtZ7CO8xShY0g6qirQ8rwqdzPh7IEYCPqZsY9nhhBMca4PAQBT/WhKAHY0GqHHF1+s5U+xYt9KEEANrQ2U76t3HznC14B2Oja0mFCUUAFCMAG17JGA901kH2SAGxsraXP6IaBFCAAG1sd3bayfJQAbGxNonzjCBkKsNS/AgRgY6u0AyScJvvxPs71qwABOBz0tYfZBC8SgMNBWZRvlVK/ThWAw0KnuvS56d+ZArDJJQAbUNap3azPP3U5l6RCArCBdKshtC0g2dUh+hzbL61M373XlEdsadJDlTGPAGwoNaY8+178mH7WZvtnsfE8X3HAId4rViMB2EBCnhXwA2LZfj4mDygOSMcMdUPTArBxtBPx/YkBbsrnAtcA2IXxXK1+4FIANo5GgKQfA0TZJgnLT163pvtBnBpzSyIAG0m7QKHNcm5hHEAA4dIFYANpMAXbTgK8U85cgHv+h8sQgI0jy+TK0WsLOd96347aRrPO1X0vAXfv+12oAGwczaEkm5C98dFfnTuJzakxhFTiuFv6X6gAbAhl4TO4GWsjr6/yK4i7Vov1lR4xvFN2ej3fvQRgA6gwEWAWIWMpzJRrjGnjNvQjkZC/4c6VQMoWgA2gGYixjJR2g3rXNjDAcdcqA2QScvoFgE9I+bUiYhkd3/GeH2ULwAbQKGRawNM3GOCxhNwu39Sgc3ZpxiGSUwVg00w/re8EYAPoKuB9Kmln0r/rLbHQRCauGcScXjRRrbmf1ncCsBGUPXVNuW3v0eVh8QMLyWipf/TwVUx8aZ2CUCcB2HBaxh7SXckm3NBb+XPcvpvdO67JDT9KE4CNoxNfsWlgkOeTUvnAc022n+JnoQKwIXT6EiHzKcaNhFgcxzam8ww/SxaADaCyOgCdmeOGZLr3GZtQqjJa/tYy/yPeKfZLArABtA2RXkqUOkKnpemGbdbvLYPi6vs72iEAG0DosgGWZtCPnhC1caQEuLn07aND/k81CMCG0OOqjOipU3twkyoB7sa/xBt6k/9lC8AhVtGaLeWkcOU7FGMHMo+R/QcHPIp9X9yUJl946Hf5AnBoVVgNoBbt7EZTjK3ZhCHVwyi2wbUqfXhOd78vIACHVmxu4TQhU+nm6Cnp4byltBtucgk5JuUk+X0BATi02o74LtLEqWXZpJ+EE2qyif6462SulDHD7wsIwKEVdoE7yjvjwKY/079/PompF+FP0y1+X0AADrVOnrcmHyUADJMA/0Um/R79a+d/8QKwoXS7hOyqzTtJMuAP6F+0/yUKwMZTS4o0NvnjJoMY4Jfo35O7/C5MADaCSr89JyfLCcnrVeVdRFskD3lQPfK3aAHYAHpE+71fsNSpapB4t/NvWeMKKvzhF1bAx/0tWwA2gOZJvV5CarBWsws99rdsAdgAYgNYbEZBgfQDaftz+ve8/+v8BWADKBdwIBrVx8r37Ua/oZ+/jqkznG5GeynAgwRgI6hw0yGeyEO2P+GI//lreAX9N2zptzqAogVgYwmXmh2c3qk9Er7m/XDvEoANoY1vv7WOJXLTu2DY0QMIWJMQ0QKwEXQZeZ5T5syKTjqsSdECsBGEy85gqC5FC8BG0F4EvFWXogXg0GomVFtCN20Amjl+dTIjsGgNXAJwSMVcnd1SZGyLiZnOU70AKj8I/AoCcEjFDGb32fbZkn4Wk/AxppQxY0tmfOFP2HcBOKTKcZgpYkbvCwg5/kkLTClDm31K93eov4IAHFqdbNP+vGK3GLGmF93lY1mVldaymNFe/QUEYGPpPPqqrIUBz2COfSMLDWvT1RcoABtL59mtexI/HFYDHwCoXaC+QAHYQNqatosF84ayBf969q1Fjl/7NXQpAIdW91K/OCun0UsD88cyn7ehx2lxAQE4pCqvqngW4wLhp+HpinvkJaT+uE1ylAAcUrFX7jxppxZvO/8WEi8VsdQGDa4gAIdUbIJftok9XQme4YxrYSw72ivyf1GhVQJwaPU1QG/b3oMFssVOG7ZOqV7gFxCAjaFv2/Zh4VT2jOjJCXdnrjoCL1gANoTOUJjVpPSiVowwGtvFBF6yAGwIYQ8JJvTazveYD/CFnSDhVOAlC8CG0Gbp1cvtJ5lDUr/XMthLADaGpLXfjfjemmowSaOCBWCjiK3+9t8XhzsJwLrpgTp7DAzmDioDE/ogAVgv9bR6uvJNuVUdwpxpIwFYJ51W3VAqPazJUgYHCcA6iVnCZoS6Fu4AF6SmjyyXNgfaDxnCXg0CsBoV0Sdu5cKAirjw1bHA6+Ea8MqdZOkeabNDNsgWgFUpf+YMdYFg89I77SMko/cCybs/OiedGXA1XAMec51kzpM2aweMmFJKSNo771QP+GpCbnS7Xfz0ZAo0cy396MHz6oMcRzgQuQe8WNrcvUuWbyKkMHeSuIN1U7I0kjW4iW2K4ROaCvyecg14TQZZdFDaHLtJNrKljeIRrZ/kWcIFnejHhzzvHE1+E3jJLnML09PGWc6PYptLfYelMhcgArB+iqcwewO0K8foshWkzOJTfjtPskl0kwyhS3ViBzF/lI20mQW2SQA2ljbii1jLAgVgpQ7WiF8c4iocSF2raXkCsELMlvFkqGuhrQRghdhq3TmhroW2EoAVYmv7Toe6FtpKX8CFuy6oPieUOl4nXts3YOilK+DboMsUp5AK6QoYLT+h3PtxQvpJV8DpAnDIFSjg0s6QdNZFPtNNyjfVj0qFVKVrlwc2jWssBQp4sscpj/zNYdcmLa8BUNVv99vGU6CAO2g9dhpqMe8JG0NdC+0UKOBV9J+jgWa1MYAugCazdIZRwI2sr+v20WAVq4H0GUBjEzUMxUiWk66G1+CMFwnAJpcAbHIJwCZXoIAXJ7bUY8GFkFYKEPAu2qeI0q42QporQMAsfGK+dtUR0loBAt5otoEs0ynQd/Do6GQNPIUI6SbRija5BGAyBGBVqOugnwIGXLp6YXg3svZgM0KDNSIGVaCALckAFdWtgzWYvkLA570fF6YKFDCbXUvYrFl9gq+r9AdUMdH0kYMCBXyTL3tcqVmFgq9jHfvc8n5UuCrgd3A/pYc2IcMp8Fb0zSoUcB+NqiOktTToJp2MgpRcbWojpLkCA3yk72QzmZiaUQEBPkYfzvW1rI2Q5goI8CBsXxVrWZ1QaGZUdRNZUToqIMAzzTCXlIk/wkSW7g4KCHBJY4AtJMxvYTaS9X2oa6GbAmxFl6LlezPPQfXyDhu6jY3uqKAk1LXQTYF2k8qZ+y5Ppx2mB+xRWaugaket5udCXQf9FCjgfAQ8wrq7t8fkIocjatMDEv2qm07KbwVN74a6EkFTwAMdtZRvMHSS3MzhgJo0L96vuumkbrRCLUNdiaApYMCFkwfYloh+jvezxf4AjBjTyq+66aTqmnhxDRdpa9ExHgEXOBD+kOatV1ktPZUaSHST/L3ax83QVdoCLqoP8LIDTuabaIDqiumn0hEJg/0dYMVIDGs0rY3e0tgm6wKfHrbLi6EZ4WwRoFRLbQIKBlEaA57CAZeynR3VY5bcKCKXm8QF7pneIGoW6YA3Mb68IZ3LYS/zs2aGFHbq53k/zEDS2my2B0CjufwGPgIuntfhrlurwsw+T0e76EcSYDaQeWHYLMcREKFgSE/D92P1qlG+7TF5ETAwvVDwpfPKhssT17Je8Vi8lT04a7k096j6woV8UJCWrnxta1u70m767VT/SxdyryABLm0EsMj9143CrvcRNtIFcPnMljNo0+rSiDk2Y4ACT5PGrQ00PJy1/V6oq6CltAd84JMuvSmvNHKZftb2rWgcANvme0301EpalSOhroSG0hzwJfoP9BT9i+Mxyx0bVpc2u7w/ik/l0xvfCMY/sbTSjYN2tcetoV6WrlfQHPBcqffbnCx00bBCAyh3HgGmA3T2vTZ6KbiA0S61nq5X0BzwAcSaBCm3WMPq63vNoK1i/XA0/bKt6xMf4Ik7fK+OTlqhfESXLRrl1hm2Jqqj+1CfVoDzusUN5nfrWIhi3njPLM/KL2XTL5/LB+XnI+BPXZdw3qt1V3B0fYvNngdrr2scpanySJBu0grwp7SmXRQN5cf48PmWP/HiOlWoe5Xm9cJb262JajnCv+17dYKgMvw/10/PK1gmpwzS1yZbK8DsvVv5jrzL54UbcwOoX+IacUK+Y8/gDLc2tHnjU43mNA9rPCHUlQhMWgFOYESHyLvt2e4nhBQmWueU0CueHq62LzSvqZdF0Gr6zAnzRQ8aAf6WU+wl73/CdvFh3JZ/8zdCDj1D73HtFyNaKtHij2teLFfpfZ0KDpo0AtyLY7QakB+iO18UYErqNp04PxqUk+V3tyu6f49aQoObamqtVBaWO8nfs00vjQBPZk1ghauLrEUneMIyKq7Vinl3znDO/aWv0RhghfVgHPny2wkEawnt9fds00sjwMW0z7vLRX7u+PQbpHTfGfQ2htop5TexG3xOsvYGS6e1V23i82189GS150SONBvosLjKLK1KyWVF0a4ws8ZLtsKzB4zm1HwQCx/1X/leJSGv0ne68BRi7YwfN5OVz2S2KHe5da98QvIQ3vrCbnOy6ssIuZdegLM6Ns7g4UeBOVq6S+7bDUvfmZYc46JlhCvVPK86eCScgqiSloCv14SG0uyRBX0rzR+X+BfKa8ezzsNxN9Fai60RuLnhhiL/YnWo/cDTNboDDPW9ykKaAq5PmXXjyWyQNZtNItwgX0dFWx/RJSdasu96E5JBN5t8rwPaJcNl348X0hIwvkClY8qsgCshQli3Hz+z+Ze3KknfreGLSxN8rwPzMH/Iulvkd+85YqQl4J444tABorBTuuvPMmFcrQnNWFLqSHHvh/Dal2Rc7O9B1fJwXG9e1epvYTktPbx9GesvLQGXjGvy1Vy5S3vdeg+vTag+jKdyS5jNBltF3HQdIaN4tppJ4HsTptiQ4sljVJwcHD1oDp8YJ5yj1q3oAfiPXkwuxkCsfA+XEbKYJTan0Sb1sWjAeUGIlj3VZvi94Jb5B+nr79m6CWdOO4a6ElZpDRgNOlIkSwWu2Ni4/XcwMRyXN8AHcj57wgJ85G/NCZ+z+i6A8/URNkWMYiKqQz84o+3Q3IcjXgE7lWeljb1P9iuzRnJPiAHNA1lWj7sUwOk6CUfWjbPgXY+BDks0OIgbarC+UxOWkbYfM8YBLFFbYeOraGh8mhHsQ7n0AHwNET6r4PtMepmlPyS+hek7r9GP560j13mLNpjXnb4RpAfgxwhy6WexCsQjRsmpgh70Y/TyabxphaMg+pqNRrp0ALx2ZDpA10MjZ1UCZzUhxaMbzqhLU2yMgi1K09fyO8KlFeD7k2cxA47sCTi5u6Ms3QVcVD4pLCT3MMVWE67CFFvrcHHsCuuzuqA5JKVFJez36xcZQ5Z57eYY4t2jEeA8yunp2F1yEJYabvBS0fb1QPYMX4znlaYADCIHEqJwdMvqLm2AdKzvLkIXxiQayx8mLogeFupKoDQCvI4TKZrJt79xDxh19kuAt4ZxG9tz2baB61KS8w1andeT9i/7WjU271zg+0/RX4Zxp6cR4H2cyAXuRqnaOxKhN55XYP2PNbWH9ZgqygapN+R8Nre0m5A5fLeSz8849irXdQmCWuF0WROeDO1ouVbv4A6MSNl2RqbF6xKxar9X3rgDR6Sw7ceFd5m19G7pXGb4WhFNPPDRXpPmzG00YnXNZtyfTRmLK7h3fKaHq7NlL4ayBLhVE5LYRPcegHiPM9w6iwMuXzkTK+H5peEB8KMlG++tr93yAvcu7FZLLSv6ruz7R6jMreEv5HePG1xiGRyXWLvmepJTTAjmJ9kXTZthtYvIDFCa+DhrS3J95Q1sHP/e6/C/eP8QVoADblW50Sv0DnnD46HuAefSW7AOT15/2xNh7BDJL1h4g7Sjn2k4//R3qSD0TLvdoYKADe44Z/Lu9X1ViDPIRPEC9kvbYPJejwYrvB2ugzjg18vJ6T+f9RvwEvwVV3j6XpvY39pB/alyJ5M3P15gO13wI4E93J9N5AMfWeuy7YsuwW8HM7sAn1ft4ovAzRrVYKsW+6HMWDgZQuLFgAN+g76/Dvzlqr+Am+OvsC0MrGV/21ZRpOtfPfAx3YyxjYE8X51vn4YuUpvq5oAvLshFnUl5kX53GuclKvnsYk7V7a6vPsMnFWs9sLnNnsGvAQec9hfaO9n92s89HuoWMHNpV5MlH/ZrsHh68+fcPqN/xTdVx0xo4fzlQlZEcWWalJcpom1eQ1yDXZrl+7ABDrLM9/loXXWvBsRIy2XxFzq4ufQczEQbSa3o3dgEve95paRbwGy2lzuUreeMzXcNYkUcxaQUrrYU06oDX1qW9zag++IT8dDD/j9pH4Dpul9WAnzQQki2p44I8fSIbmi95wLhCwfvDPr8NLellp0oxANGZjKn2IpM3f3Hc8Arnqdtm8zfuVpdZJN7wOVr50ldvVg7Yj9Qg7fOju/wfXsdhzms/bXstokGedhqL2bQckLvq3DAf2fXOfiBx0N9mE36popt6OoZG7v/55FsBfb5ouSddq4GPyo8hG+2Ku7dO2okDvhVvvOSx0O9A2ZTDZ/AjzzhfMJ+t+fUe5KRPDnINp6fIqbSlQHD9XeqxwG/yIZ+8p/3eKh3wJs9P4RdCMfj+VxDA3IXN25c8Aj5Kw64dTfavitq7NkPmXfAWb5RVT6xu9LTBtNtI9rKWxMd5eIBXbJysaHmicJMHHBelZfqprwQ59nfiA/v4I3RL6Ph1f9XcQ8fo6d9t9URYU63BlL0mvIE+qYShP2WPJt0cM5cby44fTHZwdHldz773BeyJ7njlqYNY3tKLY3SkTXSJL//6LRnG+s1nsSDfPGiUzyx/VofDos0ccD3u8aleg2p4AtgpBFfdtkXwLdL3peTr8eyGQZ08/AFK6aY5+Naf3Ru6pPzJbSCD8VovsHFAae0Xl7Pa7Q3XwCj4+ePfHs2X5+k3MNJXxxDlqwgpFx8Z3QHaOLSP4TjL8HHgQ/HhVSlu4LtqpgDftlCSv7q7VBfAD/8tHoj3/gOs99NevXlWTi7JNllHeVTEGzKPMu31Qv4VB/s05GhU2GMNHcYPHHAr9G/t70d6gvgsQCtfeL7Dxd537SAptaOIbbIvUalvZwAdWRXZd/FQINH3msYUqEr2yBbCWsL+IgLbB70Y/vd7uTeedvD+HLqFPvGc+nkNgsdrodr3D7z/iMNo5AB/uFTTz2Ffx4P9QEwRm6FdwF+pw60pFXjASq6N8X4gh4ywz4Lh74NMvPrk/ARHWRTBA74piSPh/oAeDpyOrx+RdEFdWjRHy38iHnzcB+cxQXNXs7Mja3Sb4IdH0rbpSs4fv4B9mrP/o8fd7DU/HZXOD6Pu9lnlU5pvciXFnYES+O1SZf7DWPNpESPJN3pT+yTEssd2HS1c9mJUNe1AepDtz6ohfRYXVjW0S+88qt7EXfIxIwy8k/5ENKyZ0iMncJEegCepQ7rjx0z9rNBC2wdf0O3Z7xdjq1bcRfKJeKlB2B0pwT/6yvfm4+nvGCf04IFZ5lGuMWO19Ep5kk+gqaR1UkPwBmqbuAPX+Lb/7Lm/KHkRDx0RJNDZgGUlO35coVVAKoYat2KkaSLM9INVlb/VMValtX6cA7bbeblcnnTpuf5/jMiTLoAtiAWZgKd5xdgm3ez47gSJk7dTxJSSh93wjghjAsaRj70D/AHbM6lOI87wxut+lcJWaWTv+gds/nilJ5/8Y8wvHGfTABoY7HMaz9fDGUEINeAC1LTR5ZLGymt1uO7n2CtqncfP7fkjJkk3q+ByDXglTvJ0j3SRkqrBVzHG0FvYo4P/w5K77JC6uUa8JjrJHOetJHSsxtGV1dT8P3u/g1XyqqxM05Omilec9DlHvBiaSOlLx8YqMKvMyGzomvM8YPrT6Tt/wK8NJ0HyAOjRTQMK7kGvCaDLDoobaS0ykc0hj6D7JUn2/t3/7JQiONusfSQcYMvavFTI1OuARemp42znB/FNuwDM30AbDkux1NGpxoQW77bH7pTJNOHl3O2xMevKY4BXJIm5J807SaV1QJox5NsBgD2L/fr/q1bVMIWsfXHkpjr+JnuLnmwfb+77r4TIhoD3oYsJPcL4xjgA34BpvcwW2rIzK0KMLXRzRW/B+Y7XsitNAXMYqJI03vltJ9UawsoXLyr0Bd96qHLS754dh5AB3djHRPx6IO+/4bIk6aAC6MB6sgsLPv2WdCrbOyGlLqr+vkO90XpP0Xj8dfI7U1oGW1x7157EztF9KPcS9uhyuINO+zcUGCcaNZ7nu3HbQyH9oLXZUnoPx86+f4jIk76BqdE40o2CpYnzxu+9axbngrxBaaNasv/Pzz9ADBi5BXjSF/ApGRj3ZrsHnz0NzcsPSw1ZYArjPXsPWwtQBXdPZmEsXQGzFbvM3up4qoAf3WGOKqVc95AaXtECtPynkfDu0cXmbup/PoQf87TcREqnQGzxZ+9UtfhfEHmnks2iLKBjuwGT+m646/cuvJtQu5sR69oPjlm6QuSM7acjvGz1VfUtNL7ES1Da4eN6xL3C9PsO1PMQw9bSYpmd+DZQRtXAzwQE+gT0tmoOmKlM2ByoEY0N2dfS5/XjBbUqeAC8Oo4p6zJeD4b077h+RrM4G4hPa4L+0UQXgvSdJbOgIuP3ZKm/qdKkc4kvadE+Y8GK3nIP4UGS22rx5On2Nyc7ph2wekSj1N4eMtlbSay/jL+T5nndFTESl/AOZUApC7wVULOKfitZN2lSq+ynRPH6AP8oD3g/eTojC2pn9q53cAFD4ccrzFafjZLulA7eoAw8rFKX8DodQP607+250nuulNdbPzu57KXJlsE8d8ANXKH2XuZBujMN+tspaXhfgeeztsgOwHsjrmCqDvpC7gv+8ffPuUij3c39Nol3h1+/0y7OGbv/rKM03n8Q2pnt56c8EGLTHJvWJ/zzAy+NysYveNJASBxPqO16prppAdbL4e6Cg7SfyRrIE8O5jcap1anjRNPWe1vt7Pb5w2z6bQXzeIO/4YPagzHTMn70pH+c3X3+OijcHJrTqgrYS+dG1k5S+SpHrb439KVU5tmx/D3dntvblLuNVbc2vQRXUGa+Wc+Whz8aV/cHvJVpJ1oraJCXQl76d1NsgrdQE9kMe3eS/zSYneXVpa2z/EEs+Zox2LFQ/VtiuN2E2ZBe6tb4xG4O8L+AlNpVqhNeyIZMCncc437G/2a7uR1Vgxb2rrAg/DjQ/xYTTYn/ZFuU6RvxspvX0JipCwHZyaY9XlANQxc2E8wWBcteICZmgJUfbSzVrMz6M9e1ii71K8xUelxofQcxill+Cvt6FpI2ezPtxKSKx1bxaHljHm1A69hYHq4/Wqoq+Cg4AJGG7y57PFcxH3DYwQ4aGoFTFvWleTlTH14DxeY+R6bbujG7my2si2taRfH4S1sglVwe+Wsri0PBFz9cFSwAJctHH2Bj1SwOxROspUPMWy2KQZc6kQ5ThcCw/yQ1RVY8KT9laGui3glbfFr5+zC2V/eIaQcX+84CvZ49hT9nXAbScECjKGVTpMmVnqP9r0BzyWeJXtftGZ9ECUHXOKxLbeRK/SzQTbIMbHi2X1NJoM1zppS6O5hCClv/qtXlCaY5fiCv8dntWh+WYz8vyVSFCTAbFZpkGRMC/DM9qv4eiW8ce06Gu0rBYRcG7ewlOTMWM3tgE5Ug+RcHmGKBZVeVj3pW+sVcK1aLWnsTHZueaZV89W4v5gU4mYvId/hdqX/vyP8FKw7GP9hJxPy6EBjmnj/Dq4NBXggNZkWNLRixZkmflMft55bduL6xqEsVCm2q1h0aQw1ymBZfZizrnU+YTNVm3iWNVLmNkI2VoPhRIpk64tzYtMoWICXA9Rk03r3mkLbR1LEXzQDSKNflPS2Am55efgcPgxi9ZtTwN/Ri+X9+sBjaC2wO4qFpi7gI9OSKTzzOP4yrkG74DEAABWMSURBVDK21YNe7tOIGrgOWiu6xL5tU1JPHtS7dY4QNnAB2KZiAQqZyztruFhpOsrqxbB0/oDvvgBYela6k7nyqgI0KiYFteCVDCkrBw/4+LG987S8+ySiFOR+sEJ3FdbOebTTtKyI5J3g1lcYT/xl65dTpHtbce4Odruui69t8/x46tM/Yh9bcZBlFfosTtW00uGn0AG2VxbeisVfT8k+irMJu+t1t7kSRgeYz9OGs+JOnK1sSjHxhYig9Df8GcthoSGze7Y9rGPljSyjAGaqwSnVdxjDyFu+C1ehtrPlYBu8mt2rdCU/lbmqLTzEbEB4DoZ/ZL1g3/zGm07GALwvKX4Dx8YJbx59zP4A7M1+NG6Dlel33YfYv9NZcBaIxkY1LkdcQGTAaHfPvEHM1q32hpYhALO+0qVyu0nErXZH9OeZ0mSCZckX68n5Wbaxx9MHymi/a+Q+Ni+MjsOhbtwCZlswCnNYz2qvXrU3tgwB+DACWJ6u5AuNLUdP2o4oHpbAJpn4Hh6JPatx0rf0bRtn82XYVVHKLJ61ThHQNMJkCMDMH94lfAxXuy2PZnapA9BcedBgG2CcqEDj+Bi+y7pDy6wHnlEA/s+QiOr0upAhAJP9SfEbCcbj+e3wQZxM1TX4qbiHyYNovg487yqbhHjTBpgFPV1sOzJnxXHMGcMK+kq3SoeHjAGY6Rqb/8V54m5lt8lO9ti2O+AG9pwXUa5HoyAusyIbgmT6BKDSeZ78/vOB2IQ+VTtuahlrZnXQt9KGl4EA87ZWzdtzd+JOyXPYnHY+iLecaGuq9Kytq7yvB0A87uLUU2VsauUn0Ps9BTzE+IgQGQkwG5kYIu/hK9k5ZA6bS+pn3Z0bm4RtaWYE8CVh9zfAdzQxEhO3WlWPj2vkwsbi/sWIeTUbCTA+ov9gXSo6n+6tcT4IjedlX00kEzGWSLORQ4k0v4BDXszrvIUVkuBUxlSA6p5D6ZpHRgKMUw6K6BpHZrgMxLBpumQ6e2hLEZtCQkONNtKW0DZa/OQiHoOtPSGtbE1vm0oxc7oOP8CIMhLgsrHJ6T4EWZHUDqAK89KEr1zLmkkSdjT7ak7yCk8PX1pOyMd071XHM9m4R6Q4oTYSYFViJtbz1iU1OG2fj5Z5UB/Y9D7vMVd1Ohfv60jxnRcWgK8eK3fKY+6k31/mlD9CHuNgMxY9sM1d0gyq2Q1UWjbMziEPkiD6O71qbByFA+DBALHObSIeR36TdT/3FBqIbLYOYrVG71lF6bUmlE129SJm7A31TNJHYQCY+TJ04aaDDUl2lfc2AFS8QkgUcLsulLyUmLWonZ4B9V1iN53CAHCuu0YvmrrPUu505OMg70uAe0lfoT+1Rk5nL4WIiIgXBoCZcwdX1uqnEqCr9cZEop/wReKTJcBfyt/t7jLeRbd342dzXNjPm03hAJjsXuU9djva4b7X+g75Zup53pKGekWWSYm9cEHillFKg53iGQNOuivFfAoLwG6luAO5SUdDvoMr1hb3n4zmei34sKXCEwQaYZ8gkaKwBtxb4Sm8L38s852yVbPQSx5zqEZOYuzLOmSS5JiJDWsODE2FQ6AwA7zJ6pqDcOta2cidjFXewURy8kBVm9nHQztml1UDZyYiaaAy3AAXI5zt8h62g61LGx69Qncq3rEeyixrO0G9DozvG1kZ1uGPjQAN3HugNpvCC3AWMpoi76GpTtVS2lFmVrIlw1stkb8p6vg0HrmW3G3EuZ68vZUnVuHXkRSLNrwAW6qBtLqF6cqA4fcJmU6fwyw42oH2/aUHNrPfQgMftsAc/mshGgAx1hHnkDa8AJPbvds6urpjK0PR7weObMWQrKaxqZZkxjU5uuHHLPEc8+vRYkNMzIJQVDqkCjPAzrJg9A6YSHhPGI7i6ojZo6UWFg9iDL9ngJuEuqohUdgD5kay6HFno0wVPi0b9jooxZaaOnm5fDBtnvnfxmEPmHV/eOuql8yTdo4H2gG2kPOrbxFmMW9torEx7ljTD1aGPWBC+0G1JUx7JZ75bHIJ3nhOBiwta2CjXdaZfifLa1Mq/AGTc9ZxxxL6qn0KoAemv4yrd36+9RZmjecS1nG2uvXYxXBfmLDe1BaWJgCsUNGyZQ+3nSDFs9IvElL+1WcVANr8GTmiqTXzGgGVrK9dS1uAIeik3NT+4c0FWBK6Y7rERqfH3mAmAM8WSAFgEgfdUhx3N5fZdZh62t+MgJmFwFhmcJfMPUDgwn8W7nKU47HDBGCrwgYws3teSHDhCobp2DeRt6QagItl/g9jXJrXm0dmBIwLWBqXkrMxkKSMLVw6f9QVp0PLOkCUqaOXmhIwKStw/115vnJvjnhEWxUmgB8dz/f09SaAGgoDIOaC7cLqLPcnhLlMBxj7PVYr96NbsVM0V7FmkTg0tXClC1phZgSxikGV6QCjOVailP4MoPI3E1lUAOt7Fnd6KI7f0mowWtymBLeWwZPpANcEa8zhPMVotNUWAG1wHSKGJ4BL/8TmkOkA49qVpTwpe5tF6w5reLzyRcMcVyTtcDXVZBaZDjC5uOB7OSlNLyUM6HmubGaHpW5PubvHvD7CzQdYoQIOGCcj+tFtRy9RTN3obBJ0d17dGC4yNWC20ntYCRps8RggnqPFuxE6C5mhcc2CJ9MCvjuo2yl8RldeBNC8jLAYIC6cuvggPLGT1tULmswK2ILBDq+QUxlFyKdiyWXmhKumP0Vhxyt8h6vNA9hyWvmKvYI8cV0Lfw9PvNENCe9RVeKuWvXR6e39z2uFsTGmaQAXxdt5d2evXwwMTljwpc5V6Mc8dRGTrkvWP+Et0wD+2gHHChZjiarw3/QL5iBtsroSV+E5YR8uzTSAWSxxl7dowfA2m9iSwgx1JZ7Hc8I+hIdpAKNbpebuv94WpT48x5LouK3ej9JAm1J361a2aQCTvMXb9TaPfKSPFTWufF3s/TD/ZB7Aviof536zZ8dAjctejmQBAh53j+nCX+0l9XXiEOtvD90XRRzguQAJj3kMRc/Rhi9GQ9x1Pp7dST7ThTcmqhtfBxazB5fKtQioBA8yMeCbJS4ymTviKTykkuffXo+/1HGgMpbuXh/MApW78AaD7uVHOGf7rpOVoKpucaVNC/hulMuhZ+YjYBifZurm/uyb3zxEnvHWO/i+NDHl4thOgZt1PfB+iL8yLWA0e6/mIr81Gs9mVwSoNJ7d4TtX3HRum2FQ2u58juFx99jO9B28lgOu6KLEnm6uZAyZFnB7N/eVZc0U2soqPi51mZlDafsRzJJS1uxpuqKXwvfSEcl44DZx0g36v2WFdhXXWKYFjG6Uejvk5dSCWrfsch4yajHKrH4AoxBwY/tzpfWoKTucL1VyQsdHbKAyLWByZsxGx6xWTs3VVU6tLbTKhC/pn2PLuMwyFT2EwGXtq6qnzAvYheQWsU08MnUvRQ7zE3Dg8ro79qdmDxqYRY7id0uIj7q/MzuQymqkiAKMztHS7HKW05w/rlS2snDlWmWnwAJs5VrWRPw86/idG2HvyQCv5ogCXD692TSH0cb02NoOnpVupY92nmHYBVY59q4tY6t/5uod3Nzx5R4aRRRg/3XJBtjR09ZM5wYZU1MBOJw0DkBy3BPFlyhe/54+2CdCtVUI0tWQyhE172v9JAD7KssC6RZmt2tfgJolrMXNlpCDi3BAdzY4r1YNvgRghXLqQJ0cd1+ulp/RuDDmDia6Mqcumcwi9xIhDyaMU2cTFBQJwAphP7mluy8Z3Bb0byq51YNb4f4HPy6cw3Z3OSmsDFDRu2f6YEsAVoj3ky0Do5MUQfWux7FpQw540Jo+63loF6Z3/0A/lk6t1/EqD9Pjn2W9rhKAFcKIHqlkCdhFS2sqvXZx8oF7I78DzppwkG2MFypAAFaobEqTKWU8OICtt4u3dRzdnnqRJprF9Cvhc8rT0Bz3NzbCo/CjXejq7k4CsJO2g92873hAn0yEoBvbvwAL9/DNRxBza1h8Yqc/2AAvo38fG9DzpQDsrMVJnRRzTpZ2f6yJI1VWlOPITfr5grT38TS2qTg6+eOWGOD27NSmMd0N1NaKSMCPZ4y/5f0oSYuBx03DJ3cVBvO+zQcmwJ8Auu8Z2g/bXTiTyENDdNGt6qoViYAt+Fq96/04Low9DQkZpHxBj6W3cUkbXN4hwX1D2uIyCqrOhAdlAvjw82FGmSKORMAXkMFCX4+eyCE+JIU1aWuLpl63oHeXwbHQ6mbf1k3wO0s0OwTf1G3kOztaxx+gRpEImMWW9qHHeoe9Sku7MGCHMPgHE84glMjms9/RjLc54CY4XImzEv+piLsGcb0ViYCxYdzW6yqI8ibYnkJdRV753G8pdpWkA250aba8IRvz2IwzSnwtIyn5vgCdrbkcnQ6FIhIwKcr1fsw6/mBGZSTXOyo92QGqSnYalirWhlbcbd665jo1blt3gJX61Fy1IhOwL2ILGa4pc6Yymk238Zs/S9GUtjwPVu9bO2lyaNBr61YCsDvhgKTDL87lOFthuIf2JQrAzOCjAr+1W7B2NephaqdvHUsNugRgt8qZtKCYkDXVWFAmcn/P+n5L1nOebJnDgowY6J/8bwVm2PKgfNPyziBbciSCAQanBWDPYobTR+WALr14zEtmwFOLkOvvymifkbYv0D4wbUN/g6cyR3vpof4BArBnsUg8i/gKQHz04n4SW+A0gHRW3LyvKtIpfLmpBdPu3esFSQKwZzGW1wlhAfQgni1EgpzeAP96wNegwgfgpPH83G0AnUIeskcA9qLLHVtm0g2LTlspszyREUw7t6eYfMuST46zcv2lNWWg+NMCsI+6vuZaISFDJIJoTlfyrKtb+O/s84yHmALBlQDsTZv7rrA9Z2NkjrHRNY9ES6/c/yj4Ns/vyhN1DT2SVZCaPrJc2hxoP2QIMzWMPMBnW735fgWwhj6kqqO8WaWxy1fetOb0IVeqvialp4aw4gq5BrxyJ1m6R9rskF0JRRzgAuvtSvYNWmmZEJdy6sLfFIBvL+db2x38BRsK+Xtb/BwU6upzuQY85jrJnCdt1g4YMaWUkI1DG7vyX2BmHZK5peC4NEaIr0Sb1bcPH6jMcn/3QLKubGEFfGkLfr5HcBLxlPcLBEMuAK/uvxXJLuaAF9+9S5ZvImT/nPaRBvgug/ZbqPZ9Q5lgLiFHmzbf2x8k4qg/8ZZXs+PrH7LZQ6hPSr4acjrUtZfk+g5ek0EWHZQ2x26SjcyXQcQ9osmqavD257nFhGDDCYPv/C0mdiEi/Bg/rHHkmf7YZlD1ljfREcC/DLW+wTXgwvS0cZbzo9jmUt9hqY8xM/IAW3UnFiDz68RaSqIDWIBpZjj7dzmvJiHlFpLXOaZfMTtv3SD9fBT6KNFN8k3MuGOrEvD2UjSaZSNcZyvJmUcJ2VX7TyBF4kJDejZWaVnSYVaITGoFYBXKQYYYeAnejKO9oKzKkJA1hradFVaW1ZbxbQLa22IiGU9Ea58BoamzAKxGB+rUPcRv4xW14kbWju5ZxpYB85XDzyvv74GStw9oi+eh0Xyst8L1kQCsVpYkBceY2VYzyn7rlYCzJQ/WH7E4Ax1pql5oqisAq1aRXesZG1h/pX/Ny8l27Dglz38Ls2lL+gZYPeDdTYH4EK0GF4DV6+wn9ffYQX59/DZuxbez7ZCH5FE1vrQBTS27hramRAD2V6frfdS3KzaX35PbVDaV7DyOm5qhe/EqJAD7o8td2tYGqHE/GapO5jexzR3w1V69+A4uKA59UFMBWL02N5AezRtJGSnmQfOs0S1ZTNs8TD2sC4mXQ1ZJWQKwah20vnvZCG7JhhiACeybnOWH+fJCFx5LQyUBWLWGWQGflHK2TcWVwSwOz5CL+MXFEFbPQQKwaq2yAubLgJfiOMcuInk2JV8C/BsSHf3hhUwCsHr1hbhvW1gB8yVL6MVjEAPMDbfov5Vl/TiPmE/WjBmrf2UFYD+F9rPnMSE5SEtqcOR2JYA5NKcJB42taE/hWPCoTbrXUwD2V5e38ZazbRHa/eIjzDMEBkr8tHSE7SHuUmzhQ5pz/pSoGgc1rKYAHLB2J0qAM4pmDb+MOYubdd8xmGUN9nAeGvw4x7tla2Q0nFoUgDXQqii0vYSbKcAtpjEWz9OYE+PJU8fVT2q48EY7Cc+7rl3dBOAAta/XNLTeWBETs5YtOsQe8SPplm7gR3nowbaShgteBODAhLMOcpwPtmIYoxuy12sFaHr3245DVXvbWVez+QUNKygAB6ZuUteIaQFAE/b67ApQ+SFfmhgfuroxCcCB6Hr7d5SASans4u7olkL6yfxX+uAORE8JwIGoijTn4EbMP1qIF5AKwAHoNm9LJbv73tIl9B6kBeAAVC61lktDXREPEoAD0T4WUzjA2LL6SgAOTOVdofqRUFfCkwRgk0sANrkEYJNLADa5BGCTSwAOFxX2qT6wWP1pAnC4CC2A+qk/TQAOF6GBvR8rYQTgcFF78GstmwAcLsptA+3z1Z8mAJtcArDJJQCbXAKwySUAm1wCsDG1pnqNvZoUJAAbUhesfgIClQBsSDFP1J6WJvosAdiQYpGZ/BjWcJYAbExtSkjRZhGpAGxyCcAmlwBscgnAJpcAbHIJwCaXAGxyCcAmlwBscgnAJpcAbHIJwCaXAGxyCcAmlwBscgnAJpcKwEverOVKNVNcZuukpGBeLLhXS9H4agvVAiaLp7pStX4us3XSu6ODeLHR/w7ixab2i9K2vIOqAbtWPS39k3vVf24H8WK3PgjixcgBf7wPe1XggGcENepiuia2pD4qd3gQL0YuzdSjVEN7aRMKXAKwyRUI4ILU9JHl0iarf2rfh95P0eRaUjpIV9P7l9lfzTIlbeBjLQsPBPDKnWTpHmlz5Q6ZdkyzWnm+lpQO0tX0/mX2V8tcQK5o2swIBPCY6yRznrzJ6pmqp+tr5bWkdJCupvcvs7/aoukTJ/nhLMm9/Aa8uv9WrNFiXjEMRbEgQ7tqOUl5LfmCwbka0fmX2V9t/hqyStNwaIHcwWsyyKKD0mbJCbJhnWa18nwtKR2kq+n9y+yvtns12arp1QIBXJieNs5yfhTb3O6fNrBAs1p5vhb70PFawf1l9lcrGzN8sKZXE90kk0sANrkEYJNLADa5BGCTK2IB3/zha6+9Wuc+yWv361feWUX3K1cOdZV0UeQCfoqQ8lbdSNUuJeTIc4fIx6MEYFMJAZPlcYdewnCh2RaSt1sANpUQcGmDoXPqyhkCsLl084dvv/1Wl6KvUuQMAdhcYo9oQjJfxLmbzJsCsNkkASaxzfLJ4d9lCsBmkwz4cYfnXv5gJ7n0xh+efGNoaKukiyIWcKRIADa5BGCTSwA2uQRgk0sANrkEYJPr/wBcLvP963ndLwAAAABJRU5ErkJggg==\n"
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "pca %>% rename (PC2 = V4, PC3 = V5) %>%\n",
        "        ggplot(aes(x = PC2, y = PC3)) +\n",
        "        geom_point() +\n",
        "        labs(title = \"PC2 vs. PC3\") + theme_classic()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 497
        },
        "id": "uWMhmddf8YCC",
        "outputId": "9ecdf0e6-5be9-4f18-c4a5-abfcf1efeb39"
      },
      "execution_count": 65,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAMAAABKCk6nAAAC/VBMVEUAAAABAQEDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////FXECgAAAgAElEQVR4nO19B3Tcxrnu3LybvMR5cRznpjmJ+7VvXK+7Yye2SErsEot6771bkmX1QvXee5csyapUsapVrEpKsnrvoipV2evunIf/n8EC2AV2l9uBxXcOiQEwC8zuhxnM/JVQE4YGCXYDTPgXJsEGh0mwwWESbHBUiOBn//MXv/hDrdtCKf1fv37q3WlCIeOzZ57rX8HPbH//ty8M87TBJiqGihG8htK7VSMpnf7U5Nv56/7ch+b8brr17LOrKvSZu79aS888s97LhptwDxUmmB75j4KCpyfB7uF19P4cYZuYBnufjhH+vTbf0vmlV97e7Owzt5cL2wizCwcGFSc482fFP5InsqPZ/3UANlM/Eoh8Km/LC4V0b0MXn7FsfvaYF4024T4qTHB2YjJd+yvZweyP++L24c+v0a8a0uO/nnrb1WfW/ew38zxusYkKoWIE/+rpp//e4hHdS+7bjh17eTQvVRtp/esOSnfVeOZ/v3f+GWr56bWp3rTahNuo+BAtoOh3+NY93EUYlP+yUTy74sPdz1uhUDb3lwVOPnN6pbAdEONNq024DY8IpnN/PvR6/to/jqWFL22wnS1+ppYwWM9rVkqP/byQrj2u9ZmDv9xOb7/XyweNN+EanhFMN/7rqV9/IvTEFeQXAuqwgy3+4yKluQ3+9tJbqyl9f5jWZ+jiV3/9p9aFPmi8CdcwJVkGh0mwwWESbHCYBBscJsEGh0mwwWESbHCYBBscFSC4OM9/zTDhL1SA4DlJ/muGCX/BJNjgMAk2OEyCDQ6TYIPDJNjgMAk2OEyCDQ6TYIPDJNjgMAk2OEyCDQ6TYINDneCCoSNGW/gmo31aWjYcNAnWI9QJXrOLrtjLNzu28YMmwXqEOsHjsuiRRXyzfsCo6WWUnt3Wq1qA22bCB9AmeDnfPHhAV22mdGmbaokBbpsJH0Cd4HU76bJMvjl2h25CByQDDdGPp04Db+Wb1UjKnWC3xc9QJ7hoxLAJ1otjcHOl7/Ch6EhkHIILIwiJLKK0ISGkRbAb42eE5TJpl0As2UNpvLBJCHZj/IywJPgUEHyO0uHCZkKwG+NnhBvBPxDSupz2JqS/sGOZ326JNdgt8jPCjWDou4spzS9wXdUYCDOCi4DgIcFuRSARPgTnLFxZSmkNgeCwiuAUNgTnVCKkmpXmj/0qUzq4tTJ5e5ixg0mEDcFr+MxZjlI4RtoJC+OBKZMswWmXvxE2BG8HLm8pj91CgklheUfh/8TgtMvfCBuCLY0JGWB3zFoF+P0jozk5KM3yO8KGYEqzeKi9a1I/vksk9AtSs/yMMCKYwSr05O62vaqc3VTS2qAr47AjeA/weVncOxOP/Brim6kj7AjeBIQel/Zh9/Mi2+617CC0yZ8IO4ILowipKlsSNWaySwZLfSakNhDCjmBasHxNqbA5UoU0Lha2Ob1SZsPhshvC0S3Qn41lAhB+BHNEC1SOl3bP4MC9Cgi+GLxG+QHhRXBJua0IVHaUztQRdqvTXOF/irEUiGFFcDdCponlpgKX26VT1ZhxR+6SNWXBaJn/EE4E4wpJTCxQNL69jF+6Uj7XMhLCieDvgOAzfKdwfKNOW2Unf5p8iNLTU/cGo2H+RDgRfF/gt4r4Fq4PbKNweu9qMeHPduXEyxAIJ4Jp1tDxObxoZSLKONK4vbBZIBwpXLWpnlCMDWYD/YCwIlgO5Pe/RE3DQVoQRcjLBjSjDVeCfyAKjKDrYfMJmksbCkYmeFtiVTaNyu6QON3u3BcCnZ1LWwr/mwGzm+g2VEJ8FVv3WuAb6k8YmOC7NrkjWNotFw/jOjcfzjWnRWNbr7MOeT1yJKXltYQjlYz3EjYwwTuARHRuRo1gdqu4YRZ6LIpUfSjMnIk4YS7/l1DCXluJD9fBbLPvYWCCs4Gte1ACMRV5iYAcK1H4/zWl1+HIeTg3DErDocT5fXqBocwsDUww3ZOc/CMWrn3KyWuELNYXDs0QrexQ4z8FSi+JM64YI0krjUzww8akFYtSv5JTN4X2Ev6vk1f6Sjjw13wooQXeZ/DPSJbxRia4hcBVNyyVCIN09+ZxgyzUsrT3LkWl3KYk6S6WOqOCCf7NDHhT/QcjExwncBXPitZTN6ilA59z2WA5dEbauTkSuB0Lgi2yO4Ct9DOMTHB3gaqh0i7qGvbmirtXp6wV+nV7LB9p3GI3G8Tvl8D/dzcJB0sakyoHA91mn8PIBBcNShwpmy+NZBSegHLprBb8tXydysyjVxWKk+l9lE4WNpWD1HTfwcgE2+GEOJMW0NYmpAQL2vXiDiyQjsNKCvRM3WCre/OOMCKYZiKJ9aBo45ekbiynF6FQi5BZrB7sLad0t/g06BrhRDBNAOoOQCmG2NZEJOUJXVQlJp0+KhamXXDyUDzpCT13Z8cJ+pd5GI/g4pWrSzRObSbkvVNYEjrza8/ZenFbPhBfjSXxWYFpZcBgOIJLogmJKxX3Hgzo/JPtTHWht4qOC6W1iQz/rBzRcC2ldYVi8u7igLfZnzAcwT/ahmEAWD+Lds4L4cyX3MesOmP26VcklrcwsSWJeByEZvsNhiP4sG0pJOA27IiWshORPibHus9JXUTPRYgEd6XjWWGa6oV1CsMRTFuJ0gvKYzQwpX/+9N7I3nl6feM9Wsio/Gx9fOKG2Ulspxm1rgLJNJkatLb7AcYjmF67IZU3ENILC9ZYgTnhbxhdi2KMscjpM/DvIO/BVaCe8BREPHLrNgXz5+oh364BCWbYk1TtB9nuZaBwSVkZ0w9GpL5DHMBMOdKnuucJXlKFkEgdrKKMSvADoOy2bffHHrBfGwSXCY7UMnSBip0ISdRaZSmQAR/Z7rpesGFUgnEyvUncE2WRYKQzyJHad4bcbS1sls/uhgoJYcG0rGorFwvi81DziJ+/hQ+gU4Lza5DKh5xVwGmyLdpKQ85kQ6E8wZHgOCaefJ3vrkIppSvju168z4c4dErwEFcMWICpo+Jed04dCJvPOBLct3zbZ9Je5QKaBltX7+JHD33xTfwNnRLcBhhwVuEGVBCNoccJ5d8Lf+/unTNj0tWDL0pk/lfsivm1KtX5lJA/kBpMQF29BCffJCoAX8MrHP7OnVgEOiV4q8BAB2cVyoGraUxkeQeJ+4wkf42F1BsSweNKmbQLUHszWmVFwEdGR6eeCsDX8AajhLaedl1NpwTTXT3mOLd9PJryd0JicB1zEvl7OYkLNMjzstE58tECW7k63Qib0HZeOTVuCytAU7u6rq9Xgl0D++kaKJV94PjaJeT/sc0bskPWKPi/il/gfg2SejOIX0AVsDpgERmhqb1cf8C4BN+EX2AtlDBFw68U7H40RbH7G/zfHKUhRBRkgdVHgyB+AVWgpRGWlhNS6baL2tRgBG8jpLEUAqsdIfGo+osUfpM/nlD2387K3TeB9JhOUFwmfh5Md+KC8C1kKL1QZHcEphF8+XD/mDsSGSMRjEujN6QoSCcOgDv/4Za/hePxSkZ/r9ytfPwK2OR9IiyVbU8ILJgH+bqNuW50OhvAlGi/8lB2ZVEhxnZzqQvomuDdcbGyKM8WpuOtr6xzXUHk/xGWQ6rgDkw/3XyY1o0rG60zK/cop77FNLYMcxMQCcghn+AdaXIJIZJHu7iGnglGE2bpCd/PqKqqrPStOqF26Di0c3MsTIFF0xX85BPi87xoxXCLb92uXsvFSwLjbroQt+iZ4Gvw/SQ3kz2MK0UwpHVVRAr/QyzIbDgArfH/l9D1bcdm4GfRAuQ+9SUewyXdVzfDan+Ok/NLie1p1ISeCbaAMcZZ226ZMMh+OVy+ip0Zp95fRdE0LjP6iTuj+4olZiKAP98T6lPAQ+TyLfykEamHbq/07OzDWpUsFx7TKwQTjTiFngmm19rW3SnbLd/5o+LbLtEakP8qFur2zuaZOQSsvi78m7iWkJi6o2D2XSTMy/r6uMXlq2fcc1kJZvhNXdTJEV4lNYVZ4RJXb3RdE4wozdSKqtHSkdo/2h+ILT/OS12sNGfr+RNzL8LsGWPCW45cDdzXkAE01vDqfXhLuw43HyPfu7qYrggubU6q2C0baE5lsMMRcX2/zOh1ij2bahBt7thkpotQ+pQEef07WmhAb4w9UF8z188Q3myXMzZdETwDvpLdselwTBynhM5XiVk+n5m0g5b3IK+yn+Fn7lC9nq+k/0yCLMEqn1xnTCl96LSDwusXposu7cd0RfA38J3tHupJcIzbRmH4uhFQ2k245+jAp1k//R97Ot+GiTP5d0fZIQsLvvNuExLXcXQODTJwjbBU2j+3WyHVurtgR+nyWa7tA3VF8D7CfcdkgFArX/EydkDkFYNffd5P+ElK2jnrtf3Bc/RvfGcfpTBgb8TuoTpI36xN2tgLD/2GFKEV3Ab//vRvBxPP1my6Ipju7jjWwc4ib12GrQz2VrfoqcVXvmKUoVTrc2cMN7k/ssvb4s5JWjSh7TE+7Ktp08HfJc1/X0+JsnEdLrHSA96+sR5cRVcEZ++466LGiQ25dDH0c/6TCMM2M7KT3sL/+WsFw5VlZXGyhmrhWNLJQVAJx2t5/TX2texty+1yc9Awvm4aHV/rgqIavHxYiAlxvTfGg5vpiWCIXrbVVaU7crPYd232dp8SN7CEX8PKdU3z7C8OlkKzvP0aoL2MpqU43c8RyhE46EPyzLebjJEtAqAFnbG0kbfPk5Q/eiIYlAlVXVVqLSPsmb7n1vBuq0nqn/j2A0JaSj3Wiq/znvILW45dpwUDUqd77fOPPlKDMB4bS/OCrnJcjNaEjcpl9xnBLEcbKBWGffDMx65lJI4wBMFPVk2FUBs0d9LI28wuB+00GhLniLoWZeOXzM1La75edk0Q9O+U7RfGETLQtvfD1PMef41N4i0zuS0Ctn2zeLSeBWPhxjxCNQnXlmXn/VvYed+Du4UuwTkP7I+ANmGzY0UrmlieE9aPoFmA/vFKwjJ4FtRpfSryM2at85udwipK2D6HksuVICaWvQCeDGiiWITie1B0RoIprb3ExX30I1VmwdXgeRpGyCR2P1sDf2Ax2frTS7AnCqP/Ajse3CxkCU4jpJX9aHhvMxfUl8v13DvY77KAWTw/w/YOjZpp6ymaKJdJMzH0isySvdhOyDsXzovLTii38ubblcLcDhUZFv4lV9tasphdvz37PqJy7GOh/LYHdwpVgjF38y6Nk+sISZWWoyv4D3M/W8bdrLJqLvklN60DbOXqYBUteQb3tV8PPZJzCnW/9ur75UwarXyj4mN6C18tubQngTdzKazKr/Pzd98lbzmRTWsiVAk+Dd90jcZJOCdpVR9xio7RETLy+OzqdaKN33ws2/npeDxpazOWQEtbZQq0/DVS/LvtwgvckxmPE1iFwWQYs+DeTK3rx4DJ87VOzZ3657iDUCXYAjZUGuJCNGqXZfNegD9LZKktMAPgG9cd2IZn/kRiTrJrrYmOXkR5xH8nKXaKb2iqASqCy4tOSDsPcuhZ/qT6EKFKMC1aNEszVgZY18iM+mfDr5IAMpDSRRMOvstIW1pPSeL/dUEyC2p5j/fc/ErCE+N3oSQMy/Ok3XUxMe8JR7qM7erDCIohS7A2LOvnTB8pX6VgJMJ+4ozsMCNszpOjRztrkKmKOy0SJ9EDUNogXOXB2AnuOfp7A5B8SjJvHJ9HT7oNCzSFX6p1cEzNYwuXeRb9Rz8EZ07jaZ2bEJljKALViCyjVc5lS2YXxlhcveod1ZhUx3+DbH85vs99/HbVBnwTKY3Pbrj3FmZhMlBeDdZ8zxIS61Gcct0QPI9wP4U8+AEmKU/G4i9lwd/ixYZ7GqhS+Bstbv/xqljq1IeoOPys6TDLLzHgcysRmTBlP9jngzQSNvMz10sBQHry1nnkbq4bgmHSlQwFfMLnK082Eg59TkivpeoM/uG9/9XgFj0LMQDaK6A4XhRHHFNjgYjDuzWRBsAqI9K2By35HCbNYwhp2Y0PJBnfFxTS73ljL3lyE50QbEX9H7OzmE5Ikux9lNFlVO6tZHvPBQWG0n86I3hHorh7sInwr5PdvWvhQ+AHdJJfGPImkhqLsikszdECeDyvUbNoQdV2jewk425DJwQzOU+P+HYgv3x8sGM9m1ARbOagZ2c6svfpXwh5DUsTtaj/EP41tw3p0+/WIg3sJ+8wQqb66ovcqEFairF5wH6hMy9bxCag+Ao1JlPY24i/j257okqiQSJ4ckychpDKmqUeOGE0PtKEG3SAnE8MzzAczuQLK1NbxDoyUFghdVk68CS+tx3xjKxcDcbutpcjCRvDG6nd+0lNEue5aoFhzwQuU8bhoA6fxh1NWy4up28pmgjTxGcf0QK254aTqDaCQTBGHlO15z3/KSHfqZ2AfprYBD73NnlhF2zH8TMwxcSw3feGfPNv/D3+epUW3hJWwzWLVdVJv2JhSN/D/78BYfQ5YVBgjxDxkz/wTCK6HbM2NHGoYZE3sTKu7M8zXQghPznUrgCCQHAxSgjVPC7m4/dR/dC+tgMfgRkseAVCklCy9Vi/GfAmtrSTwlWVn8IQwVUolbt1q6NSI1vxD9PO9BiKYZF8nz34SZuEroXSDJFizg81g689cR/+N2/PP8bjCzlSaM3l/XNbfxE7CioUTu2vGVWiP0TdVEcQCJ4KrY9QE/Wx76cpBCzvF98Ua2yMI8OhT9sb4DF3NEJyLrjkl5CqEYrdCFgHkyG++YoS4G3aB+3nGuF+4ZB38QggR2YSJMyoU/PS15xhK/defAEPpgCwBSUpJGA843B9BLrdachlAkXwJCk0Aqw11af8+KXsJ7EMB1t0w+ETp8MwIqOpVUwyGNztm3hEcYWPdztlVrSgTVQeZuocz7/hvgVqH2brrqPCAyR+4/Kp9SD0Cy1K4Qt7AGqD4QeyrocpQhwtwIn0HH5mJGXv4xHqt8bXlEZAlgARfEz2jB22Pc7UelLhdgJOI7NVTWJA9x0JXRsGu2hsDeelBEsreb2MWCIFctfAaVsQlqSIWlL4jrgtiQlbPP+GA4VLqAyhHYTDgwVWbzmMSziQiTsoplwIpWlQasiU/ajPgpVcJrf5Xmx/EenTURqymAARjLnlRC3JT31nsSkWJCTrLa92Jl1joEF1OyRIwW5XTaglrv4vIOeypilCuTf6zhnV9ZmvCsck9Vu7CbhCd8fD+d1j+6vNJ7PZ0swWmh5UI/jd0TYrC21GyI80vxGJbVgdTYnWE9JUyx/9Qs80LZvpABB8dMRqK0TmjuBfdENMDJ8R4OrWLWtuTAcL2h1mYPgN6xrsJ4IXXLQk+ZCbQX8ybgVxhqX0YSoRzaIdp7YVAVxhgOtqIhrjLZvZ9i0bvwXl6E/oz9yIO5dZke6aYh33QwPI2+V+VQ8J3oOP9vE2XbltAlpxf4XPIkoWb9JHK/epfG5DszSZYc4IUoWZY2XBZxpSDPj8QbW6p/jwH2FzdP/CKaVKfAqDSlkTtjPXo68nQhijIrVlEflDGq+S7d4c9ZZwwzeW8h6ZbaMOXtmRs4Xd3BicGNeEho0S5la39roX5dgB/icYXUek3VJ0JmIjcw6Yu40E+5dPJtm3H1YKddSuB4GBhTHrYCSJ5gvEIrjgBCz2JnHvV4BgQp75uP1jarm3pEniRC/tYZ+cdRLQAzpsuriTzWPrcTv2gqo4hziUmrgRx4HWxT1fJKmXLkGPnstqHgNj/qseNctfBJ+cIRqbgGw12na8OJY1+d/s7sQWd85eGKied2xdTOW3sQODwYao1kP1A65g1yrIe+cTtzhuRn0dasURcB+bRRdfA33Ib4tPPPN7uwhyl80Yc+AVkgDCr4VoBowzha80ru3izu5XrQjB8KqczYqP4+XZPLfwXzUF92BCJC7v7eLjz1Q8Fxy3ed1CXPCKwix8hvBVLjfKIqTx7ueJe+i9YTUz39hJSEetdXhGNOng6aMANxHfpVv4Ok2UdgyEnTK0UxgCy+a1Nt8MOJsOha7wr7P6pV3d2f2qFSG4uvgFTh610ie2jpg5DqM1/558zFaFWTVICzFAt93vWtaOxB+ldhAtYQvK4L8UQunk1oezR1zjGoemDb9ktWwxN9xBZVQBQGmZ/V054NwC938COXCNw+fYM8U7inNuMMRqVYI+Ujh/aGprN7ya8ZnOhsXTWa2rO4WfCIZ3DkR4ai48uJy52726QFjf94SXDL317SGas4cNsZZ5rdPqMjsZB+SeULyaRaVQg1zoqy+1RNvzy0mkRRGFkewGc//cB/Ow51r042HeP9T2/8bBUczsvuKKlUkThqq1hJv69XH/J1AArOq5DR/0z5dqye9zZfxqyy7WCAgxMaSIt4gJNzPq1j9Nyzct8zA6tZ8IRlnhAGa6zcbnskjW6rijZzAPKCh4pVzr6kMfTMBlZoe58Pl/7xsj/D9Er/xL2IARD6yShqJJ9DRhtjqy/S6mNygWoxzEzL8hCkXsgSuqJVHibkIBTSbaRo0gK1Sb7ruDgrEdxdcUrHjrAsvx8gpi6pd3Sf0CfP80aNgbKb3w3VVK70dpKGFcw1+TLKT2NAaI3CEdAIK/v0lFYzj2Fsqzl8EUnODvY3A+qSsdR5XaPAxJV5PpSkfwGRablq7GWqeOYGyOcublC3ivtWj0gvilVBRuEJX30Obj8kXV8UM7HqAayBnccpvWuQrgchSJugQxRRXrbmsT1oSPkzay/s4ljyAh2sZCPXl2O38RjJKJHRZhHZfApDUoU2UTqg2izTJE1SxvYK8JATUCG87gOf+w/QLb2xk6azYKs1LYa22lZRheCFaV5E38+VsR8q8PCHnt+j0Zp1o2szd21opu+7hUFgRgJfUDLKcVdnz5lN5NJSl28bLOiTb652jJ3EGiWgF+glSUNYcYwUXCiBxRQC27NqOMybJ2rrCkY7oy8od4Hm0OpjPLoaCQ0YCFFHMiZF8MRO0MpUum7kiERL/CQw0T3i5WVFw8LzpIb2RCawy98hwTUP1cg1rEOFRZNJdr2z2bqjpHXoyqHNQyv+0C2cxyt60NA2UGSTBc1cGj/T27ud8EHY/GjwPzmkEsXCYMOnekVQwuUFehShgSqSvjLaYSm+Hb6UUwa5JrT6H2gFVMRXF24lYMZZdIu7LL1hD6AWx/64xVBPMXrpLArDvgdU66VX2BaKtVtVDmhrUyviscdQEQk18WSnSmrHlSPiYIAI4rKE+jZvpZkoU+PldYjMZpbE2H+C6xOlfxXScQS18OiNC4XNwBJ4ZG0jmm8K25C5580ILiknIptawYiiSNplZQJv3FGbca1nlN4FGKGlxBcdZkd8LqYxo9RzcJaIhsogUWB1EH2IwvRjr85FAeBha44PB59+A/gssFbjYDWWQ/EyZ+C6oSZlVTRar2Y9Q/OmecUHzy8mJpyXcnmVST29HwCVFKOYvITqqQt3fiiWP4fregJZsTvLCO7lI73vc0e1QqBpy8/+iqFrwBOjoeBuMFecTCfc263aJlzAxBXu92kTV98nXqIfxH8CJoKI9BUSKsfj/GFy1Gzf9MtNf5vs9i9vs2lnrOiT4T8rWvmsl7ZybmkSQfC5e/zM4MhcNghQXjPwvgP+plocaLdlTaXOmT5EebYTLDdcmkbkXCj+LSwPX65eHS3SpH76SQ1H3VSIdSxdFrICOSLcULErVFL+7ATwRbu7IZc+Hd4aMecGOruJrxNcY0YcIaVmuu9APbgsTCS1TjRqf6TSoQxXjCAH+/Mout0pydBa13soLKyoX0UTlqrGqNl45+yLdjoAd9qZB2vQbTmW7uf0taDuszh0gEFQB8HZh/jSYxNkPgc1N2ymrgHMULPYg6wQVDR4y28A0vu01w8dZDlHI9uyh+HS37FfEtw1YBch/tGbwqSLuIql8hUC+a2TQQvvLCqrUuwSQL4xoNFN7x2891k9PVaxYsTlg7/kEcwKJzDPzuVelQtTjtp0sduZNHViRGvw3rY14m8RmsdW2YhbTKWxqASdhKVU+5BXWC1+yiK/byDS+7IHh6bAqTHOdFQZNxvPxq8NRN7Nm7ZP/r1o+LHc1mUJHioXKa1bfvDeavryrYGiN9HAJagQtvDExQtwgLDjQ3zhajOfxO/HeVEzxNcfNfcXN4wMtf404yvvqOwgJ9utq9fQy+RmdBntbKDFZsuJXOX2Mw92rtxa3UCR6XRY8s4hteXvN1TYf0ARIw2SeSibFhso9AVwBaWXqyRbKf9+Pf8cIX+TfjSOV/i8eLUTaVu+1vchGmHMuka6ygLBAKKTww7TQvouDyX/ZP0vewGKnxArszP/bRAzs3iMT29862bbAZlNU1RkjD4caUBuc8/FldgM8EC2jugHrwgrWZnDHkopxoBdu5NindG2dzbYKX8w0vH13V1UmMKlQSobgGBRcP6K5WAx+jMBjXfzgWP982IfUfRJYFsvWJ4vOLxZ12bHa7Df6Jy8AhigRzlhai8jhBWHz+BBeqzM68iUfBxf9hz1q1ojvJyftJLPx6D136LJbGzj3zpYLg4tJ6/PcUFtQ1xeEDQxzBzt5Jylm+98jB28p+zwPNutjS2WbHsPh78WqfrDDUCV63ky7L5Btedj5EHyKio1xhrG2egsoa7BBgZP7Xu8zNUwM/sAk21kBFr3VnN4lChrIiWCUlHC7k6S2qsOUU6w+vcm1Nf9iRZJNi4lFcApULs6hqINi1iZ9B0pXEgkBYeIoWcbaD7gWXWdjpjR78tM5womFEQi8NazRoMazm7b0cPYM6wUUjhk2wXhyDG/wHB52+g1cnN+bvkLIDohw1L0409LwaRSLPSo6udkA3hO9lL1kc11uwskKK+RFwsyOSxN+FkVdMWMG1Cky79tDu4m3EQs2WNUjcDmEYBj/RX8prPEChEjg4oRJMfEGgAKmU2UnVpIGD6HLhhQ2vDH6VZFmv2czmYHUpquujG9uR8KJwrIhnTQeAHIQHBlM8xtjRhoLKoT0qpPjIyZJPcommTK6cAqAY7DQAACAASURBVG/k3xEFStir/BdE6uXfMWJBbWWtiemDOebHJIFdPfT25hX96l4AFur9c/b6KN9LIF1XYvgv2utkJUJmy3/2bRtKCvffuSPugics6t7fay9fhDCNYwz0qOcgT9k/xRObU2EuzhNoSMFXJsqTrjzFNteZESBMmhesn4SrObAuvzhsfum2brPLLJvXOKxWhOcuwkXuGkfMi4nzWLeYkbbW64CYNgSI4Ix4MmCw9GuvOP6QyqLCRrO52RLQCv2dkLfAeqZ0mNB9lT82S2wUxxT506t2llupbu6+mM81LWI40maltqRYr0a3YnmvPrCyFdu5wsPYQ/YnJYojMiSFVnebKb1RYVMsHFRcpp0TZjnxqYcozVdOk0+u8mhtrQ7/Eryx2WD8lgeU4yR5dfQDem9YpZoToknjRhBxnUlCoGrpymXCGFki9MNvTwvf+85qyeEqFSuNsILhLNGOEMaNvN4kI7lNx58+SJxmZeP/2ZXoCKyUHDJFTzOxBb4Avgdcz73RpfFhqjI3A4SUceYxal09rALROvxKMMgi6giTVptthYR/ViZv9xEmOOVSKEL8eUvOovHbTL4LRu02yQMYj74wuRwj7mt1j8Ob7LUNzFdpOk6Z+vNFU71F0hgIShHoxQOJryaulCdcsNckrk5IsjMXQTlqT7snC95kzgJhgjRmh9st8SvBPRyJtcdi23SZgBofXrMQBQxjSJaxCNwxuyqTHgIh+8DxEmdSGxOT7Q1ryvrFtntE6Vda92FLZSkdtBTWGzvb8Uc0rwaJ9l2Mub11mtlHBUBVtdK+/7yNYNkgDQQ7S18P1d3XDvuVYJhIveWc4DS2Gvk77qTga7YaZT23IWVRSNDgag2XZ1fPVI9iBZOuZmLMAzfwN9sklS/O4u+hLY0foTZspyekHjpmRxhozZ1N6kBU0NvJeSX8SfDRpOdJPKyNKkUqflzgXPS+PsrcrJg4oga+ZlGCc7wfeiDCaIxyzpHc9ZdoKM/w/UyLiAqeUmUYftL9U0/yjiUg4VECD77uJ0A2YDEBkAKnhixW2Hs8OKRWy4a9wjjmvg2tHwlGs7gd9ETXFBI1UorbC2I4nhcjsfMBtK0SRq4m8D8TnUIVTrAFGczK7phtkSW3FpDAejCbYIn+hf0++Xx7JVV2BUSP7wDvdGF2c6oPS1EqS67tH2yqWtPBlN8TWCuisvYjwShTGsMnHCm2X1YSRr9iWTYgA43kP7hF6aP94G11dh4PWCh7KT3s33qP5PQidNSsbQ8oza5Okm3yW1raN64t+NdeWi88TmksZk4CLRKfK6kX/3UW4/9FtlsDPszic+BcoGJfUQfw5xANmt8MHppXskyVOWjDFGoZiqy+sfuota1o5ixBNMWYDKGgj2QBiSoZ6FAawnUJVtssr2eLz97hxSMwFLwsWrs3o1mTlpfdSSEJh8Cf0S8R7YIKfxJ8tWkykIQWd7Pbi791SzWNQxd6WxF+G8N5o5r7RgppyJZEaP3dc9VJFA7XxZe4SrQazIbAOulbVBRlxV2UMrtvh+hZoo4p4vJ1YjOuzxi4wnp/8Nd+UhEGCYGQZJ1pLlpNJDURf+i2Y+MSMySqTwlj6ecy4zpUROFEF6xoWBA/nIVCAWhjtpGD7e5kXcXCX41ltFJR0zCG2V/98X8IiSyk59fexWu8k3e9nCk6bZHHy0E97UUkFi9QOGWARhwVrxAQUSXaW4AS55LNDwjtF7kLy65li5jviTSHBY3KF6yBRPQ1tX5DEnGSAt0bI0kNtBchsgs2u0i3wpq5QVJSEmkDK69rbNLeJ5G8zxcg+wX+zy3eRzM+so0VADQJ0Qh14mfAOsAPg4d3BG9u3Fd1wr61QdfbtCwnh+ZnCNOhMpHV6VI0WJAdcbuKhnySw4bKomvAGuwxJwNQCs+wu3zW1gcQA3YctQe7DJjF36lHUHSdSy2rp9y0srEC7A5ainVzYdnCBCNS7Fq0wXNpB+sPoD3LaPmRzPFOcgq4Da8IBqlgskpNeMPhnAYmpuvF352QppL1PuTbQGuOtHQLtSkFyAphRhZ5i32CBYUqGdtwvtu6FXYRnlYS3+Mr8bNgj0BenwzjegKld7dgjX62m0oXwLHdd6qcCsBiG9c4QNM12/vrekUwLkJUYr+MJTLEosVWCiiPUJ7/xROwV/3oAV8/wdzqrFQblqRdMEhSghvKGHswjTM3wUULEZIM7dvNLg6jIDMNAlXecPGeslhdT6p4G4zFY6QLv4/8lQPTBh9Y7XhFMI64KjUVKam47OBSEjOuaG1hEpB3LfQ+bKdS686NyxOTt39jS3IDVtOlFYqemz9xEJuEo0pZfDLSY8T+WcLe8X/H3wyW5NUsmccfChOChbuqJShcz8vPeOho7QMouwoE0vJBFGOvCLZ2J/Gqars0Em+TQUsvOByFah/mkVV3soRH6+CbROVz9XAUSKTEV0/2j5p5V+yQYOuHx5q2lPSLaJ2KURckQ544JLiq0D9alJ/yf94Nz3Exgjm4ewm/zaKzmS/wQPnEkCndq01Fk53NRTClbm29ATsreXarwc0iuopaGPBrynTrXph2Rc2eGXoBI9GWUWkhmviiEZan/lye4PuKZ5a23PJF9B//LZMejBrYrdFshXqsbDH3TPg9GGGhmex+Nqqmy+IlixIPqFvbrVuhieV6lRPWLcv4eJ2PSqMl38JS897WuygVC6BEA1uolanPr/ADwTOjY7lB4J1Ipa0CD7UJiChhUogzaCaaVCoTNc/mhjow7rppzbhMmDzJ5r6lmSrOeBkJcWwVAkrBknjm/qIFe2/eknFNl6hWdBPX4Yt5nVraE/ieYIwsyIJsgH5GGesqxybpoBScg34LBy8dxZ96kI1iNpiBXnSP4tP0p+SYaVQNVvkE5UkURgXIXbxaJQKrsPiOh7X5jv3a/N5LIJWVoZm7EIdMLxWCFSzZPcqa4i18TzAuaplDQlvGpBylB5mbfxJbtL5Iy3vHN2Y2Ztbur0d9F4fKJmZifW39HeWHcfqt5ompBAZhKc2pJKyR5CRCuaQ7yEpdqRQgFF2K4ghxfzRRx51vWnudZ9Ij+J5gXN2yDgWiqmEOFfKzOpLql5l5zXyUOteSncWhOkP90uhv7TrqPpoQFK+Sv8+FXplIPo/idgaNXVzAFsbNBgih5GmMrODCD+/g/fVbi4PR5Vlba5AqdjaAU6KT2MO8c8QFlt5N3ggQ2UVpBc4Fy5DDGuck3MNOik+KNATYtFnEwWeg9AYO5Q86J3CZKKg1lHqMq0mkgV2sRZ3A38qGNPvOcAvsj2U2ZmC31YAOJlXEX/3hhKmicidzv51f3eUmyVo5heXIXS88QZb6igDOkm/4B3bjA9gwQ3JZUIlwc6BtvVcERV7pB/ib4FbKDmrhmW5Ey+593SYOS+jwcIc0rEuw1hNeotpJAx9UJ9WcR664JrfP4+bwTeK6sMcnd58oK6vJ369wuq1YfWmrKV54XYcQ/E0wLDg73qifMHjdFBA18UD8kbx/HCDMKgPTh96w++hJ51MqCPleV/Msw6HqVTfxIjodJtviaUAMvvdrrxu3g623wfU5SbaSAY/sHu5/yxCG3/XBO7rPLuPhTs4y9T2p0ly0Cv2GD9fg/FXZflDEOIjawoGqDjMhB+RLM3J8lqrT4gt8nLD5v41HiTqoca7Vju1joWUHYGJW3XH+r1P4h+D7IMDaZMunyH/L/tz/DJgsndl1K3f8hBoZLXvdtb+IiyEaZCbiO/Z8gxS1aLVH4fJip93ectCTI/x1K/o5sWfk8GRpGphXBb3YvvZ2WRQy8AfB5fVASYRqfibH4tKNcRQj/KKRNyw119PSxg6yDBmsB/c6CV5gXdJmNhfWopxTxRMoT7lSYlYTTIEtRneV3Owf1yJVr7B4QPk0pw6pVmGPwpCEPwjGZCoFqMphIRoxJAtJyENLqLdwkgMHmjq9ihNMVijpuSDQ3uwDsC8pMV2+jyqrdFoyrtF8/MzHslVXN+y0GHk4x/FKuoU/CEbX37voYsE0Q3vEjguHGuEh+KkHVailEvDlLI9OhYOFpHi62qqGhnc8KrBjMEjC9Cvtmx2idySFTQqO12AC6o+ApEGDPwi+i/MZ+nBgJ7F7oHhyDzPSYUQcjya1PJUc4AihkJ5kJsVIdhjWCNub1gGwGq58JJqomBrBxPkbSgs3uZak6Al+mWTdnbJUuYjEQJCQOOjiWoe5VMWBCkaVsGFX6saBHPMmnJ6q/tHD4qtXRR69rNX0sqKpfX3nYRgSCIyHP4b8811u3hNtu15TOQwvgDV8wNaaut1ezbK9N1S3+IJlnA/MKEIIAQrhcP91QqLcCK3sDTBrBmTL3h8f5yQP4cdOltcYrdjeml7fCADBhYvm5u5x1qt8hQQi2nUUdYlrqhWlBqVmKsZYef1qzER95Dy/NTAY8D/BZTGERKCNuz88M+S43jiOG2YNJFo6wcV1es//ZzW1eRTYcM/bLMzytWUreoS/CV4Xgxr8zT387rn35IT0CkgWJWT2gDg+tdROsJV5fUqNoWKQ4GeCs/mc9SB95K4JrCPcyrsJagyb7AnCPbRVq/WWjPmS9dvkgjLQ6Wtk2NYz/EzwfsZv+4p/0oadwprVjRVzjPw25cMSvlb5DIv+/Ue2UyTM7GvIFBw3apJ2fp4GBgN+JhiTlZ33Jia6zE3JKWCN1MzhqPWgzLZuL3vaeAQbdL844/AJWpH4CDqAv9/BJ5vU91IyRNwTHoJyycGUy1qH5e9gQHPALwdwhSEaGXBd4uWmyVxmfaeysxhrOkQgY1V6Bgj+sN+NeseWZDkcO6pYnUEiz8o2ui21bLlQULgZGQsJuMDD2Engc/0hBAm2KG13ypcMdBbYzymUBNML3wyRvy1OiprEu3wqmM5CmbowI9AXQo/g2T5U51jrEpLq2sNH9JsZwCbjGvll9YmQIxiD7vzg8cevKyd0ikmWNg4nRUOsw1WFlGYM83V09+Ai5AjGcL+eRskor6suSt41y6XbyM1WiR9Lzt93l2nY3usOwSd4fEILuYeKFRa0busUy08qHMXRaNPRdTyNuOOIuhA+zOSUF4jCplrPCDrBsHZROLLnTRxqb0CrCYi4IPcLxpQvVx2qEVXB1toBSp0SJpBnWsTeWqJO/SHoBHfw5rdELySZvPGJ/dPCANronvYHIUSHIu0geKI3yDkAiqbBJsE+AwyMapF63MJwaVRlyFm4RkUZtJmQKHHULxyQPAkfiTiHh+HWlJUQjmcnpfcjK55JOEQRdIKtw+IbOUoo3AR0uu5u1Cu4YlsrwYgxidI9c6Jt4/aG+ETRSK82l3OUnfBKvBpCCDrB3uHBUqfBwnLPO/Rn6PMpaAX4AYlBF6ksOMTdmMCjwWdx/UMCQSb4QQqJu+i6mofYQEiEfeBJiMvSH2luxdWQW2Q2PBC82h3/Rf0gyASD6LeG7y/LUUVlBL9bn7QpROWimCgYBZVidKwbay5TQyHIBKf4VfSrXB7dbJgwUJxxC9OuyNsFW5n19J7UGgedX+he5+qu83yHJoJM8AKBgn6+vywHRFWTlJXgmGQL/ph/sQwC7YlhGY7FkCb22v75hAzhRYhiup3qEsGeZK3tNN8X4b40sH+uzH4a+rM81TLGpOaaKzAXsAtei4EW2eOBBrnuJzoJKQSbYH8i+yeF+wMsgaJIE9v6Z7KMYCi2UX76OBxbxcrwMl82peo3HsRHDTYMTPAyQirJvSmy2yRUltv1PIqQwguCMV5D5cdLIiW59qlkMmCWqk1QyMPABBNHAWWiYk5XvMcWrRKkHvaxe+8O/Eaug2qgT/FlyBP8sF8r16HPVAGEdFEeGuF4iAHCKn7l/HIQv1TDpDqUEfIEQ6fzLJ0UhCW3c6YoG5HYTy2rGCgt2lxOIvFXtS9X0pnU9Z3/XMAQ6gRjFAYP7dGPfmevV86fOtymijw4/ZRYxJwRVyHwYh37D/RKGOPHWX4AEOoEoztYRYSHRZu0TTAxN8edByirgvQRYiRcTOGyWy1sD5h0uo6dGMoIdYLpnmjydQWizuVGOsm9imE067I8DnEyIWkOHM+fqWJvx1QTekbIE1xBoHG7lsUP+lkAvmXCqSbiiRt9+gmv15WdFtlH9YH8TZ6GEgkNGI1gTCeuqcudzVNQzsQ4ShHyVdCNmqS1Y1iI7Eakk0qwCB2BEVyydC9d0nKiVpBXBl0QXBTv1E+1vBg7ZbbQm+dMLqD08YjuPL4WRKwcov1B3YIRXD/2w0GVJqc0cVpVBwSX37ZYMpn04ngVUkc1o7dl46JHwuw8gpBEC76JmT6aqMyhjQBG8BvWJ796Qi1vOq0a+gQflQ27kOVMludE4LuGXJKMMfdOY7C2KXgARB1zKb3aKEmeBfxMImnlfFwLdTCC/2GlwtNb+rbTqqFPMGiXG/EyUNecFXPLmbykv6wqpny5Vgr/N6A3VMGgmrOsTPnP/SrKtu+xJjhqmXQGRnAHELSf/MK5/VroExwvs9DsgNQJKEyGAhDXWHiGZ/VlzqHWJsKBauXphHS1ziOkA1+J4SScXaI4DpdUShWj/sAILoNcfieWOl9vhj7BMyUFHy2d3mkrFtCenXYlmPEUsicy643dcBjJRmEZN6vEozFox4WmWmAkkE71DL5MyhS4vX3EedUQJPjebmWmwT3jHb3NMW6DpWx2121cLtaDFtUjH2MMa5xBo/hDzJmDAa2ReBRvbR/aeJXDFXUFRvDqP9+m9MjzzhPnhh7B22z90QkgdGkrcQdIm4AxWgBM5IUCTDHHAC0EZ3CIg1v2D0PMqxnB7+GjnPm506qhRzBIj+3Cdl+LJ0l2gqyzI5fa9AXC3Kp2EXOXkVLb5U9OkyV/ALorg9qopj4VwHZg3+BVtvOS06pBIvj45ANap2BqXF15qK6rWREsejB/A4kRD92KJ7Gb5ouaRVxAgTEWhKNvT+mT07qOjMYIfg7Df+X92WnV4BC8jkkW6e7JpxzObSQO+dhBcuEqxsZaoc7TssAfLViH5i/bEnH0tk6rPqiQrhTW1j6IkBs0MIJbdbMI64KGzkMnBIfgVKbEA+39NoeTNzbYJb/D4PJqiWblwAw7GcOjU1hc2dw/M4J5cP8ruNNBrAw7fb35AkEGIzg36qXa1f8Sr2bsICE4BEN6lCTWM+u5Ud2ysKPLnFaYaVbKqyImxeSDPdrISkHzYKerh20PBYiziMz5C13FsgkOwZcq4QsRRErPkwaOj6DlPEaOPV/VfZkiDsKY/wW9k3jKLCJ+/0vNq02XNEhDiGb4eF2AEfzoq/ihLsP4BWmSVX4DJgh7CXmKqKS6zBWImkPZhHq0u5cU5lEJW4hov/GDUOp8K0NLK3hggb3/mq7ACK7ealUdl362QV0mHWW9rLn9cZRTWRWiZ0Tu9ElOPHxL71G6MqkdZ25/n0VCLx7o4xaHCBjBL1tp6buuqgaV4MGM4HX2x4fB0VImg5TJnErBxN3BD6F0cmt1HzKMYuk8EaIShzc7n6+EDhjBrwl/77iqGlSCMQROO57ezLp3nziRukpY3LSCIQ3k3EGKM+IQ8ApSIy1UuzpmRBVzNZRecfkyB69Xz8MjBxR6Ibi8pRQezSLMgGuI1lP3hk9QmT2gKMMhchJxlHwx5AkzuTgu7jpHXIYjRfXELOd1QgWM4J89++yz8Oe0asiIKjPk7A0gJMIhAwMudV5wCCQOS6Ne0u6DGiSJm0k/mbUQJ1lbh+1pwNZlzoBJL+c6rxMqYATf4XBaNWQIxgQQvJPhkmeyfQ0mjXJwUrmcSOrgi/l8T4hKCknB31NEjAaP0o+Io5uSPfoSUkUnCcH1aFUJgYBr8yEaqcQkOoe/l/VYUP9qhw64JZyMKsellXI4BouBT4ikPNTEhQN6EVDrkWBqzci0Sav6ERZgAyZQsoURDLQKpUPJuEbzxA9Bag5ynk4l9mIqGMObnphmpOR2uiRYgVPbYcVSCFxNXdxkPJ8BF3QkrRTz3B7C+Wm8jFmccqgViZY7M5yKIFFqOdV0DP0TzFAMXDV08vqMk8+dhHfokofDvjm/VaivzCWrk8WP+zAKwSAJqQz6YS3BMYRy7yHuCF23irAyIjfL3Atod+lkBdyjQguGIZheOlCKagNVSQalWamkga2zcvWCVl17dCakul4mVfYwDsECHrDJk4QFVWIchJtUCuGvlaPSDtehqkbK6ZCHoQimJ+umbJbtomO32iAMU+6xmA/aHVxQlYLrBMYi2A7rVAWWgO1zK6ADtNYWXtl6US7Yw9AE37Ep9V3ica96qzVPWrav123SO0MTTA83binaxtLTqbHaFgFFL+v4NesUxiZYjsrOXqRgmGMLP2sohA3BKAnRtNrAKGfzAticgCFsCEZR1g6tkwuEk+/YB+gwBNQJLhg6YrSFbzLap6VhyEadE5zVstq3wuaHvqpmtfPqpLk3G9Mb1Ales4uu2Ms3O0R7c50TzLDC2UBtRKgTPC6LHlnEN+sHjJpeRumw9983QraKms70xFq48VVrjxOgBhsqBKf33wbMLmcEL3/wgK7aLCwkcqYaoQd3IhVP01QKHqVuZ2MLMaj34HU76bJMvjl2h25Ca0ZDDNHZCSTK0YvNOc7AFHuJX5rjf6gTXDRi2ATrxTG4udJ3+FCU0+mP4E1tRzrOnNx0cLEcvSoWc+wte/QEIy+TIAhDQ9fVVAEB1cSMHTRdx6F29E9wnubyFS3vPLwq5jE1gnmH3gnOr6otQ54icBTt4XUXAcGOmYj1B70TPNpJLy1tTmI8Xd5AoloHTzc9Qu8E9/BiGHaKgnXOYw7pBXon+LDAbzufXKl4fPJgfUcOVoXeCaYnhy9XzrIeT53hQf6q67Gqzi76h+4JtkdBBCFRFbe/aEScOrvoF4YjGAIyEO28HFpIQIIb+6FBQYbhCD4GRF1wXc8OEPWM1F++QjWGuJ5hOIJpT0IGuKpzcmOO/aF1vdZCmLxKDid0DuMRTPNcau7BAivL8XAW9OKl/mhSEGEogqfGNLiieuJ6o7gp0h4GFR7sWAszBus7OrQjjEQwhK6MUT0Tr2DOau81KqIfIal69UHSgpEIxsgratpA9EWS+alApIZ7alfIOqd2VNcwEsHfC7zFqp6BHrxWtn9trwFlVuowEsF0RmzDq6onrjaIneTB9XTrFCyDoQj2LZbpO8wsh0mwFgr17BUswSQYUTa1yTw7y5AbQPDs4DTHhzAJRkAwpgnKQ9YYHRvLSjAJRqilengyuvd51cq6gkkwAgIwdVAcWdNosAdq5dCDSTDiViqprUiiBmvq+sFqjS8RdgQ/7ln7WzeqdfKXrVegEXYE1xaIW++62lSj2HeEHcHQMzu5rlbWnlStqA9TSCLsCAaxtE6CtfsEYUfwuWTSrdx1NcMg7AgON4QrwcXn3c2TpnOEKcGQdueEtFuyTbchGlwhTAmGnKa1bHt5UZgp2JAIC4IX1+xp5+oLocOr0c2J1XbD3hKtqLQGQDgQnC7QV1vcOdZrQj7z716DZrIgoPxWLBgQ4UCwXOwIAVVShO3xOSdR3kz2UOvWeREGDVRJw4PgmQKPYowvdBjnaqKbrOM2F/5vOB201vkZ4UBw+VckRdTsopeoaLuxvWryXpoLR6YGq21+RzgQLEOhow8hhqGNie2j25DfzhFmBGMmPLs8q+MIeZG44bGmT4QZwTQSGLZzT8k+Hyd7S99N6w+WWMVnjWEbH24EfycPrXTvq/osiXQb4WhfdqwoAgNkHRf+f/n1k6C00acIN4Ktg2JSbRllwa3/ABRy2iV04x0WkxN/T2uhx7+ncfJCCOFGsBw440qzO4jSj5O0GhJsAKOdcCYYrTvW2B+cRshwStciv6nC/p0Fu4PQNJ8hrAk+kuA4dy5aNAszxp+e05jUvk7pRYHn7oFvms8Q1gSroFyYUFeSG0QPlEtGdAiTYCVO24dxQNnm3aFDbgetSd7BJJjjZDSpk8/l0/KX7pMqhMyuJBzMC1rTvIJJMAdoiAdRTDTdWuH4XX72MaaE3x6khnkJk2AO4LApFJ44av4vw0mdhu8Id4LzGpGU61AApbFmasOxhIwMXJt8inAnuLdonFU6veNm1Rq3IS6iRbfz6HAnuLqqvEpmGS8sk1JLA9ggXyPcCZ5NHEOtPKxKqog2tQ9VpV06QrgTbF3QcKy9qr87mlwy3AKC9ZoUCxDuBKuhrnzYhh09aw1Ngh2xWuC0t7hj2bQsj+7rPEavgaRNglWw/ZtvFbPmQwLjNYPVGC9hEsywiZB2mkuhAVpRTnUAk2AExqP9TuvsfB3r/k2CEWgcPUrrbHkrQvSaJsskmCFZINhd74b1CdX1k2w2vAk+UqMal0/mjenmrovwBejtLvNChArCmuACoOqye3UtO3bySdgy+NRx/7XKtwhrgsH4mSyXH3k4YuA12J4dNEsp3yqPJ+RzJqI+b/ZgvSAPqJIn0SqLEg7cp/ScYuELjhAHoOoctp+ekJIR0HZ6g7AmmO5LSlgt3z8FNK6gdCRsueyqpC4hP9D9cKSlVHN2lRhN7XFIIbwJtgdmTjrI83/zV+40XARjKp7Ftopom/coSK2sEEyCFZhNSB9hU9ZIytLSB7gspwf/SbpIoi4QV5MjQWliBWESLEdRU5J4lpVshneQgrqRrca9ZJJwjWsRCwPePg9gEiwHvHvtA7/Tw71nSDNq8EOsCwebtrkYyJZ5DJNgOVq4FDpXJToLM2wSLMca5VRZBTOEGkMC1BqfwCSY0gutm4nzpVXNRjl/s1pXdFyoKwtLk2DmJqxX1yOXMAnGaTJZ7bqePmESzCxjMazD5tQPKzW4zg9nZRgisJJJMKXp0UzKfBK9+rnB7BRCIu4Hs1U+gkmwhOmyuBwomhwe5Ab5AibBEvbb4nJwI63BQW6QL2ASjMjq1uYopdNi3vysxR12ZBCB1JRnGtfSqzEWh0kwoQUQPQAAB39JREFUoBSin2Upjx1Nf0KLoB9fpbQ4JyjN8gVMggEQRZqoZLw7AceX0YmEtNJruneTYEAOEKliKYnWtKcewf+tgW+VT2ASjEgnZJzdoWt90+7R7XGxC1gEB/sE4XqBSbAGQPwRwUQd1nicXq8Kcos8g0mwBtbLRu3COXVlNjz6gkmwPXYO2gSbYwoVxAjdWHDYwyTYDmBvNxEKQwmZaTsKttD1gtYmb2ASbIdq6iYbF4YtKL27Syt5VvGWA35tlBcwCRZR2jWmbhalTQWCa6hW2EJ4+HAHQGq8Fv5smxcwCRYxljF7vTKJPKtaIVHTzx/yqJE7/myc5zAJFtFI1CRpRR1N1OzbK+Gj2f5qmHcwCRYxTyCpibT7YNE2O+nkRqLlBl4U5xhsK1RgEizCOjWxp6RT2Cuw2UpZYZU8+A7gp0XXeKn8oPqoHgIwCVbFDOLoI6oIn0WZ09K+wDbLA5gEq4LZdihjVPaQ7HkQsQqfllCFSbAqkN8pymMQwvKkbB/SpbUJaKs8gUmwA7ZWTT24UNV5sFxZT6hzJVCN8hgmwfa4Dp03J2u369iFuSd1EBzNJNge6dwJ3CAwCbbHVSBYz/FllTAJdsDG9/7w0WLX1VxjACEbfHEdr2AS7ABMT3lNfuTCxO89sLnbAdcJuvuLOsEFQ0eMtlBqWVXfVg4fgncBMdtkB44I+90qfp1ZcJ3rruv5F+oEr9lFV+yl9NGpHrZy+BD8BBfBVaXIld1c+v2rAmz1Kgfd2la94eOy6JFFUOhhK1/NGFhNta7xcKbVZwqt/xDPCKZH2ve557tWeQiVhqf33wakYog/TrBQnlc/JsGxrkHRXkHp4xhCNgaxNV5B/clct5Muy4RCD1k5XIZoykwqG8n2b8klGnljex4LcHu8gDrBRSOGTbBeHHM2rW7afizDwTAimK5ulKaZbTRJ9BfXBcxlkjpKf9CK+Y1uLqMD2hhvYBKswKHOQx/DtjBa04zOeXqHkINJsBzgZYhTSYxFmaVeaSMhbfXj5GASLMd44PUhZRHRyK1gN8cXMAmWAzVJoPUtjiOkfbBb4xOYBMth7UrIFiyV7z/hoq5OYBJscJgEq+KefmZRLmASrIKsKNWADrqESbAK2oI2KdiN8BFMglVQS92FVJcwCVbBBiaMLB6SMq7cde3QhkmwGg6NBouO7gLPI4LdFG9hEqyNOFFw6QZ+aD9cU/0UVJgEa6OZQHB396qCM2Id/7bGQ5gEa+N+Q9I8V/XMjk4TixQHeodsmCWTYA+wU+krLmCOh2Zb/odJsAfo6kBnWStSJTTdXUyCPcBEgd8qwW6EmzAJ9gAljUh0ZrAb4SZMgg0Ok2BtWBc1m1TqulpowyRYGzOFV+3XripdHLkgpB8Ck2BtJLuhc7hAWLrZkIVJsDZAFl3byXnLmik3RsGKyXW0h+DBJFgbT+qQZG33z+mJrwvcjgpZERaHSbCHWMNCaQ2pE+JW8CbBHqIbI3gGzS0LdlOcwiTYQyxFfpOKXNcMLkyCPYR1WHzdsyEaI1oOk2C3sKRuf51mtzMJdgfgi9Yg2I3wDCbB7qBtyKp7XcIk2B1MFvhNDHYjPINJsDso60iSLwS7EZ7BJNjgMAk2OEyCDQ6TYIPDJNhfCBEhpkmwf3A6ksSGRBAXk2D/IEVYObcOdiMAJsH+ATiupQS7EQCTYP9gpEDwsmA3AmAS7Cek998Z7CYgTIINDpNgg8Mk2OAwCTY4TIINDpNgn+Jk9fiFwW6DEibBvgRGgz8Q7FYoYBLsS9wBgicGuxUKmAT7ElYgOLR8/02CfYozdeJDQkApwSTY4DAJNjhMgg0Ok2CDwyTY4DAJNjhMgg0Ok2CDwyTY4DAJNjhMgg0Ok2CDwyTY4DAJNjhMgg0Ok2CDowIEf/dWDW+Rkur1Jby5e/Vg3j050DdcWlGC6fIZ3qJ2e68v4QWiBgTx5lPfnxzgO2ZWmGDv0ffbgN7ODjV+CuLNS94vDMp9A0twelAN0mbcCOLNy4eUBOW+Og3QZ8JdmAQbHIEiuGDoiNEWvslon5aWHaD72t+bl4N090B/cxo4gtfsoiv28s2ObQG6qcq9eTlIdw/0N6eBI3hcFj2yiG/WDxg1PZCJDuT35uUg3T3Q35wGiOD0/tvg+y1nX3P5gwd01eZA3JdDfm9eDtLdA/3NaeB68LqddFkm3xy7QzdtCNB97e/Ny0G6e6C/OQ0cwUUjhk2wXhyDmyt9hw8N5Kpffm/8F8B7B/ebU3OZZHiYBBscJsEGh0mwwWESbHCELcF3fvbaa6/WekRz2/31lffXUjrizTdahnKaWE8RvgQ/S6mlZTdauUsp/elPh/a8WWSptCbYjfIDwppguir+0Evlwva2tegBpfXnBLtRfkBYE1xWb8h8KYf7mZcfBbE9/kL4Evyzd955u0vxgurigUOvHwpme/yF8CX4WdwceQ4saY7cofveOBfcBvkJ4U4wjWuSRw8/f+Tx/1wPbnv8hbAnuLDDn17+fBed9PQbb7wxPLhN8gvCluBwgUmwwWESbHCYBBscJsEGh0mwwWESbHD8f3Q8AvHG+uvjAAAAAElFTkSuQmCC\n"
          },
          "metadata": {}
        }
      ]
    }
  ]
}