{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMlS058Qf9HIDsERwPAtW7U",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/DCEG-workshops/statgen_workshop_tutorial/blob/main/src/09_functionalGenomics.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "N1M4q7iaMEQo",
        "outputId": "ce1947b9-9851-4c8a-faf5-df50f5103ce9"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive/\n"
          ]
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive/')"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!ls /content/drive/MyDrive/statgen_workshop/data/workshop9"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "CWs1sO03MfhW",
        "outputId": "f5b8fd52-3640-4c67-fba9-fabe985b7bc8"
      },
      "execution_count": 24,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "assemblies  genome_hg38.fa  genome_hg38.fa.fai\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "analysis_dir=\"/content/09_analysis/\"\n",
        "input_dir=\"/content/drive/MyDrive/statgen_workshop/data/workshop9\"\n",
        "os.environ['analysis_dir']=analysis_dir\n",
        "os.environ['input_dir']=input_dir"
      ],
      "metadata": {
        "id": "VhYWG7_oMbbG"
      },
      "execution_count": 27,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "git clone https://github.com/DCEG-workshops/statgen_workshop_tutorial.git"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "H74zch9bUEpZ",
        "outputId": "be88ce5f-5650-47a4-bed6-1cda729b48ee"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Cloning into 'statgen_workshop_tutorial'...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install cutadapt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "C44akAkIMzch",
        "outputId": "74eaf7b6-1036-4572-b319-af7395858ef2"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting cutadapt\n",
            "  Downloading cutadapt-4.6-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (287 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m287.2/287.2 kB\u001b[0m \u001b[31m6.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting dnaio>=0.10.0 (from cutadapt)\n",
            "  Downloading dnaio-1.1.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (110 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m110.5/110.5 kB\u001b[0m \u001b[31m11.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting xopen>=1.6.0 (from cutadapt)\n",
            "  Downloading xopen-1.8.0-py3-none-any.whl (16 kB)\n",
            "Collecting isal>=1.4.1 (from xopen>=1.6.0->cutadapt)\n",
            "  Downloading isal-1.5.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (259 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m259.9/259.9 kB\u001b[0m \u001b[31m10.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: isal, xopen, dnaio, cutadapt\n",
            "Successfully installed cutadapt-4.6 dnaio-1.1.0 isal-1.5.3 xopen-1.8.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cutadapt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Jefyj66kM7sK",
        "outputId": "e50e7840-afb6-48f5-ea6e-b4ea78ce07ce"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: \n",
            "ERROR: You did not provide any input file names. Please give me something to do!\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "cat statgen_workshop_tutorial/src/09_functionalGenomics/script1_toRetrieveGenomeRegion_HPRC_DCEG_GSTM.sh"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "69_ZeA5NNEX9",
        "outputId": "b7d1cfb3-cbc6-49b1-f51a-fada5c324e49"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "#!/bin/bash\n",
            "#\n",
            "# this file is myjob.sh\n",
            "#\n",
            "#SBATCH --job-name Task1-RetrieveGenomicRegion\n",
            "#SBATCH --mail-type BEGIN,END\n",
            "#\n",
            "# load cutadapt/4.0 (uncomment the following line on Biowulf)\n",
            "# module load cutadapt/4.0\n",
            "\n",
            "## Author: Oscar Florez-Vargas\n",
            "\n",
            "##########################\n",
            "## == Set Parameters == ##\n",
            "\n",
            "genomicRegion=\"hg38_chr1_109655000_109742000\"\n",
            "\n",
            "# Internal Primers\n",
            "forwardPOS1=\"GCCTGGAATATGGTAGGATCTCAAC\"\n",
            "reversePOS1=\"GTGCTCTGGCATATCAACATGTAAA\"\n",
            "forwardNEG1=\"GTTGAGATCCTACCATATTCCAGGC\"\n",
            "reverseNEG1=\"TTTACATGTTGATATGCCAGAGCAC\"\n",
            "\n",
            "\n",
            "# NOTE: POS primers are in the UCSC genome orientation, whereas \n",
            "# NEG primers correspond to the reverse-complement counterpart\n",
            "\n",
            "# Define the error rate\n",
            "error_rate=0.1  # 10% error rate\n",
            "\n",
            "## Set path to where FASTA files are being stored\n",
            "PATH_TO_FILES=${input_dir}/assemblies\n",
            "OUTDIR=${analysis_dir}/retrievedRegions\n",
            "\n",
            "##########################\n",
            "\n",
            "\n",
            "# Create the genomic region folder\n",
            "if [ ! -d $OUTDIR/$genomicRegion ]; then\n",
            "  mkdir -p $OUTDIR/$genomicRegion\n",
            "  chmod 775 $OUTDIR/$genomicRegion;  \n",
            "fi\n",
            "\n",
            "# Create the \"fasta_files\" folder\n",
            "if [ ! -d $OUTDIR/$genomicRegion/fasta_files ]; then\n",
            "  mkdir -p $OUTDIR/$genomicRegion/fasta_files\n",
            "  chmod 775 $OUTDIR/$genomicRegion/fasta_files;\n",
            "fi\n",
            "\n",
            "\n",
            "## Loop for the set of primers POSITIVE and NEGATIVE strand\n",
            "for i in $(ls \"$PATH_TO_FILES\"/*.fa.gz); do \\\n",
            "\n",
            "\t## Steps for performing in-silico PCR on the POSITIVE strand\n",
            "        filename=$(basename \"$i\")\n",
            "        temporal_file=\"$OUTDIR/$genomicRegion/fasta_files/tmpPOS_${filename%???}\"\n",
            "        output_file=\"$OUTDIR/$genomicRegion/fasta_files/setPOS_${filename%???}\"\n",
            "\n",
            "        ## Perform in-silico PCR\n",
            "      \tcutadapt --discard-untrimmed \\\n",
            "       \t-g \"$forwardPOS1;max_error_rate=$error_rate...$reversePOS1;max_error_rate=$error_rate\" \\\n",
            "        -o \"$output_file\" \"$i\" 2> /dev/null\n",
            "\n",
            "        ## Check whether the output file resulting from the in-silico PCR on the POSITIVE strand is empty,\n",
            "\t## and if it is, proceed to conduct the in-silico PCR on the NEGATIVE strand\n",
            "        if [ ! -s \"$output_file\" ]; then\n",
            "\t\trm \"$output_file\"\n",
            "\n",
            "\t\t## Steps for performing in-silico PCR on the NEGATIVE strand\n",
            "\t\tfilename=$(basename \"$i\")\n",
            "\t        temporal_file=\"$OUTDIR/$genomicRegion/fasta_files/tmpNEG_${filename%???}\"\n",
            "        \toutput_file=\"$OUTDIR/$genomicRegion/fasta_files/setNEG_${filename%???}\"\n",
            "\n",
            "\t        ## Perform in-silico PCR\n",
            "        \tcutadapt --discard-untrimmed \\\n",
            "\t        -g \"$reverseNEG1;max_error_rate=$error_rate...$forwardNEG1;max_error_rate=$error_rate\" \\\n",
            "        \t-o \"$output_file\" \"$i\" 2> /dev/null\n",
            "\n",
            "\t\t## Check if the output file is empty and remove it if it is\n",
            "\t        if [ ! -s \"$output_file\" ]; then\n",
            "        \t        rm \"$output_file\"\n",
            "        \tfi\n",
            "\n",
            "        fi\n",
            "\n",
            "done\n",
            "\n",
            "\n",
            "## Loop for rename the headers (sequence identifiers) of the reads in a FASTA file with the name of the file itself\n",
            "cd $OUTDIR/$genomicRegion/fasta_files\n",
            "\n",
            "for file in *.fa; do\n",
            "    filename=\"${file%.fa}\"\n",
            "    sed -i \"s/^>.*/>$filename/\" \"$file\"\n",
            "done\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "bash statgen_workshop_tutorial/src/09_functionalGenomics/script1_toRetrieveGenomeRegion_HPRC_DCEG_GSTM.sh"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HDmtAkH3UcSA",
        "outputId": "1d3e2718-0288-4833-f89c-8b7d769f17cd"
      },
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g GCCTGGAATATGGTAGGATCTCAAC;max_error_rate=0.1...GTGCTCTGGCATATCAACATGTAAA;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setPOS_HG005.1.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG005.1.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.487 s (486777.782 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         1 (100.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                0 (0.0%)\n",
            "Reads written (passing filters):             1 (100.0%)\n",
            "\n",
            "Total basepairs processed:        68,652 bp\n",
            "Total written (filtered):         68,602 bp (99.9%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: GCCTGGAATATGGTAGGATCTCAAC...GTGCTCTGGCATATCAACATGTAAA; Type: linked; Length: 25+25; 5' trimmed: 1 times; 3' trimmed: 1 times\n",
            "\n",
            "Minimum overlap: 3+3\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "Overview of removed sequences at 5' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "\n",
            "\n",
            "\n",
            "Overview of removed sequences at 3' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g GCCTGGAATATGGTAGGATCTCAAC;max_error_rate=0.1...GTGCTCTGGCATATCAACATGTAAA;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setPOS_HG005.2.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG005.2.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.419 s (419023.752 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         1 (100.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                0 (0.0%)\n",
            "Reads written (passing filters):             1 (100.0%)\n",
            "\n",
            "Total basepairs processed:        68,665 bp\n",
            "Total written (filtered):         68,615 bp (99.9%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: GCCTGGAATATGGTAGGATCTCAAC...GTGCTCTGGCATATCAACATGTAAA; Type: linked; Length: 25+25; 5' trimmed: 1 times; 3' trimmed: 1 times\n",
            "\n",
            "Minimum overlap: 3+3\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "Overview of removed sequences at 5' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "\n",
            "\n",
            "\n",
            "Overview of removed sequences at 3' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g GCCTGGAATATGGTAGGATCTCAAC;max_error_rate=0.1...GTGCTCTGGCATATCAACATGTAAA;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setPOS_HG00741.1.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG00741.1.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.383 s (383080.244 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         0 (0.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                1 (100.0%)\n",
            "Reads written (passing filters):             0 (0.0%)\n",
            "\n",
            "Total basepairs processed:        86,992 bp\n",
            "Total written (filtered):              0 bp (0.0%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: GCCTGGAATATGGTAGGATCTCAAC...GTGCTCTGGCATATCAACATGTAAA; Type: linked; Length: 25+25; 5' trimmed: 0 times; 3' trimmed: 0 times\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g TTTACATGTTGATATGCCAGAGCAC;max_error_rate=0.1...GTTGAGATCCTACCATATTCCAGGC;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setNEG_HG00741.1.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG00741.1.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.018 s (18121.719 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         1 (100.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                0 (0.0%)\n",
            "Reads written (passing filters):             1 (100.0%)\n",
            "\n",
            "Total basepairs processed:        86,992 bp\n",
            "Total written (filtered):         86,942 bp (99.9%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: TTTACATGTTGATATGCCAGAGCAC...GTTGAGATCCTACCATATTCCAGGC; Type: linked; Length: 25+25; 5' trimmed: 1 times; 3' trimmed: 1 times\n",
            "\n",
            "Minimum overlap: 3+3\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "Overview of removed sequences at 5' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "\n",
            "\n",
            "\n",
            "Overview of removed sequences at 3' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g GCCTGGAATATGGTAGGATCTCAAC;max_error_rate=0.1...GTGCTCTGGCATATCAACATGTAAA;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setPOS_HG00741.2.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG00741.2.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.371 s (371129.036 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         0 (0.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                1 (100.0%)\n",
            "Reads written (passing filters):             0 (0.0%)\n",
            "\n",
            "Total basepairs processed:        68,594 bp\n",
            "Total written (filtered):              0 bp (0.0%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: GCCTGGAATATGGTAGGATCTCAAC...GTGCTCTGGCATATCAACATGTAAA; Type: linked; Length: 25+25; 5' trimmed: 0 times; 3' trimmed: 0 times\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g TTTACATGTTGATATGCCAGAGCAC;max_error_rate=0.1...GTTGAGATCCTACCATATTCCAGGC;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setNEG_HG00741.2.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG00741.2.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.019 s (18591.404 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         1 (100.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                0 (0.0%)\n",
            "Reads written (passing filters):             1 (100.0%)\n",
            "\n",
            "Total basepairs processed:        68,594 bp\n",
            "Total written (filtered):         68,544 bp (99.9%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: TTTACATGTTGATATGCCAGAGCAC...GTTGAGATCCTACCATATTCCAGGC; Type: linked; Length: 25+25; 5' trimmed: 1 times; 3' trimmed: 1 times\n",
            "\n",
            "Minimum overlap: 3+3\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "Overview of removed sequences at 5' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "\n",
            "\n",
            "\n",
            "Overview of removed sequences at 3' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g GCCTGGAATATGGTAGGATCTCAAC;max_error_rate=0.1...GTGCTCTGGCATATCAACATGTAAA;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setPOS_HG01175.1.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG01175.1.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.493 s (492683.411 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         0 (0.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                1 (100.0%)\n",
            "Reads written (passing filters):             0 (0.0%)\n",
            "\n",
            "Total basepairs processed:        87,117 bp\n",
            "Total written (filtered):              0 bp (0.0%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: GCCTGGAATATGGTAGGATCTCAAC...GTGCTCTGGCATATCAACATGTAAA; Type: linked; Length: 25+25; 5' trimmed: 0 times; 3' trimmed: 0 times\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g TTTACATGTTGATATGCCAGAGCAC;max_error_rate=0.1...GTTGAGATCCTACCATATTCCAGGC;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setNEG_HG01175.1.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG01175.1.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.033 s (32718.658 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         1 (100.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                0 (0.0%)\n",
            "Reads written (passing filters):             1 (100.0%)\n",
            "\n",
            "Total basepairs processed:        87,117 bp\n",
            "Total written (filtered):         87,067 bp (99.9%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: TTTACATGTTGATATGCCAGAGCAC...GTTGAGATCCTACCATATTCCAGGC; Type: linked; Length: 25+25; 5' trimmed: 1 times; 3' trimmed: 1 times\n",
            "\n",
            "Minimum overlap: 3+3\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "Overview of removed sequences at 5' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "\n",
            "\n",
            "\n",
            "Overview of removed sequences at 3' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g GCCTGGAATATGGTAGGATCTCAAC;max_error_rate=0.1...GTGCTCTGGCATATCAACATGTAAA;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setPOS_HG01175.2.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG01175.2.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 1.749 s (874665.260 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       2\n",
            "Reads with adapters:                         0 (0.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                2 (100.0%)\n",
            "Reads written (passing filters):             0 (0.0%)\n",
            "\n",
            "Total basepairs processed:    12,835,171 bp\n",
            "Total written (filtered):              0 bp (0.0%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: GCCTGGAATATGGTAGGATCTCAAC...GTGCTCTGGCATATCAACATGTAAA; Type: linked; Length: 25+25; 5' trimmed: 0 times; 3' trimmed: 0 times\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g TTTACATGTTGATATGCCAGAGCAC;max_error_rate=0.1...GTTGAGATCCTACCATATTCCAGGC;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setNEG_HG01175.2.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG01175.2.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.724 s (362044.930 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       2\n",
            "Reads with adapters:                         1 (50.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                1 (50.0%)\n",
            "Reads written (passing filters):             1 (50.0%)\n",
            "\n",
            "Total basepairs processed:    12,835,171 bp\n",
            "Total written (filtered):         86,989 bp (0.7%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: TTTACATGTTGATATGCCAGAGCAC...GTTGAGATCCTACCATATTCCAGGC; Type: linked; Length: 25+25; 5' trimmed: 1 times; 3' trimmed: 1 times\n",
            "\n",
            "Minimum overlap: 3+3\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "Overview of removed sequences at 5' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "\n",
            "\n",
            "\n",
            "Overview of removed sequences at 3' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g GCCTGGAATATGGTAGGATCTCAAC;max_error_rate=0.1...GTGCTCTGGCATATCAACATGTAAA;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setPOS_HG01243.1.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG01243.1.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.437 s (437320.948 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         1 (100.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                0 (0.0%)\n",
            "Reads written (passing filters):             1 (100.0%)\n",
            "\n",
            "Total basepairs processed:        86,968 bp\n",
            "Total written (filtered):         86,918 bp (99.9%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: GCCTGGAATATGGTAGGATCTCAAC...GTGCTCTGGCATATCAACATGTAAA; Type: linked; Length: 25+25; 5' trimmed: 1 times; 3' trimmed: 1 times\n",
            "\n",
            "Minimum overlap: 3+3\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "Overview of removed sequences at 5' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "\n",
            "\n",
            "\n",
            "Overview of removed sequences at 3' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g GCCTGGAATATGGTAGGATCTCAAC;max_error_rate=0.1...GTGCTCTGGCATATCAACATGTAAA;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setPOS_HG01243.2.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG01243.2.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.407 s (406885.386 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         1 (100.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                0 (0.0%)\n",
            "Reads written (passing filters):             1 (100.0%)\n",
            "\n",
            "Total basepairs processed:        87,021 bp\n",
            "Total written (filtered):         86,971 bp (99.9%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: GCCTGGAATATGGTAGGATCTCAAC...GTGCTCTGGCATATCAACATGTAAA; Type: linked; Length: 25+25; 5' trimmed: 1 times; 3' trimmed: 1 times\n",
            "\n",
            "Minimum overlap: 3+3\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "Overview of removed sequences at 5' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "\n",
            "\n",
            "\n",
            "Overview of removed sequences at 3' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g GCCTGGAATATGGTAGGATCTCAAC;max_error_rate=0.1...GTGCTCTGGCATATCAACATGTAAA;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setPOS_HG01361.1.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG01361.1.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.366 s (366269.588 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         1 (100.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                0 (0.0%)\n",
            "Reads written (passing filters):             1 (100.0%)\n",
            "\n",
            "Total basepairs processed:        68,661 bp\n",
            "Total written (filtered):         68,611 bp (99.9%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: GCCTGGAATATGGTAGGATCTCAAC...GTGCTCTGGCATATCAACATGTAAA; Type: linked; Length: 25+25; 5' trimmed: 1 times; 3' trimmed: 1 times\n",
            "\n",
            "Minimum overlap: 3+3\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "Overview of removed sequences at 5' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "\n",
            "\n",
            "\n",
            "Overview of removed sequences at 3' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g GCCTGGAATATGGTAGGATCTCAAC;max_error_rate=0.1...GTGCTCTGGCATATCAACATGTAAA;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setPOS_HG01361.2.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG01361.2.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.471 s (471001.625 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         1 (100.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                0 (0.0%)\n",
            "Reads written (passing filters):             1 (100.0%)\n",
            "\n",
            "Total basepairs processed:        87,104 bp\n",
            "Total written (filtered):         87,054 bp (99.9%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: GCCTGGAATATGGTAGGATCTCAAC...GTGCTCTGGCATATCAACATGTAAA; Type: linked; Length: 25+25; 5' trimmed: 1 times; 3' trimmed: 1 times\n",
            "\n",
            "Minimum overlap: 3+3\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "Overview of removed sequences at 5' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "\n",
            "\n",
            "\n",
            "Overview of removed sequences at 3' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g GCCTGGAATATGGTAGGATCTCAAC;max_error_rate=0.1...GTGCTCTGGCATATCAACATGTAAA;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setPOS_HG02148.1.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG02148.1.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.389 s (388731.241 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         0 (0.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                1 (100.0%)\n",
            "Reads written (passing filters):             0 (0.0%)\n",
            "\n",
            "Total basepairs processed:        68,667 bp\n",
            "Total written (filtered):              0 bp (0.0%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: GCCTGGAATATGGTAGGATCTCAAC...GTGCTCTGGCATATCAACATGTAAA; Type: linked; Length: 25+25; 5' trimmed: 0 times; 3' trimmed: 0 times\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g TTTACATGTTGATATGCCAGAGCAC;max_error_rate=0.1...GTTGAGATCCTACCATATTCCAGGC;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setNEG_HG02148.1.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG02148.1.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.017 s (16935.349 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         1 (100.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                0 (0.0%)\n",
            "Reads written (passing filters):             1 (100.0%)\n",
            "\n",
            "Total basepairs processed:        68,667 bp\n",
            "Total written (filtered):         68,617 bp (99.9%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: TTTACATGTTGATATGCCAGAGCAC...GTTGAGATCCTACCATATTCCAGGC; Type: linked; Length: 25+25; 5' trimmed: 1 times; 3' trimmed: 1 times\n",
            "\n",
            "Minimum overlap: 3+3\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-25 bp: 2\n",
            "\n",
            "Overview of removed sequences at 5' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "\n",
            "\n",
            "\n",
            "Overview of removed sequences at 3' end\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "25\t1\t0.0\t2\t1\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g GCCTGGAATATGGTAGGATCTCAAC;max_error_rate=0.1...GTGCTCTGGCATATCAACATGTAAA;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setPOS_HG02148.2.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG02148.2.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.394 s (394398.689 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         0 (0.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                1 (100.0%)\n",
            "Reads written (passing filters):             0 (0.0%)\n",
            "\n",
            "Total basepairs processed:        39,575 bp\n",
            "Total written (filtered):              0 bp (0.0%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: GCCTGGAATATGGTAGGATCTCAAC...GTGCTCTGGCATATCAACATGTAAA; Type: linked; Length: 25+25; 5' trimmed: 0 times; 3' trimmed: 0 times\n",
            "This is cutadapt 4.6 with Python 3.10.12\n",
            "Command line parameters: --discard-untrimmed -g TTTACATGTTGATATGCCAGAGCAC;max_error_rate=0.1...GTTGAGATCCTACCATATTCCAGGC;max_error_rate=0.1 -o /content/09_analysis//retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/setNEG_HG02148.2.fa /content/drive/MyDrive/statgen_workshop/data/workshop9/assemblies/HG02148.2.fa.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 0.014 s (14039.516 µs/read; 0.00 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:                       1\n",
            "Reads with adapters:                         0 (0.0%)\n",
            "\n",
            "== Read fate breakdown ==\n",
            "Reads discarded as untrimmed:                1 (100.0%)\n",
            "Reads written (passing filters):             0 (0.0%)\n",
            "\n",
            "Total basepairs processed:        39,575 bp\n",
            "Total written (filtered):              0 bp (0.0%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: TTTACATGTTGATATGCCAGAGCAC...GTTGAGATCCTACCATATTCCAGGC; Type: linked; Length: 25+25; 5' trimmed: 0 times; 3' trimmed: 0 times\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "ls ${analysis_dir}/retrievedRegions/hg38_chr1_109655000_109742000/fasta_files/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "v66AzHVYUrab",
        "outputId": "082a56c3-7454-4133-9822-7a3a646df826"
      },
      "execution_count": 32,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "setNEG_HG00741.1.fa\n",
            "setNEG_HG00741.2.fa\n",
            "setNEG_HG01175.1.fa\n",
            "setNEG_HG01175.2.fa\n",
            "setNEG_HG02148.1.fa\n",
            "setPOS_HG005.1.fa\n",
            "setPOS_HG005.2.fa\n",
            "setPOS_HG01243.1.fa\n",
            "setPOS_HG01243.2.fa\n",
            "setPOS_HG01361.1.fa\n",
            "setPOS_HG01361.2.fa\n"
          ]
        }
      ]
    }
  ]
}